<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN"
	"http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
<head>
	<meta http-equiv="Content-Type" content="text/html; charset=utf-8"/>
	<title></title>
</head>
<body>
	<h2>Copy number aberrations / Call copy number aberrations</h2>
	
	<h3>Description</h3>
	<p>Call copy number aberrations from segmented data (either aCGH log ratios, or log2-transformed NGS data).</p>
	
	<h3>Parameters</h3>
	<ul>
		<li>Number of copy number states (3, 4) [3]</li>
		<li>Genome build (GRCh37/hg19, NCBI36/hg18, NCBI35/hg17, NCBI34/hg16, other) [GRCh37/h19]</li>
		<li>Cellularity</li>
	</ul>
	
	<h3>Details</h3>
	<p>After normalizing and segmenting the log ratios, this tool is used to obtain discrete copy number levels with CGHcall. Calling of three copy number states means separating losses, normals and gains. Amplifications can also be detected separately as the fourth copy number state. For each data point, the frequencies of arrays exhibiting a loss/gain/amplification are also included.</p>
	<p>If available, a phenodata column specifying the cellularities (proportion of tumor cells) can be specified, which is used to correct the data.</p>
	<p>The input data must be segmented and contain these columns: chromosome, start, end.</p>
	
	<h3>Output</h3>
	<p>The inputted original and segmented log ratios are accompanied with hard copy number calls, and probabilities assigned to each call.</p>
	<p>A plot is also returned that shows the frequencies of gains (blue) and losses (red).</p>
	
	<h3>References</h3>
	<p>CGHcall:<br />
		Wiel et al. (2007) CGHcall: calling aberrations for array CGH tumor profiles. Bioinformatics 23: 892-894</p>
</body>
</html>
