<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN"
	"http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
<head>
	<meta http-equiv="Content-Type" content="text/html; charset=utf-8"/>
	<title></title>
</head>
<body>
	<h2>Copy number aberrations / Fetch probe positions from CanGEM</h2>
	
	<h3>Description</h3>
	<p>Fetches microarray probe positions from the CanGEM database.</p>
	
	<h3>Parameters</h3>
	<ul>
		<li>Accession number (...) [empty]</li>
		<li>Human genome build to use (GRCh37/hg19, NCBI36/hg18, NCBI35/hg17, NCBI34/hg16) [GRCh37/hg19]</li>
		<li>Username (...) [empty]</li>
		<li>Password (...) [empty]</li>
		<li>Session ID (...) [empty]</li>
	</ul>
	
	<h3>Details</h3>
	<p>The probe positions are fetched from the CanGEM database and appended to the input file. For a list of available platforms, please see <a href="http://www.cangem.org/browse.php?what=platforms&amp;order=Manufacturer">this list</a>. The positions have been obtained through a megablast analysis of the probe sequences as described in the referenced paper, except for Agilent miRNA arrays where the positions of the miRNA genes have been retrieved from Ensembl. For Affymetrix arrays, the sequences used are the target sequences used to design individual probes for a probe set.</p>
	
	<h3>Output</h3>
	<p>The input file appended with three new columns: chromosome, start, end.</p>
	
	<h3>References</h3>
	<p>CanGEM:<br />Scheinin et al. (2008) CanGEM: mining gene copy number changes in cancer. Nucleic Acids Res 36: D830</p>
	<p>Ensembl:<br />Flicek et al. (2010) Ensembl&apos;s 10th year. Nucleic Acids Res 38: D557-62</p>
</body>
</html>
