<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN"
	"http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
<head>
	<meta http-equiv="Content-Type" content="text/html; charset=utf-8"/>
	<title></title>
</head>
<body>
	<h2>Copy number aberrations / Import from CanGEM</h2>
	
	<h3>Description</h3>
	<p>Load a microarray data set from the CanGEM database, perform background-correction and normalization, and append chromosomal locations of the microarray probes.</p>
	
	<h3>Parameters</h3>
	<ul>
		<li>Accession number (...) [empty]</li>
		<li>Username (...) [empty]</li>
		<li>Password (...) [empty]</li>
		<li>Session ID (...) [empty]</li>
		<li>Agilent 2-color filtering (yes, no) [yes]</li>
		<li>Background treatment (none, subtract, normexp, rma) [normexp]</li>
		<li>Background offset (0, 50) [50]</li>
		<li>Intra-array normalization (none, median, loess) [loess]</li>
		<li>Inter-array normalization (none, quantile, scale) [none]</li>
		<li>Affymetrix normalization (gcrma, rma, mas5)</li>
		<li>Human genome build to use (GRCh37/hg19, NCBI36/hg18, NCBI35/hg17, NCBI34/hg16) [GRCh37/hg19]</li>
	</ul>
	
	<h3>Details</h3>
	<p>Imports data from the <a href="http://www.cangem.org/">CanGEM database</a>. The given accession number can be that of an experiment, a series, a data set, or a single microarray results. The original data files are downloaded, preprocessed and combined with chromosomal position information for the probes.</p>
	<p>For password-protected data, there are two possible ways of authenticating. The first one is to use your username/password pair. This has the downside that these will be saved in Chipster session and/or workflow files. To prevent this, also another approach is provided: First log in on the <a href="http://www.cangem.org/">CanGEM web page</a>, then notice the session ID in the lower-right corner of the page, copy this ID on the clipboard (do not include the text &quot;session ID: &quot;, only the 32-character gibberish after that), and paste it in the session parameter of this tool. This way Chipster can access the files from the database as long as you stay logged in in your web browser. After you log out, or the session expires (after 24 minutes), Chipster is no longer able to access password-protected sensitive data in CanGEM.</p>
	<p>For normalization, similar settings are provided as in the separate Affymetrix and Agilent 1 and 2-color normalization tools, so they can be consulted for more details. The only difference is that as Agilent miRNA arrays can contain multiple probes for a single miRNA, these measurements are summarized using the RMA algorithm to yield a single value for an individual miRNA. The recommended normalization settings for aCGH arrays are normexp, 50, loess, none; and for miRNA arrays none, 0, none, quantile.</p>
	<p>The optional filtering step for Agilent 2-color arrays requires the following fields to be 0 (for both red end green channels): ControlType, IsBGNonUnifOL, IsBGPopnOL, IsFeatNonUnifOL, IsFeatPopnOL, IsManualFlag, IsSaturated and SurrogateUsed, and the following to equal 1 (for both channels): IsFound, IsPosAndSignif and IsWellAboveBG.</p>
	<p>The normalized data is automatically paired with chromosomal positions for the probes, according to the selected build of the human genome. The positions have been obtained through a megablast analysis of the probe sequences as described in the referenced paper, except for Agilent miRNA arrays where the positions of the miRNA genes have been retrieved from Ensembl. For Affymetrix arrays, the sequences used are the target sequences used to design individual probes for a probe set.</p>
	
	<h3>Output</h3>
	<p>A normalized data set and a phenodata file.</p>
	
	<h3>References</h3>
	<p>CanGEM:<br />Scheinin et al. (2008) CanGEM: mining gene copy number changes in cancer. Nucleic Acids Res 36: D830</p>
	<p>For Affymetrix and Agilent normalization references, please see the respective tools.</p>
</body>
</html>
