<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN"
	"http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
<head>
	<meta http-equiv="Content-Type" content="text/html; charset=utf-8"/>
	<title></title>
</head>
<body>
	<h2>Copy number aberrations / Count overlapping CNVs</h2>
	
	<h3>Description</h3>
	<p>Counts overlapping CNVs from the database of genomic variants.</p>
	
	<h3>Parameters</h3>
	<ul>
		<li>Human genome build to use (GRCh37/hg19, NCBI36/hg18, NCBI35/hg17) [GRCh37/hg19]</li>
	</ul>
	
	<h3>Details</h3>
	<p>The list of known copy number variations (CNVs) is downloaded from the Database of Genomic Variants. For each are of interest (probes, bins, genes, or chromosomal regions) two values are calculated: the number of entries in the database that overlap with the area of interest, and the number of base pairs within known CNVs per megabase of sequence of the are of interest. These are reported in columns cnv.count and cnv.per.Mb, respectively.</p>
	
	<h3>Output</h3>
	<p>The input data table appended with columns cnv.count and cnv.per.Mb.</p>
	
	<h3>References</h3>
	<p>Iafrate et al. (2004) Detection of large-scale variation in the human genome. Nat Genet 36: 949-951</p>
</body>
</html>
