<html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=us-ascii">
</head>
<body>
<h2>Variants / Annotate variants</h2>
<h3>Description</h3>
This tool annotates the variants in a VCF file. <br>
<br>
<h3>Parameters</h3>
<ul>
<li>Genome (Human (hg38)) [Human (hg38), Human (hg19)] </li></ul>
<h3><br>
Details</h3>
<p>Variants are read from a VCF file and assigned to chromosome, genes and transcripts 
using the Bioconductor package VariantAnotation. 
Gene symbols, names and Entrez Gene IDs are taken from the organism specific annotation packages generated by the Bioconductor project.
</p>
<p>In addition, coding variants are separately treated. Data on coding variants includes their locations in the transcripts 
and proteins, inferred consequences of the variants, and specific sequences of the codons and the alleles. 
Gene symbols, names and Entrez Gene IDs are taken from the organism specific annotation packages generated by the Bioconductor project.
In additions, for the coding non-synonymous variants, the <a href="http://genetics.bwh.harvard.edu/pph2/dokuwiki/about" >PolyPhen</a>  -predictions are fetched and printed into a table.
</p>
<h3>Output</h3>
<p>The analysis output consists of the following files:</p>
<ul>
<li>all-variants.tsv: Results for all variants in the VCF file. </li>
<li>coding-variants.tsv: Results for the coding variants in the VCF file. </li>
<li>polyphen-predictions.tsv: PolyPhen predictions for the non-synonymous coding variants in the VCF file. </li>
</ul>
</body>
</html>
