<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html lang="en">
<head><title></title></head>
<body>

<h2>Matching genomic regions / Coverage BED </h2>

<h3>Description</h3>

<p>
Returns the depth and breadth of coverage of features from A on the intervals in B.
</p>

<h3>Parameters</h3>

<ul>
<li>File A is BAM format (yes, no) [no]</li>
<li>Force strandedness (yes, no) [no]</li>
<li>Report a histogram of coverage (yes, no) [no]</li> 
<li>Report the depth at each position (yes, no) [no]</li> 
<li>Treat split BAM or BED12 entries as distinct BED intervals (yes, no) [no]</li> 
</ul>

<h3>Details</h3>

<p>
coverageBed computes both the depth and breadth of coverage of features in file A across the features
in file B. For example, coverageBed can compute the coverage of sequence alignments (file A) across 1
kilobase (arbitrary) windows (file B) tiling a genome of interest. One advantage that coverageBed
offers is that it not only counts the number of features that overlap an interval in file B, it also
computes the fraction of bases in B interval that were overlapped by one or more features. Thus,
coverageBed also computes the breadth of coverage for each interval in B.
</p>
<p>
Selecting "File A is BAM format [yes]" is analogous to using "-abam" option instead of "-a" in the command line version.
</p>

<h3>Output</h3>

<p>
The output is a tab-delimited text file.
</p>

<h3>References</h3>

<p>This tool uses the BEDTools package. Please cite the article:</p>

<p>Quinlan AR and Hall IM, 2010. <a href="http://bioinformatics.oxfordjournals.org/content/26/6/841.abstract?sid=62330781-9bde-4e76-8f13-b5e323cf8034">BEDTools: a flexible suite of utilities for comparing genomic features. Bioinformatics. 26, 6, pp. 841–842.</a> </p>

<p>Please see the <a href = "http://code.google.com/p/bedtools/">BEDTools homepage</A> for more details.</p>
</body>

