<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html lang="en">
<head><title></title></head>
<body>

<h2>Matching sets of genomic regions / Increase regions in BED</h2>

<h3>Description</h3>

<p>
Add requested base pairs to increase each feature.
</p>

<h3>Parameters</h3>

<ul>
<li>The number of base pairs to subtract from the start coordinate (Integer) [0]</li>
<li>The number of base pairs to add to the end coordinate (Integer) [0]</li>
<li>Define additions based on strand (yes, no) [no]</li>
<li>Define additions as a fraction of the feature's length (yes, no) [no]</li>
<li>The fraction to subtract from the start coordinate (Decimal) [0.0]</li>
<li>The fraction to add to the end coordinate (Decimal) [0.0]</li>
</ul>

<h3>Details</h3>

<p>
This tool increases the size of each feature in a feature file be a user-defined number of bases. 
It restricts the resizing to the size of the chromosome (i.e. no start < 0 and no end > chromosome size).
</p>

<h3>Output</h3>

<p>
The output is a BED file.
</p>

<h3>References</h3>

<p>This tool uses the slopBed function of the BEDTools package. Please cite the article:</p>

<p>Quinlan AR and Hall IM, 2010. <a href="http://bioinformatics.oxfordjournals.org/content/26/6/841.abstract?sid=62330781-9bde-4e76-8f13-b5e323cf8034">BEDTools: a flexible suite of utilities for comparing genomic features. Bioinformatics. 26, 6, pp. 841–842.</a> </p>

<p>Please see the <a href = "http://code.google.com/p/bedtools/">BEDTools homepage</A> for more details.</p>
</body>

