<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html lang="en">
<head><title></title></head>
<body>

<h2>Matching genomic regions / Intersect BED</h2>

<h3>Description</h3>

<p>
Reports overlaps between two feature files such as BED, VCF and GTF. One of the files can also be BAM. Note that when A and B files are compared, the B file is loaded into memory. Therefore to minimize memory usage, one should set the smaller of the two files as the B file.
 
</p>

<h3>Parameters</h3>

<ul>
<li>Write the original entry in A for each overlap (yes, no) [no]</li>
<li>Write the original A entry once if any overlaps found in B (yes, no) [no]</li>
<li>Only report those entries in A that have no overlaps with B (yes, no) [no]</li>
<li>Write the original entry in B for each overlap (yes, no) [no]</li>
<li>Force strandedness (yes, no) [no]</li>
<li>Write the original A and B entries for overlapped A features (yes, no) [no]</li>
<li>Write the original A and B entries for all A features (yes, no) [no]</li>
<li>For each entry in A, report the number of overlaps with B (yes, no) [no]</li>
<li>Minimum overlap required as a fraction of A (0.0-1.0) [0.000000001}</li> 
<li>Require that the fraction overlap be reciprocal for A and B (yes, no) [no]</li>
<li>File A is BAM format (yes, no) [no]</li>
<li>Write uncompressed BAM output (yes, no) [no]</li>
<li>When using BAM input, write output as BED (yes, no) [no]</li>
<li>Treat split BAM or BED12 entries as distinct BED intervals (yes, no) [no]</li>
</ul>

<h3>Details</h3>

<p>
By far, the most common question asked of two sets of genomic features is whether or not any of the
features in the two sets “overlap” with one another. This is known as feature intersection. IntersectBed
allows one to screen for overlaps between two sets of genomic features. Moreover, it allows one to have
fine control as to how the intersections are reported. IntersectBed works with both BED/GFF/VCF
and BAM files as input.
</p>

<h3>Output</h3>

<p>
The output is a tab-delimited text file.
</p>


<h3>References</h3>

<p>This tool uses the BEDTools package. Please cite the article:</p>

<p>Quinlan AR and Hall IM, 2010. <a href="http://bioinformatics.oxfordjournals.org/content/26/6/841.abstract?sid=62330781-9bde-4e76-8f13-b5e323cf8034">BEDTools: a flexible suite of utilities for comparing genomic features. Bioinformatics. 26, 6, pp. 841–842.</a> </p>

<p>Please see the <a href = "http://code.google.com/p/bedtools/">BEDTools homepage</A> for more details.</p>
</body>

