<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html lang="en">
<head><title></title></head>
<body>

<h2>BWA MEM for singe end reads and own genome</h2>

<h3>Description</h3>
<p>This tool aligns single end reads to the reference genome sequence given by the user. 
The alignment is constructed using the BWA MEM algorithm. 
The genome needs to be supplied in fasta format and the reads in two fastq files.
</p>
<p>
<li>More information: <a href="http://bio-bwa.sourceforge.net/bwa.shtml">BWA manual</a></li>
</p>
<h3>Parameters</h3>

<p>
By default, the tool runs the mapping using the default parameters of BWA MEM algorithm.  
These parameters are intended for, Illumina, 454 or IonTorrent reads files where the length of the reads is more than 70 base pairs.
If your data is from PacBio sequences, you should use the <i>Data source: PacBio subreads</i> setting to switch to parameters suitable for PacBio reads.
</p>


<h3>Output</h3>

<p>As a result the tool returns a sorted and indexed BAM-formatted alignment, which is ready for viewing in the Chipster genome browser.</p>
</body>