<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html lang="en">
<head><title></title></head>
<body>

<h2>BWA-SW algorithm for single-end reads</h2>

<h3>Description</h3>

<p>Aligns single end reads to selected reference genome using the BWA_SW algorithm. The reads have to be supplied in fastq format.
</p>

<h3>Parameters</h3>

<ul>
<li>Score of a match. Corresponds to the command line parameter -a. [1].</li>
<li>Gap opening penalty. Corresponds to the command line parameter -q. [11].</li>
<li>Gap extension penalty. Corresponds to the command line parameter -r. [4].</li>
<li>Mismatch penalty. Corresponds to the command line parameter -b. [3].</li>
<li>Band width in the banded alignment. Band width in the banded alignment. Corresponds to the command line parameter -w. [33].</li>
<li>Minimum score threshold divided by the given value- Corresponds to the command line parameter -T. [37].</li>
<li>Coefficient for threshold adjustment to query length. Given an l-long query, the threshold for a hit to be retained is a*max{T,c*log(l)}. Corresponds to the command line parameter -c. [5.5].</li>
<li>Z-best heuristics. Higher -z increases accuracy at the cost of speed. Corresponds to the command line parameter -z. [1].</li>
<li>Maximum SA interval size for initiating a seed. Higher value increases accuracy at the cost of speed. Corresponds to the command line parameter -s. [3].</li>
<li>Reverse alignment limit. Minimum number of seeds supporting the resultant alignment to skip reverse alignment. Corresponds to the command line parameter -N. [5].</li>
<li>How many valid alignments are reported per read. Maximum number of alignments to report. Corresponds to the command line parameter bwa samse -n. [3].</li>
</ul>




<h3>Details</h3>

<p>This tool uses <a href="http://bio-bwa.sourceforge.net/bwa.shtml">BWA short read aligner</a> to align a set 
of FASTQ formatted sequences against a reference genome.
Aligning is performed with Burrows-Wheeler Transform based BWA-SW algorithm that allows gaps in the alignments. 
This algorithm  is designed for longer reads (&gt;200 bp) than the BWA aln algorithm. It performs  
heuristic Smith-Waterman-like alignment to find high-scoring local hits.</p>

 
<p>Note that this aligner is more accurate but significantly slower than Bowtie</p>


<h3>Output</h3>

<p>As a result the tool returns a sorted and indexed BAM-formatted alignment, which is ready for viewing in the Chipster genome browser.</p>

</body>