<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN"
	"http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
<head>
	<meta http-equiv="Content-Type" content="text/html; charset=utf-8"/>
	<title></title>
</head>
<body>
	<h2>CNA-seq / Correct for GC content</h2>
	
	<h3>Description</h3>
	<p>Takes the counts per bin for a CNA-seq data set and corrects them for GC content and mappability.</p>
	
	<h3>Parameters</h3>
	<ul>
		<li>Method (ratio, median, none) [ratio]</li>
		<li>Degree of smoothing (...) [0.65]</li>
		<li>Distribution family to use (gaussian, symmetric) [symmetric]</li>
	</ul>
	
	<h3>Details</h3>
	<p>Correcting for GC content is necessary because it affects enzyme chemistry and therefore also the depth of sequencing coverage. Mappability represents the uniqueness of sequences in the genome and also needs to be corrected for. The R package performing the correction is QDNAseq.</p>
	
	<h3>Output</h3>
	<p>The input data set with raw counts replaced by corrected ones. The data is also log2-transformed.</p>
	
	<h3>References</h3>
	<p>QDNAseq:<br />
		Scheinin et al. (2014) DNA copy number analysis of fresh and formalin-fixed specimens by shallow whole-genome sequencing with identification and exclusion of problematic regions in the genome assembly. Genome Res In Press.</p>
	<p>GC content:<br />
		Benjamini and Speed (2012) Summarizing and correcting the GC content bias in high-throughput sequencing. Nucleic Acids Res 40: e72</p>
	<p>Mappability:<br />
		Koehler et al. (2011) The uniqueome: a mappability resource for short-tag sequencing. Bioinformatics 27: 272-274</p>
</body>
</html>
