<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN"
	"http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
<head>
	<meta http-equiv="Content-Type" content="text/html; charset=utf-8"/>
	<title></title>
</head>
<body>
	<h2>CNA-seq / Define CNA-seq experiment</h2>
	
	<h3>Description</h3>
	<p>This tool takes as input a number of files with binned read counts, combines them into a single table and generates the phenodata table required for further analyses.</p>
	
	<h3>Parameters</h3>
	<ul>
		<li>Experiment type (genome=hg19, binSize=1kbp, type=SR50; genome=hg19, binSize=5kbp, type=SR50; genome=hg19, binSize=10kbp, type=SR50; genome=hg19, binSize=15kbp, type=SR50; genome=hg19, binSize=30kbp, type=SR50; genome=hg19, binSize=50kbp, type=SR50; genome=hg19, binSize=100kbp, type=SR50; genome=hg19, binSize=500kbp, type=SR50; genome=hg19, binSize=1000kbp, type=SR50) [genome=hg19, binSize=15kbp, type=SR50]</li>
		<li>Is first mate read (yes, no, any) [any]</li>
		<li>Is second mate read (yes, no, any) [any]</li>
		<li>Is not primary read (yes, no, any) [any]</li>
		<li>Is not passing quality controls (yes, no, any) [no]</li>
		<li>Is duplicate (yes, no, any) [no]</li>
		<li>Minimum mapping quality (0..n) [37]</li>
	</ul>
	
	<h3>Details</h3>
	<p>Takes as input a number of bam files, divides the specified genome build into bins of the specified size, and counts the number of reads in each bin. The implemented package is QDNAseq.</p>
	<h3>Output</h3>
	<p>Data table with read counts per sample and accompanying phenodata table.</p>
	
	<h3>References</h3>
	<p>Scheinin et al. (2014) DNA copy number analysis of fresh and formalin-fixed specimens by whole-genome sequencing: improved correction of systematic biases and exclusion of problematic regions. Manuscript submitted.</p>
</body>
</html>
