<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN"
	"http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
<head>
	<meta http-equiv="Content-Type" content="text/html; charset=utf-8"/>
	<title></title>
</head>
<body>
	<h2>CNA-seq / Filter copy number bins</h2>
	
	<h3>Description</h3>
	<p>This tool takes as input a table on binned read counts (from Define CNA-seq experiment), and filters out bins that are known to behave spuriously.</p>
	
	<h3>Parameters</h3>
	<ul>
		<li>QDNAseq blacklist (remove, keep) [remove]</li>
		<li>ENCODE blacklist (remove, keep) [remove]</li>
		<li>minimum mappability (0-100) [0]</li>
		<li>sex chromosomes (remove, keep) [remove]</li>
	</ul>
	
	<h3>Details</h3>
	<p>By default, bins in the QDNAseq or ENCODE blacklists are filtered out, as are the sex chromosomes. A minimum mappability can also be specified.</p>
	<h3>Output</h3>
	<p>Data table with the chosen bins filtered out.</p>
	
	<h3>References</h3>
	<p>Scheinin et al. (2014) DNA copy number analysis of fresh and formalin-fixed specimens by whole-genome sequencing: improved correction of systematic biases and exclusion of problematic regions. Manuscript submitted.</p>
	<p>Mappability: Koehler et al. (2011) The uniqueome: a mappability resource for short-tag sequencing. Bioinformatics 27: 272-274</p>
</body>
</html>
