<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html lang="en">
<head><title></title></head>
<body>

<h2>Statistics / Correlate miRNA with target expression</h2>

<h3>Description</h3>

<p>Given a list of miRNAs and matching gene expression array data,
 this tool identifies which of the target genes correlate positively or negatively with the miRNA expression.
 Note that the miRNA and gene expression data sets must have the same number of samples.</p>

<h3>Parameters</h3>

<ul>
<li>Order column miRNA [EMPTY] </li>
<li>Order column gene [EMPTY]</li>
<li>ID type (probe_id, entrez_id) [probe_id]</li>
<li>Correlation method (pearson, spearman, kendall) [pearson]</li>
<li>P-value threshold (0..1) [0.05]</li>
<li>P-value adjustment method (none, Bonferroni, Holm, Hochberg, BH, BY) [BH]</li>
</ul>

<h3>Details</h3>

<p>You need to have miRNA-seq and gene expression array data from the same samples, and indicate the matching samples with numbers in a given phenodata column.<br>

<br>The analysis consists of two parts:<br>

<br>Firstly, predicted target genes for the list of miRNAs are fetched from the TargetScan and PicTar databases, and only those genes that are predicted in both databases are included. 
Then, another intersection is made with the list of genes
present in the matching gene expression data set.<br>

<br>Secondly, the correlation is calculated between the expression of a miRNA and each of its target genes, using Pearson, Spearman or Kendall's measure of correlation.
Pearson correlation is parametric, so it can be sensitive to outliers, while the rank-based Spearman method is more robust. In cases when the direction
of the change between two-conditions, or time-points, is more important than the absolute values of the expression profile, Kendall's correlation method is recommended.

The statistically significant results are reported in separate tables for the positively and negatively correlating genes.  
  
</p>

<h3>Output</h3>

<p>Output consists of two text files, one listing the significantly positively correlating target genes and one listing the significantly negatively correlating ones.</p>

</p>

<h3>References</h3>
<p>This tool leverages some of the functionality from the RmiR Bioconductor package. More info on the databases of predicted miRNA targets can be found at:<br/>
<br/>targetScan<br/>
http://www.targetscan.org/
<br/><br> PicTar<br/>http://pictar.mdc-berlin.de/</p>



</body>


