<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html lang="en">
<head><title></title></head>
<body>

<h2>RNA-seq / Differential expression using Cuffdiff</h2>

<h3>Description</h3>

<p>
This tool performs differential expression analysis using the Cufflinks package version 2.0.2.
</p>

<h3>Parameters</h3>

<ul>
<li>Output type (concise, complete) [concise]</li>
<li>Annotation GTF (Human (hg19), Mouse (mm10), Mouse (mm9), Rat (rn4)) [Human (hg19)]</li>
<li>Chromosome names in my BAM file look like (chr1, 1) [chr1]</li>
<li>Allowed false discovery rate (0-1) [1]</li>
<!--  <li>Upper-quartile normalization (yes, no) [yes]</li>-->
<li>Enable multi-mapped read correction (yes,no) [no]</li>
<li>Bias correction for stranded data (yes, no) [yes]</li>
<li>Genome used for bias correction (Human (hg19), Mouse (mm10), Mouse (mm9), Rat (rn4)) [Human (hg19)]</li>
</ul>

<h3>Details</h3>

<p>
Given GTF and BAM files, Cuffdiff performs differential expression analysis of genes and transcripts using the 
Cufflinks algorithm. It is currently not possible to use replicate samples in Chipster, so you have to merge 
all samples belonging to the same experiment group into a single BAM file.
</p>
<p>
Cufflinks can detect sequence-specific bias and correct for it in abundance estimation. 
Note that bias correction works only if your data was produced with a strand specific protocol. 
</p>
<p>
By default, Cufflinks will uniformly divide each multi-mapped read to all of the positions it maps to. 
If multi-mapped read correction is enabled, Cufflinks will re-estimate the transcript abundances dividing each multi-mapped read probabilistically based on the initial abundance estimation, 
the inferred fragment length and fragment bias, if bias correction is enabled.
</p>
<p>
FDR-adjusted p-values (q-values) are calculated. The concise output files include only those genes or transcripts which have a q-value 
lower than the given FDR. The value of the Significant-column is adjusted accordingly (yes/no) in all output files.
</p>
<h3>Output</h3>

<p>
<p>The concise analysis output consists of the following files:</p>
<ul>
<li>de-genes-cufflinks.tsv:	Table containing the results of the statistical testing for differentially expressed genes, including fold change estimates and p-values.<br></li>
<li>de-genes-cufflinks.bed:  The BED version of the results table for genes contains genomic coordinates and fold change estimates for quick navigation in the Genome Browser.<br></li>
<li>de-isoforms-cufflinks.tsv: Table containing the results of the statistical testing for differentially expressed isoforms, including fold change estimates and p-values.<br></li>
<li>de-isoform-cufflinks.bed: The BED version of the results table for isoforms contains genomic coordinates and fold change estimates for quick navigation in the Genome Browser.<br></li>
<li>cufflinks.log: Contains useful information of the analysis run.</li>
</ul>
<p>If the complete output is selected, the following optional outputs may also be generated:</p>
<p>1. FPKM tracking files</p>
<ul>

<li>genes.fpkm_tracking.tsv: Gene FPKMs. Tracks the summed FPKM of transcripts sharing each gene_id</li>
<li>isoforms.fpkm_tracking.tsv: Transcript FPKMs</li>
<li>tss_groups.fpkm_tracking.tsv: Primary transcript FPKMs. Tracks the summed FPKM of transcripts sharing each tss_id</li>
<li>cds.fpkm_tracking.tsv: Coding sequence FPKMs. Tracks the summed FPKM of transcripts sharing each p_id, independent of tss_id</li>
</ul>
<p>2. Count tracking files</p>
<ul>

<li>genes.count_tracking.tsv: Gene counts. Tracks the summed counts of transcripts sharing each gene_id</li>
<li>isoforms.count_tracking.tsv: Transcript counts</li>
<li>tss_groups.count_tracking.tsv: Primary transcript counts. Tracks the summed counts of transcripts sharing each tss_id</li>
<li>cds.count_tracking.tsv: Coding sequence counts. Tracks the summed counts of transcripts sharing each p_id, independent of tss_id</li>
</ul>
<p>3. Read group tracking files</p>
<ul>
<li>genes.read_group_tracking.tsv: Gene read group tracking. Tracks the summed expression and counts of transcripts sharing each gene_id in each replicate</li>
<li>isoforms.read_group_tracking.tsv: Transcript read group tracking</li>
<li>tss_groups.read_group_tracking.tsv: Primary transcript FPKMs. Tracks the summed expression and counts of transcripts sharing each tss_id in each replicate</li>
<li>cds.read_group_tracking.tsv: Coding sequence FPKMs. Tracks the summed expression and counts of transcripts sharing each p_id, independent of tss_id in each replicate</li>

</ul>
<p>4. Differential expression tests</p>
<ul>

<li>gene_exp.diff.tsv: Gene differential FPKM. Tests differences in the summed FPKM of transcripts sharing each gene_id</li>
<li>isoform_exp.diff.tsv: Transcript differential FPKM.</li>
<li>tss_group_exp.diff.tsv: Primary transcript differential FPKM. Tests differences in the summed FPKM of transcripts sharing each tss_id</li>
<li>cds_exp.diff.tsv: Coding sequence differential FPKM. Tests differences in the summed FPKM of transcripts sharing each p_id independent of tss_id</li>
</ul>
<p>5. Differential splicing tests</p>
<ul>
<li>splicing.diff.tsv: This tab delimited file lists, for each primary transcript, the amount of overloading detected among its isoforms, i.e. how much differential splicing exists between isoforms processed from a single primary transcript. Only primary transcripts from which two or more isoforms are spliced are listed in this file.</li>
</ul>
<p>6. Differential coding output</p>
<ul>
<li>cds.diff.tsv: This tab delimited file lists, for each gene, the amount of overloading detected among its coding sequences, i.e. how much differential CDS output exists between samples. Only genes producing two or more distinct CDS (i.e. multi-protein genes) are listed here.</li>
</ul>
<p>7. Differential promoter use</p>
<ul>
<li>promoters.diff.tsv: This tab delimited file lists, for each gene, the amount of overloading detected among its primary transcripts, i.e. how much differential promoter use exists between samples. Only genes producing two or more distinct primary transcripts (i.e. multi-promoter genes) are listed here.</li>
</ul>
<p>8. Read group info</p>
<ul>
<li>read_groups.info.txt: This tab delimited file lists, for each replicate, key properties used by Cuffdiff during quantification, such as library normalization factors.</li>
</ul>
<p>8. Run info</p>
<ul>
<li>run.info.txt: This tab delimited file lists various bits of information about a Cuffdiff run to help track what options were provided.</li>
</ul>



</ul>

</p>


<h3>References</h3>

<p>This tool uses the Cufflinks package for statistical analysis. Please read the following article for more detailed information:</p>

<p>Trapnell C, Williams BA, Pertea G, Mortazavi A, Kwan G, van Baren MJ, Salzberg
SL, Wold BJ, Pachter L. <a href="http://www.ncbi.nlm.nih.gov/pubmed/20436464">Transcript assembly and quantification by RNA-Seq reveals unannotated transcripts and isoform switching during cell differentiation. Nat Biotechnol. 2010 May;28(5):511-5.</a>
</p> 

</body>

