<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html lang="en">
<head><title></title></head>
<body>

<h2>RNA-seq / Assemble reads into transcripts using Cufflinks</h2>

<h3>Description</h3>

<p>
This tool assembles reads into transcripts using version Cufflinks 2.0.2.
</p>

<h3>Parameters</h3>

<ul>
<li>Upper-quartile normalization (yes,no) [yes]</li>
<li>Enable multi-mapped read correction (yes,no) [no]</li>
<li>Bias correction for stranded data (yes, no) [yes]</li>
<li>Genome used for bias correction (Human (hg19), Mouse (mm10), Mouse (mm9), Rat (rn4)) [Human (hg19)]</li>
<li>Chromosome names in my BAM file look like (chr1, 1) [chr1]</li>
</ul>

<h3>Details</h3>

<p>
Given aligned RNA-Seq reads, Cufflinks assembles the alignments into a parsimonious set of transcripts. 
It then estimates the relative abundances of these transcripts based on how many reads support each one, 
taking into account biases in library preparation protocols. It is recommended to use the TopHat aligner 
to map your reads to the reference genome. You can merge the resulting GTFs from several samples to 
one using the Cuffmerge tool, and use it in differential expression analysis using Cuffdiff.
</p>
<p>
Upper quartile normalization can improve robustness of differential expression calls for less abundant genes and transcripts. 
It excludes very abundant genes when normalizing expression values for the number of reads in each sample by using the upper quartile of the number of fragments mapping to individual loci.
</p>
<p>
Cufflinks can detect sequence-specific bias and correct for it in abundance estimation. Note that bias correction works only if your data was produced with a strand specific protocol. </p>
<p>
By default, Cufflinks will uniformly divide each multi-mapped read to all of the positions it maps to. If multi-mapped read correction is enabled, Cufflinks will re-estimate the transcript abundances dividing each multi-mapped read probabilistically based on the initial abundance estimation, the inferred fragment length and fragment bias, if bias correction is enabled.

</p>

<h3>Output</h3>

<p>The analysis output consists of the following:</p>
<ul>
<li>transcripts.gtf: This GTF file contains Cufflinks' assembled isoforms. The first 7 columns are standard GTF, and the last column contains attributes, some of which are also standardized ("gene_id", and "transcript_id"). There is one GTF record per row, and each record represents either a transcript or an exon within a transcript.</li>
<li>genes.fpkm_tracking.tsv: This file contains the estimated gene-level expression values in the generic FPKM Tracking Format. Note, however that as there is only one sample, the "q" format is not used.</li>  
<li>isoforms.fpkm_tracking.tsv: This file contains the estimated isoform-level expression values in the generic FPKM Tracking Format. Note, however that as there is only one sample, the "q" format is not used.</li> 
<li>skipped.gtf: This file contains the skipped loci.</li>
</ul>
</p>


<h3>References</h3>

<p>This tool uses the Cufflinks package for statistical analysis. Please read the following article for more detailed information:</p>

<p>Trapnell C, Williams BA, Pertea G, Mortazavi A, Kwan G, van Baren MJ, Salzberg
SL, Wold BJ, Pachter L. <a href="http://www.ncbi.nlm.nih.gov/pubmed/20436464">Transcript assembly and quantification by RNA-Seq reveals unannotated transcripts and isoform switching during cell differentiation. Nat Biotechnol. 2010 May;28(5):511-5.</a>
</p> 

</body>

