<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html lang="en">
<head><title></title></head>
<body>

<h2>Retrieve listed entries from public databases</h2>

<h3>Description</h3>

<p>
This tool retrieves sequence files or other database entries from public databases.
The available databases are:
<li>UniProt</li>
<li>EMBL</li>
<li>Ensembl Gene</li>
<li>Ensembl Transcript</li>
<li>Ensembl Genomes Gene</li>
<li>Ensembl Genomes Transcript</li>
<li>InterPro</li>
<li>MEDLINE</li>
<li>PDB</li>
<li>RefSeq nucleoide</li>
<li>RefSeq protein</li>
<li>Taxonomy</li>
<li>Trace Archive</li>

</p>

<h3>Parameters</h3>
<li><b>Database</b> database to be used.</li>

<h3>Input file</h3>
<p>The input file should be a list of database entries or names to be retrieved 
from the selected database. Each entry name should be on a separate 
row in the entry name file. For example, a file containing following accession 
numbers would retrieve five protein sequences from UniProt database:
<pre>
A2A1A1
A5ABL2
G7J032
O04298
O24248
</pre></p>
<p>
Note that this tool retrieves data only based on the entry names or accession numbers. 
However you can use wildcard character (*) to retrieve several entries that share 
similar identities. For example in the case of UniProt database name <i>ABC*</i> 
would retrieve all the UniProt sequences that whose name start with ABC. 
</p>
<p>
Only the first word of each row will be used as a sequence identifier and rest 
of the line is ignored. Thus the sequence name lines may contain comments and other information
after the sequence ID. Also a table, where the first column contains the sequence name or ID, 
can be used as a name list.</p>  
<p>
For example following table could be use to retrieve the full database entries of the
listed five proteins from the UniProt database.
<pre>
<table width="519" border="1"> 
<tr><td>NCS2_COPJA</td><td>Coptis japonica (Japanese goldthread)</td></tr>
<tr><td>INUA_ASPNC</td><td>Aspergillus niger (strain CBS 513.88 / FGSC A1513)</td></tr>
<tr><td>PHBP_MEDTR</td><td>Medicago truncatula (Barrel medic) (Medicago tribuloides)</td></tr>
<tr><td>DAU1_DAUCA</td><td>Daucus carota (Wild carrot)</td></tr>
<tr><td>PRU1_PRUAV</td><td>Prunus avium (Cherry) (Cerasus avium)</td></tr>
</table>
</pre>
</p>

<h3>Output</h3>

<p>
Output is a text file. The format depends on the database selected. In the case of 
sequence databases, a full database entry, that contains all the annotation, is retrieved.
This data can be further translated into FASTA format with the 
<a href="http://chipster.csc.fi/manual/seqret.html">Sequence format conversion tool</a>
, if needed. 
</p>


</body>

