<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html lang="en">
<head><title></title></head>
<body>

<h2>RNA-seq / Differential expression analysis using DESeq</h2>

<h3>Description</h3>
Differential expression analysis using the DESeq Bioconductor package. If you have more than 2 experimental groups, please use DESeq2. 
Both DESeq can accommodate a second experimental factor like pairing.

<br>
<br>
<h3>Parameters</h3>

<ul>
 
<li> Column describing groups [group]</li>
<li> Column describing additional experimental factor [EMPTY]</li>
<li> Apply normalization (yes, no) [yes]</li>
<li> Dispersion estimation method (parametric, local) [parametric] </li>
<li> Use fitted dispersion values (when higher than original values, always) [when higher than original values]</li>
<li> Multiple testing correction (none, Bonferroni, Holm, Hochberg, BH, BY) [BH]</li>
<li> P-value cutoff (0-1) [0.05]</li>
<li> Plot width (200-3200 [600]</li>
<li> Plot height (200-3200) [600]</li>
</ul>

<h3>Details</h3>


<br>This tool takes as input a table of raw counts from the different samples. The count file has to be associated with a phenodata file describing the experimental groups. 
These files are best created by the tool "Utilities / Define NGS experiment", which combines count files for different samples to one table, 
and creates a phenodata file for it. You can perform the analysis with DESeq or DESeq2, the former being more conservative.
<br><br>
When normalization is enabled, size factors for each sample are calculated the following way: Geometric mean is calculated for each gene across all samples. 
The counts for a gene in each sample is then divided by this mean. The median of these ratios in a sample is the size factor for that sample. This procedure
 corrects for RNA composition bias (which can arise for example when only a small number of genes are very highly expressed in one experiment condition but not in the other). 
<br><br>
In addition to the experimental groups, you can give an additional factor like pairing to the analysis. If given, p-values in the output table are 
from a likelihood ratio test of a model including the experimental groups and experimental factor vs a model which only includes the experimental factor.

<br><br>
A dispersion value is estimated for each gene through a model fit procedure, which can be performed in a "local" or "parametric" mode.
The local mode is more robust and should be used when there are no replicates.
Users can select to replace the original dispersion values by the fitted ones always, or only when the fitted value is higher than the original one 
(more conservative option).
<br><br>
You need to have biological replicates of each experiment condition in order to estimate dispersion properly. 
If you have biological replicates only for one condition, DESeq will estimate dispersion using the replicates of that single 
condition. If there are no replicates at all, DESeq will estimate dispersion using the samples from the different conditions as replicates.


<h3>Output</h3>

<p>The analysis output consists of the following files:</p>

<ul>
<li>de-list-deseq.tsv: Table containing the results of the statistical testing, including fold change estimates and p-values.<br></li>
<li>de-list-deseq.bed: The BED version of the results table contains genomic coordinates and log2 fold change values.<br></li>
<li>deseq_report.pdf: A PDF file containing:
	<ul>
	<li>A scatter plot where the significantly differentially expressed genes are highlighted. <br></li>
	<li>Plot of dispersion estimates as a function of the counts values, with the fitted model overlaid. <br></li>
	<li>Plot of the raw and adjusted p-value distributions of the statistical test. <br></li>
	</ul></li>
</ul>

<h3><br>References</h3>

<p>This tool uses the DESeq and DESeq2 packages for statistical analysis. Please read the following article for more detailed information:</p>

<p> S Anders and W Huber: <a href="http://genomebiology.com/2010/11/10/R106">Differential expression analysis for sequence count data. Genome Biology 2010, 11:R106.</a></p> 
<p> M Love, W Huber and S Anders: <a href="http://biorxiv.org/content/biorxiv/early/2014/02/19/002832.full.pdf">Moderated estimation of fold change and
dispersion for RNA-Seq data with DESeq2.</a></p> 

</body>
