<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html lang="en">
<head><title></title></head>
<body>

<h2>RNA-seq / Differential expression analysis using DESeq</h2>

<h3>Description</h3>
Differential expression analysis using the exact test of the DESeq Bioconductor package ("nbinomTest"). 
Please note that this tool is suitable only for two group comparisons. 
For multifactor experiments you can use the tool "Differential expression using edgeR for multivariate experiments", 
which uses generalized linear models -based statistical methods ("glm edgeR").

<br>
<br>
<h3>Parameters</h3>

<ul>
<li> Column describing groups [group]</li>
<li> Apply normalization (yes, no) [yes]</li>
<li> Dispersion estimation method (parametric, local) [local] </li>
<li> Use fitted dispersion values (when higher than original values, always) [when higher than original values]</li>
<li> Multiple testing correction (none, Bonferroni, Holm, Hochberg, BH, BY) [BH]</li>
<li> P-value cutoff (0-1) [0.05]</li>
<li> Plot width (200-3200 [600]</li>
<li> Plot height (200-3200) [600]</li>
</ul>

<h3><br>Details</h3>

<br>
<br>This tool takes as input a table of raw counts from the different samples. The count file has to be associated with a phenodata file describing the experimental groups. 
These files are best created by the tool "Utilities / Define NGS experiment", which combines count files for different samples to one table, and creates a phenodata file for it.
<br><br>
<br><br>
When normalization is enabled, size factors are calculated by summing the counts for each sample, or 
using the library size given by the user in the phenodata.tsv. The former allows to correct for RNA composition bias
 (which can arise for example when only a small number of genes are very highly expressed in one experiment condition but not in the other). 
<br><br>
A dispersion value is estimated for each gene through a model fit procedure, which can be performed in a "local" or "parametric" mode.
The former is more robust, but users are encouraged to experiment with the setting to optimize results.
Users can select to replace the original dispersion values by the fitted ones always, or only when the fitted value is higher than the original one 
(more conservative option).
<br><br>
You need to have biological replicates of each experiment condition in order to estimate dispersion properly. 
If you have biological replicates only for one condition, DESeq will estimate dispersion using the replicates of that single 
condition. If there are no replicates at all, DESeq will estimate dispersion using the samples from the different conditions as replicates.
<br><br>
Statistical testing is performed using a negative binomial test.

<h3>Output</h3>

<p>The analysis output consists of the following files:</p>

<ul>
<li>de-list-deseq.tsv: Table containing the results of the statistical testing, including fold change estimates and p-values.<br></li>
<li>de-list-deseq.bed: The BED version of the results table contains genomic coordinates and log2 fold change values.<br></li>
<li>ma-plot-deseq.pdf: A scatter plot where the significantly differentially expressed genes are highlighted. <br></li>
<li>dispersion-plot.pdf: Plot of dispersion estimates as a function of the counts values, with the fitted model overlaid. <br></li>
<li>p-value-plot-edger.pdf: Plot of the raw and adjusted p-value distributions of the statistical test. <br></li>
</ul>

<h3><br>References</h3>

<p>This tool uses the DESeq package for statistical analysis. Please read the following article for more detailed information:</p>

<p> S Anders and W Huber: <a href="http://genomebiology.com/2010/11/10/R106">Differential expression analysis for sequence count data. Genome Biology 2010, 11:R106.</a></p> 

</body>
