<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html lang="en">
<head><title></title></head>
<body>

<h2>Plot normalized counts for a gene</h2>

<h3>Description</h3>
Given a table of read counts for an experiment, this tool makes a plot of the normalized counts using the DESeq2 Bioconductor package. 

<br>
<br>
<h3>Parameters</h3>

<ul>
<li> Gene name</li>
<li> Show names in plot (yes, no) [no]</li>
</ul>

<h3>Details</h3>


<p>This tool takes as input a table of raw counts. The count table has to be associated with a phenodata file 
describing the experimental groups. These files are best created by the tool "Utilities / Define NGS experiment", 
which combines count files for different samples to one table, and creates a phenodata file for it.
</p>
<p>The tool normalizes counts by sequencing depth and adds a pseudocount of 1/2 to allow for log scale plotting. 
The counts are grouped by "group". 
</p>


<h3>Output</h3>

<p>The plot of normalized counts.
</p>

<h3><br>References</h3>

<p>This tool uses the DESeq2 package. Please read the following article for more detailed information:</p>

<p> M Love, W Huber and S Anders: <a href="http://biorxiv.org/content/biorxiv/early/2014/02/19/002832.full.pdf">Moderated estimation of fold change and
dispersion for RNA-Seq data with DESeq2.</a></p> 

</body>
