<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html lang="en">
<head><title></title></head>
<body>

<h2>Differential expression analysis using DESeq2</h2>

<h3>Description</h3>
Differential expression analysis using the DESeq2 Bioconductor package. This tool allows you to have more than two experimental groups and 
account for a second experimental factor.

<br>
<br>
<h3>Parameters</h3>

<ul>
 
<li> Column describing groups [group]</li>
<li> Column describing additional experimental factor [EMPTY]</li>
<li> Cutoff for the adjusted P-value (0-1) [0.05]</li>
</ul>

<h3>Details</h3>


<br>This tool takes as input a table of raw counts. The count table has to be associated with a phenodata file describing the experimental groups. 
These files are best created by the tool "Utilities / Define NGS experiment", which combines count files for different samples to one table, 
and creates a phenodata file for it.
<br><br>
DESeq2 performs an internal normalization where geometric mean is calculated for each gene across all samples. 
The counts for a gene in each sample is then divided by this mean. The median of these ratios in a sample is the size factor for that sample. This procedure
 corrects for library size and RNA composition bias, which can arise for example when only a small number of genes are very highly expressed in one experiment condition but not in the other. 
<br><br>
DESeq2 fits negative binomial generalized linear models for each gene and uses Wald test for significance testing. 
As small numbers of replicates make it impossible to estimate within-group variance reliably, DESeq2 uses shrinkage estimation for dispersions and fold changes. 
A dispersion value is estimated for each gene through a model fit procedure.
You need to have biological replicates of each experiment condition in order to estimate dispersion properly. 
If there are no replicates, DESeq will estimate dispersion using the samples from the different conditions as replicates. 
<br><br>
DESeq2 detects automatically count outliers using Cooks's distance and removes these genes from analysis. It also automatically removes genes whose mean of normalized counts is below a threshold 
determined by an optimization procedure. Removing these genes with low counts improves the detection power by making the multiple testing adjustment of the p-values less severe.
<br><br>
In addition to the group information, you can give an additional experimental factor like pairing to the analysis. If given, p-values in the output table are 
from a likelihood ratio test of a model including the groups and the additional factor vs a model which only includes the additional factor.
<br><br>




<h3>Output</h3>

<p>The analysis output consists of the following files. Note that if you have more than two experimental groups, the output figures sum up information from all pairwise comparisons.</p>

<ul>
<li>de-list-deseq2.tsv: Table containing the significantly differentially expressed genes. The columns include 
	<ul>
		<li>baseMean = the average of the normalized counts taken over all samples</li>
		<li>log2FoldChange = log2 fold change between the groups. E.g. value 2 means that the expression has increased 4-fold</li>
		<li>lfcSE = standard error of the log2FoldChange estimate</li>
		<li>stat = Wald statistic</li>
		<li>pvalue = Wald test p-value</li>
		<li>padj = Benjamini-Hochberg adjusted p-value</li>
	</ul></li>
<li>de-list-deseq2.bed: The BED version of the results table contains genomic coordinates and log2 fold change values.<br></li>
<li>summary.txt: Textual summary of the differential expression results, including information on filtering and outliers. <br></li>
<li>deseq2_report.pdf: A PDF file containing:
	<ul>
	<li>MA scatter plot where the significantly differentially expressed genes are highlighted. <br></li>
	<li>Plot of dispersion estimates at different count levels, showing
		<ul>
		<li>black dot = dispersion estimate for each gene as obtained by considering the information from each gene separately</li>
		<li>red line = fitted estimates showing the dispersions' dependence on the mean</li>
		<li>blue dot = the final dispersion estimates shrunk from the gene-wise estimates towards the fitted estimates. The values are used in the statistical testing. </li>
		<li>blue circles = genes which have high gene-wise dispersion estimates and are hence labelled dispersion outliers and not shrunk toward the fitted trend line</li>
		</ul></li>
	<li>Plot of the raw and adjusted p-value distributions of the statistical test. <br></li>
	</ul></li>
</ul></li>

<h3><br>References</h3>

<p>This tool uses the DESeq2 package. Please read the following article for more detailed information:</p>

<p> M Love, W Huber and S Anders: <a href="http://biorxiv.org/content/biorxiv/early/2014/02/19/002832.full.pdf">Moderated estimation of fold change and
dispersion for RNA-Seq data with DESeq2.</a></p> 

</body>
