<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html lang="en">
<head><title></title></head>
<body>

<h2>RNA-seq / Count aligned reads per exons for DEXSeq</h2>

<h3>Description</h3>

<p>
Given mapped reads in a BAM file, this tool counts the reads that fall into each non-overlapping exonic part using the script dexseq-count.py.
</p>

<h3>Parameters</h3>

<ul>
<li>Does the alignment file contain paired-end data (yes, no) [no]</li>
<li>Reference organism []</li>
<li>Chromosome names in the BAM file look like (chr1, 1) [1]</li>
<li>Does the BAM file contain paired-end data (yes, no) [no]</li>
<li>Was the data produced with a strand-specific protocol (yes, no, reverse) [no]</li>
<li>Mode to handle reads overlapping more than one feature (union, intersection-strict, intersection-nonempty) [union]</li>

</ul>

<h3>Details</h3>

<p> 
Indicate if your data was not produced with a strand-specific RNA-seq protocol, so that a read is considered overlapping with a feature regardless of whether it is mapped to the same or the opposite strand as the feature. 
Otherwise the read has to be mapped to the same strand as the feature.
</p>

<h3>Output</h3>

<p>
Output is a table with counts for each non-overlapping exonic part. In order to use the output in DEXSeq, you need to select all samples and run the tool "Utilities - Define NGS experiment".
</p>

<h3>References</h3>

<p>
This tool is based on the <a href = "http://www-huber.embl.de/users/anders/HTSeq/doc/index.html">HTSeq package</a> by Simon Anders and it is available in the DEXSeq Bioconductor package.</p>

</body>

