<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html lang="en">
<head><title></title></head>
<body>

<h2>RNA-seq / Map aligned reads to exons for DEXSeq</h2>

<h3>Description</h3>

<p>
Given mapped reads in a BAM file, this tool counts the reads that fall into each non-overlapping exonic part using the script dexseq-count.py.
</p>

<h3>Parameters</h3>

<ul>
<li>Does the alignment file contain paired-end data (yes, no) [no]</li>
<li>Organism (human hg19.68, mouse mm10.68, mouse mm9.62, rat rn4.68) [human hg19.68]</li>
</ul>

<h3>Details</h3>

<p> 
Note that the chromosome names in your BAM file need to the chr prefix (e.g. chr1).
</p>

<h3>Output</h3>

<p>
Output is a table with counts for each non-overlapping exonic part. In order to use the output in DEXSeq, you need to select all samples and run the tool "Utilities - Define NGS experiment".
</p>

<h3>References</h3>

<p>
This tool is based on the <a href = "http://www-huber.embl.de/users/anders/HTSeq/doc/index.html">HTSeq package</a> by Simon Anders and it is available in the DEXSeq Bioconductor package.</p>

</body>

