<html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=us-ascii">
</head>
<body>
<h2>Differential exon expression using DEXSeq.</h2>
<h3>Description</h3>
Infers differential exon usage from RNA-seq data using the Bioconductor package DEXSeq.
<br>
<br>
<h3>Parameters</h3>
<ul>
<li>Reference organism </li>
<li>Threshold for adjusted p-value (0-1)[0.05]</li>
<li>Common dispersion (0-100) [0.1] </li>
</ul>
<br>
<h3>Details</h3>

<p>Prior to running this tool, you need to count reads in your BAM files per exon (or non-overlapping exonic bins) using the tool "Count aligned reads per exons for DEXSeq". This produces a count file for each sample.
The individual count files need to be combined to a count table using the tool "Utilities / Define NGS experiment", which produces also a phenodata file.
Once you have marked the experimental groups with numbers in the group column of the phenodata, the count table can be used by DEXSeq to analyze differential exon expression. 
Please note that you need at least two and preferably more biological replicates for each group.</p>

<p>As small numbers of replicates make it impossible to estimate within-group variance reliably, DESeq2 uses shrinkage estimation for dispersions by borrowing information from other exons that are expressed at the same level.
You need to have biological replicates of each experimental group in order to estimate dispersion properly. If dispersion values cannot be estimated, the given common dispersion value is used for all exons. 
In this case no graphical output is generated.</p>

<p>p-values are corrected for multiple testing using the Benjamini-Hochberg method. If a gene has at least one exon below the given threshold, all its exons will be included in the result list.</p>

<h3>Output</h3>
<p>The analysis output consists of the following files:</p>
<ul>

<li>dexseq-genes-with-significant-exons.tsv: Table containing the results of the statistical testing, including fold change estimates and p-values, for genes which contain differentially expressed exons.</li>
<li>dexseq-exons.pdf: Visualization of genes which contain differentially expressed exons.</li>
<li>dexseq-MAplot.pdf: Log2 fold change plotted against log2 mean normalized counts.</li>
<li>dexseq-dispersion-plot.pdf: Dispersion plot.</li>
</ul>
<h3>References</h3>
Anders S, Reyes A, Huber W. Detecting differential usage of exons from RNA-seq data. Genome Res. 2012 Sep 5.
</body>
</html>