<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html lang="en">
	<head>
		<title>DimontPredictor</title>
	</head>
	<body>
		<h2>DimontPredictor</h2>
		<h3>Description</h3>
		<p>DimontPredictor allows for predicting binding sites in new data using a previously trained Dimont model. For training a Dimont model see tool &quot;Dimont&quot;.</p>

		<h3>Inputs</h3>
		<ul>
			<li>A Dimont model as available from the output of a Dimont run.</li>
			<li>An annotated FastA file with sequences, e.g., from ChIP-seq experiments; can be generated using the &quot;Dimont data extractor&quot; tool.</li>
		</ul>

		<h3>Parameters</h3>
			<ul>
				<li><b>p-value:</b> The maximum p-value allowed for predicted binding sites.</li>
				<li><b>Value tag:</b> The tag for the value information in the FastA-annotation of the input file, default as generated by &quot;Dimont data extractor&quot;.</li>
				<li><b>Weighting factor:</b> The value for weighting the data, a value between 0 and 1. Recommended values: 0.2 for ChIP-seq/ChIP-exo, 0.01 for PBM data.</li>
			</ul>
		<h3>Details</h3>

		<p>This tool may be useful if you, for instance, want to predict binding sites of a previously discovered motifs in other data sets, or if you want to try different p-values for filtering predictions.</p>

<p>Input sequences must be supplied in an annotated FastA format as generated using the &quot;Dimont sequence extractor&quot; tool. In the annotation of each sequence, you need to provide a value that reflects the confidence that this sequence is bound by the factor of interest. Such confidences may be peak statistics (e.g., number of fragments under a peak) for ChIP data or signal intensities for PBM data.</p>

<p>For instance, an annotated FastA file for ChIP-seq data comprising sequences of length 1000 centered around the peak summit could look like:
<br /><br /><code>
> peak: 500; signal: 515<br />
ggccatgtgtatttttttaaatttccac...<br />
> peak: 500; signal: 199<br />
GGTCCCCTGGGAGGATGGGGACGTGCTG...<br />
...<br />
</code><br />
where the anchor point is given as 500 for the first two sequences, and the confidence amounts to 515 and 199, respectively. The FastA comment may contain additional annotations of the format <code>key1 : value1; key2: value2;....</code>code></p>

<p>Accordingly, you would need to set the parameter &quot;Value tag&quot; to <code>signal</code>code> for the input file.</p>

<p>The parameter &quot;Weighting factor&quot; defines the proportion of sequences that you expect to be bound by the targeted factor with high confidence. For ChIP data, the default value of 0.2 typically works well. For PBM data, containing a large number of unspecific probes, this parameter should be set to a lower value, e.g. 0.01.</p>

<p>The parameter "p-value" defines a threshold on the p-values of predicted binding sites, and only binding sites with a lower p-value are reported by DimontPredictor. The Dimont tool uses a p-value threshold of 1E-3, which is also the default value of DimontPredictor.</p>
		<h3>Output</h3>
		<ul>
			<li><code>dimont-predictor-log.txt</code>: <b>Logfile</b>, logfile of the DimontPredictor run.</li>
			<li><code>dimont-predictor-predictions.txt</code>: <b>Predictions</b>, binding sites predicted by DimontPredictor.</li>
			<li><code>dimont-predictor-logo-rc.png</code>: <b>Sequence logo (rc)</b>, The sequence logo of the reverse complement of the predictions.</li>
			<li><code>dimont-predictor-logo.png</code>: <b>Sequence logo</b>, the sequence logo of the predictions.</li>
		</ul>
		<h3>Reference</h3>

<p>If you use Dimont, please cite<br /><br />

J. Grau, S. Posch, I. Grosse, and J. Keilwagen. A general approach for discriminative de-novo motif discovery from high-throughput data. <em>Nucleic Acids Research</em>, 41(21):e197, 2013.
</p>
	</body>
</html>
