<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html lang="en">
	<head>
		<title>Dimont</title>
	</head>
	<body>
		<h2>Dimont</h2>
		<h3>Description</h3>
		<p>Dimont is a universal tool for de-novo motif discovery. Dimont has successfully been applied to ChIP-seq, ChIP-exo and protein-binding microarray (PBM) data.</p>

		<h3>Inputs</h3>
		<ul>
			<li>An annotated FastA file with sequences, e.g., from ChIP-seq experiments; can be generated using the &quot;Dimont data extractor&quot; tool.</li>
		</ul>

		<h3>Parameters</h3>
			<ul>
				<li><b>Position tag:</b> The tag for the position information in the FastA-annotation of the input file, default as generated by &quot;Dimont data extractor&quot;.</li>
				<li><b>Value tag:</b> The tag for the value information in the FastA-annotation of the input file, default as generated by &quot;Dimont data extractor&quot;.</li>
				<li><b>Standard deviation:</b> The standard deviation of the position distribution centered at the position specified by the position tag.</li>
				<li><b>Weighting factor:</b> The value for weighting the data, a value between 0 and 1. Recommended values: 0.2 for ChIP-seq/ChIP-exo, 0.01 for PBM data.</li>
				<li><b>Starts:</b> The number of pre-optimization runs. Default value is fine for most applications.</li>
				<li><b>Initial motif width:</b> The motif width that is used initially, may be adjusted during optimization.</li>
				<li><b>Markov order of motif model:</b> The Markov order of the model for the motif. A value of 0 (default) specifies a position weight matrix (PWM) or position-specific scoring matrix (PSSM), a value of 1 specifies a weight array matrix (WAM) model.</li>
				<li><b>Markov order of background model:</b> The Markov order of the model for the background sequence and the background sequence, -1 defines uniform distribution.</li>
				<li><b>Equivalent sample size:</b> Reflects the strength of the prior on the model parameters. Default value is fine for most applications.</li>
				<li><b>Delete BSs from profile:</b> A switch for deleting binding site positions of discovered motifs from the profile before searching for futher motifs.</li>
			</ul>
		<h3>Details</h3>

		<p>Input sequences must be supplied in an annotated FastA format as generated using the &quot;Dimont data extractor&quot; tool. In the annotation of each sequence, you need to provide a value that reflects the confidence that this sequence is bound by the factor of interest. Such confidences may be peak statistics (e.g., number of fragments under a peak) for ChIP data or signal intensities for PBM data. In addition, you need to provide an anchor position within the sequence. In case of ChIP data, this anchor position could for instance be the peak summit. For instance, an annotated FastA file for ChIP-exo data comprising sequences of length 100 centered around the peak summit could look like:
<br /><br /><code>
> peak: 50; signal: 515<br />
ggccatgtgtatttttttaaatttccac...<br />
> peak: 50; signal: 199<br />
GGTCCCCTGGGAGGATGGGGACGTGCTG...<br />
...<br />
</code><br />
where the anchor point is given as 50 for the first two sequences, and the confidence amounts to 515 and 199, respectively. The FastA comment may contain additional annotations of the format <code>key1 : value1; key2: value2;....</code>code></p>

<p>Accordingly, you would need to set the parameter &quot;Position tag&quot; to <code>peak</code> and the parameter &quot;Value tag&quot; to <code>signal</code> for the input file.</p>

<p>For the standard deviation of the position prior, the initial motif length and the number of pre-optimization runs, we provide default values that worked well in our studies on ChIP and PBM data. However, you may want adjust these parameters to meet your prior information.</p>

<p>The parameter &quot;Markov order of the motif model&quot; sets the order of the inhomogeneous Markov model used for modeling the motif. If this parameter is set to 0, you obtain a position weight matrix (PWM) model. If it is set to 1, you obtain a weight array matrix (WAM) model. You can set the order of the motif model to at most 3.</p>

<p>The parameter &quot;Markov order of the background model&quot; sets the order of the homogeneous Markov model used for modeling positions not covered by a motif. If this parameter is set to -1, you obtain a uniform distribution, which worked well for ChIP data. For PBM data, orders of up to 4 resulted in an increased prediction performance in our case studies. The maximum allowed value is 5.</p>

<p>The parameter &quot;Weighting factor&quot; defines the proportion of sequences that you expect to be bound by the targeted factor with high confidence. For ChIP data, the default value of 0.2 typically works well. For PBM data, containing a large number of unspecific probes, this parameter should be set to a lower value, e.g. 0.01.</p>

<p>The &quot;Equivalent sample size&quot; reflects the strength of the influence of the prior on the model parameters, where higher values smooth out the parameters to a greater extent.</p>

<p>The parameter &quot;Delete BSs from profile&quot; defines if BSs of already discovered motifs should be deleted, i.e., &quot;blanked out&quot;, from the sequence before searching for futher motifs.</p>

		<h3>Output</h3>
		<ul>
<li><code>dimont-log.txt</code>: <b>Logfile</b>, logfile of the Dimont run.</li>
<li><code>dimont-predictions-*.txt</code>: <b>Predictions</b>, binding sites predicted by Dimont.</li>
<li><code>dimont-logo-rc-*.png</code>: <b>Sequence logo (rc)</b>, the sequence logo of the reverse complement of the motif discovered by Dimont.</li>
<li><code>dimont-logo-*.png</code>: <b>Sequence logo</b>, the sequence logo of the motif discovered by Dimont.</li>
<li><code>dimont-model-*.xml</code>: <b>Dimont model</b>, the model (as XML) of the motif discovered by Dimont. Can be used in &quot;Dimont Predictor&quot;.</li>

		</ul>
		<h3>Reference</h3>

<p>If you use Dimont, please cite<br /><br />

J. Grau, S. Posch, I. Grosse, and J. Keilwagen. A general approach for discriminative de-novo motif discovery from high-throughput data. <em>Nucleic Acids Research</em>, 41(21):e197, 2013.
</p>
	</body>
</html>