<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html lang="en">
	<head>
		<title>Dimont sequence extractor using own genome</title>
	</head>
	<body>
		<h2>Dimont sequence extractor using own genome</h2>
		<h3>Description</h3>
		<p>Dimont sequence extractor prepares an annotated FastA file as required by Dimont from a genome (in FastA format) and a tabular file (e.g., BED, GTF, narrowPeak,...).</p>
		<p>This version of the tool allows to use custom genomes.</p>

		<h3>Inputs</h3>
		<ul>
			<li>The genomic regions to be extracted in a BED-like file format, e.g., BED, GTF, narrowPeak.</li>
			<li>The input genome to which the genomic regions refer, single FastA file.</li>
		</ul>

		<h3>Parameters</h3>
			<ul>
				<li><b>Chromosome column:</b> The column of the Regions file, which contains the chromosome information.</li>
				<li><b>Start column:</b> The column of the Regions file containing the start position of the genomic region.</li>
				<li><b>Second coordinate:</b> The second genomic coordinate with meaning specified by parameter &quot;Meaning of second coordinate&quot;.</li>
				<li><b>Meaning of second coordinate:</b> The meaning of the second genomic coordinate. This may either be the position of the peak summit relative to the position in Start, or the end position of the peak.</li>
				<li><b>Statistics column:</b> The column containing the peak statistics information (or another measure of peak confidence).</li>
				<li><b>Width:</b> The width of the genomic region to be extracted. Recommended values: 1000 for ChIP-seq and 100 for ChIP-exo.</li>
				<li><b>Regions file has row names:</b> Select &quot;yes&quot; if the Regions file has row names in its first column, e.g., is in a general TSV-like format.</li>
			</ul>
		<h3>Details</h3>

		<p>The regions specified in the tabular file are used to determine the center of the extracted sequences. All extracted sequences have the same length as specified by parameter &quot;Width&quot;.</p>

<p>In case of ChIP data, the center position could for instance be the peak summit. An annotated FastA file for ChIP-seq data comprising sequences of length 1000 centered around the peak summit might look like:
<br /><br /><code>
> peak: 500; signal: 515<br />
ggccatgtgtatttttttaaatttccac...<br />
> peak: 500; signal: 199<br />
GGTCCCCTGGGAGGATGGGGACGTGCTG...<br />
...<br />
</code><br />
where the anchor point is given as 500 for the first two sequences, and the confidence amounts to 515 and 199, respectively.</p>

		<h3>Output</h3>
		<ul>
			<li><code>extracted.fa</code>: <b>Extracted sequences</b>, the sequences extracted from the given genome using the supplied region specifications.
		</ul>
		<h3>Reference</h3>

<p>If you use Dimont, please cite<br /><br />

J. Grau, S. Posch, I. Grosse, and J. Keilwagen. A general approach for discriminative de-novo motif discovery from high-throughput data. <em>Nucleic Acids Research</em>, 41(21):e197, 2013.
</p>
	</body>
</html>
