<html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=us-ascii">
</head>
<body>
<h2>RNA-seq / Differential expression using edgeR for multivariate experiments</h2>
<h3>Description</h3>
Differential expression analysis for multifactor experiments using the generalized linear models (glm) -based statistical methods of the edgeR Bioconductor package.
<br>
<br>
<h3>Parameters</h3>
<ul>
<li>Apply TMM normalization (yes, no) [yes] </li>
<li>Main effect 1 </li>
<li>Main effect 2 </li>
<li>Main effect 3 </li>
<li>Treat main effect 1 as factor (yes, no) [yes] </li>
<li>Treat main effect 2 as factor (yes, no) [yes] </li>
<li>Treat main effect 3 as factor (yes, no) [yes] </li>
<li>Include interactions in the model (yes, no) [no] </li>
<li>Apply TMM normalization (yes, no) [yes]</li>
<li>Analyze only genes which have counts in at least this many samples (0-10000) [1]</li>
<li> Plot width (200-3200 [600]</li>
<li> Plot height (200-3200) [600]</li>
</ul>
<br>
<h3>Details</h3>
<p>This tool takes as input a table of raw counts from the different samples. The count file has to be associated with a phenodata file describing the experimental factors. 
These files are best created by the tool "Utilities / Define NGS experiment", which combines count files for different samples to one table and created a phenodata file for it.</p>
<p>
You can choose to ignore in statistical testing those genes which are not expressed in either experimental group, or are expressed in very low levels (less than 5 counts). 
These genes have little chance of showing significant evidence for differential expression, and removing them reduces the severity of multiple testing adjustment of p-values.
</p>
<p>
Trimmed mean of M-values (TMM) normalization is used to calculate normalization factors in order to reduce RNA composition effect, 
which can arise for example when a small number of genes are very highly expressed in one experiment condition but not in the other.</p>

<p>
Dispersion is estimated using Cox-Reid profile-adjusted likelyhood (CR) method. Trended dispersions are estimated prior to estimating tagwise dispersions.</p>

<p>Statistical analysis to identify differentially expressed genomic features (genes, miRNAs,...) is performed
using a multivariate regression model. A maximum of three different variables, and their interactions
 can be specified for the model. It is highly recommended to always have at least two biological replicates for each experiment condition.</p>
 
<p> Note that unlike with basic edgeR or DESeq2, no filtering is done to the results table. 
Please notice that in the results table you will have all the comparisons between the different effects defined in the parameters -section. 
You can filter the table for example based on the adjusted p-values (FDR-columns) using the tool Utilities / Filter table by column value. 
 </p>
 
<h3>Output</h3>
<ul>
<li>edger_glm.tsv: Table containing the statistical testing results, including fold change and p-values.</li>
	<ul>
		<li>logFC = log2 fold change between the groups. E.g. value 2 means that the expression has increased 4-fold</li>
		<li>logCPM = the average log2-counts-per-million</li>
		<li>LR = likelihood ratio statistics </li>
		<li>PValue = the two-sided p-value</li>
		<li>FDR = adjusted p-value</li>
	</ul></li>
<li>dispersion-edger-glm.pdf: Biological coefficient of variation plot.</li>
</ul>
<br>
<h3>References</h3>
<p>This tool uses the edgeR package for statistical analysis. Please cite the following articles:</p>

<p>MD Robinson, DJ McCarthy and GK Smyth. <a href="http://www.ncbi.nlm.nih.gov/pubmed/19910308">
edgeR: a Bioconductor package for differential expression analysis of digital gene expression data. Bioinformatics, 26 (1):139-40, Jan 2010.</a></p>

<p>DJ McCarthy, Y Chen and GK Smyth. <a href="http://www.ncbi.nlm.nih.gov/pubmed/22287627">
Differential expression analysis of multifactor RNA-Seq experiments with respect to biological variation. Nucleic Acids Res, 40 (10):4288-97, May 2012.</a></p>
</body>
</html>
