<html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=us-ascii">
</head>
<body>
<h2>RNA-seq / Differential expression using edgeR for multivariate experiments</h2>
<h3>Description</h3>
Differential expression analysis for multifactor experiments using the generalized linear models (glm) -based statistical methods of the edgeR Bioconductor package.
<br>
<br>
<h3>Parameters</h3>
<ul>
<li>Apply normalization (yes, no) [yes] </li>
<li>Main effect 1 </li>
<li>Main effect 2 </li>
<li>Main effect 3 </li>
<li>Treat main effect 1 as factor (yes, no) [no] </li>
<li>Treat main effect 2 as factor (yes, no) [no] </li>
<li>Treat main effect 3 as factor (yes, no) [no] </li>
<li>Interaction (main, all) [main] </li>
</ul>
<br>
<h3>Details</h3>
<p>This tool takes as input a table of raw counts from the different samples. The count file has to be associated with a phenodata file describing the experimental factors. 
These files are best created by the tool "Utilities / Define NGS experiment", which combines count files for different samples to one table and created a phenodata file for it.</p>
<p>
Given an input table of raw counts, the edgeR package performs statistical analysis to identify differentially expressed genomic features (genes, miRNAs,...) 
using a multivariate regression model. A maximum of three different variables, and their interactions
 can be specified for the model. It is highly recommended to always have at least two biological replicates for each experiment condition.</p>
<h3>Output</h3>
<ul>
<li>de-list-edger.tsv: Table containing the statistical testing results, including fold change and p-values.<br>
</li>
</ul>
<br>
<h3>References</h3>
<p>This tool uses the edgeR package for statistical analysis. Please cite the following articles:</p>

<p>MD Robinson, DJ McCarthy and GK Smyth. <a href="http://www.ncbi.nlm.nih.gov/pubmed/19910308">
edgeR: a Bioconductor package for differential expression analysis of digital gene expression data. Bioinformatics, 26 (1):139-40, Jan 2010.</a></p>

<p>DJ McCarthy, Y Chen and GK Smyth. <a href="http://www.ncbi.nlm.nih.gov/pubmed/22287627">
Differential expression analysis of multifactor RNA-Seq experiments with respect to biological variation. Nucleic Acids Res, 40 (10):4288-97, May 2012.</a></p>
</body>
</html>
