<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html lang="en">
<head><title></title></head>
<body>

<h2>Retrieve sequences from NCBI</h2>

<h3>Description</h3>

<p>
This tool retrieves sequences from the NCBI sequence databases based on given query terms. The tool
is based on the <a href="http://www.ncbi.nlm.nih.gov/books/NBK179288/">NCBI Edirect<a> package.
</p><p>
You can search for protein or nucleotide sequence entries, whose annotation 
matches the given search terms. A search term must be  a single word consisting 
of letters and/or numbers. You can also use wildcard character *, to match 
any string. The search is case-insensitive: <i>Mus</i> and <i>mus</i> will produce
the same matches.

<p>The case of sequence length, a range should be defined with 
syntax <i>from</i>:<i>to</i>. For example: 120:125
</p>


The available search fields are

<li>All fields</li>
<li>Keywords</li>
<li>Author</li>
<li>Organism</li>
<li>Accession</li>
<li>Gene name</li>
<li>Protein name</li>
<li>Sequence length</li>

<p>
One search can include 1-3 search terms that are combined using the given logical operator (AND, OR, NOT).
</p>
<p>
The matching sequences are by default saved in FASTA format, but also native Genbank formats, that include the sequence annotation too, can be used.
</p>

<h3>Details</h3>

<p>
As retrieving data from NCBI server takes time, the maximum number of hits to be retrieved is set to 50 000 hits. 
However, in the case of nucleotides, 50 000 hits may already be far too big data set to download a as one "hit" may be a
complete chromosome or genome.
</p>


<h3>Output</h3>

<p>The matching sequences are by default saved in fasta format, but also native Genbank formats, that include the sequence annotation too, can be used.</p>


</body>

