<html lang="en">
<head>
<title>Chipster manual</title>
</head>
<body>
<h2>RNA-seq / Count reads per transcripts using eXpress</h2>

<h3>Description</h3>
Counts reads per transcripts using the eXpress package.  
<br>
<br>
<h3>Parameters</h3>
<ul>
<li>Quality scale used in the FASTQ file (phred + 33, phred + 64) [phred + 33] </li>
<li>How many valid alignments should Bowtie2 search per read (20, 100, 1000) [100] </li>
<li>Mean fragment length (10-10,000) [200] </li>
<li>Fragment length standard deviation (5-200) [60] </li>
</ul>

<h3>Details</h3>

<p>
This tool takes as input either paired end or single end reads in FASTQ files, and transcripts in a multi-fasta file. 
Both the read and transcript file can be zipped.
Reads are aligned to the transcripts using Bowtie2, and you should allow as many multimappings as possible. 
Note that Bowtie2 can be very slow if many alignments are allowed per read. 
</p>
<p>
Counting reads at transcript level is complicated by the fact that transcript isoforms typically have overlapping parts. 
In order to assign ambiguously mapping reads to different isoforms, an expectation maximization (EM) approach is used. 
This approach alternates between two steps: an expectation step where reads are assigned to transcripts with a probability according 
to those transcripts’ abundances (which are initially assumed to be equal), and a maximization step where the abundances are updated based 
on the assignment probability.
</p>
<p>
eXpress can resolve multi-mappings of reads across transcripts and gene families, learn fragment length distribution from data, 
and correct for sequence specific bias near the ends of fragments, which arises due to primers used in library preparation. 
In addition, eXpress also includes a model for sequencing errors including indels.
</p>
<p>
The fragment length and standard deviation parameters are used by eXpress.
While the empirical distribution is estimated from paired-end reads on-the-fly, these values parameterize the prior distribution. 
If only single-end reads are available, this prior distribution is also used to determine the effective length
</p>
<p>
eXpress reports the following abundance measures for each transcript: estimated counts, effective counts, FPKM and TPM.
The authors of eXpress recommend using rounded effective counts differential expression analysis with tools like edgeR.
You can combine several count files to a count table using the tool "Utilities / Create NGS experiment"
</p>

<h3>Output</h3>
<p>The analysis output consists of the following files:</p>
<ul>
<li>full-express-output.tsv: The full result file of eXpress sorted by bundle identifier. </li>
<li>effective-counts-express.tsv: Result file containing rounded effective counts. </li>
</ul>

<h3>Reference</h3>

Please cite the following article if you use this tool:<br>
A Roberts and L Pachter: <a href="http://www.nature.com/nmeth/journal/v10/n1/full/nmeth.2251.html">Streaming fragment assignment for real-time analysis of sequencing experiments</a> Nature Methods 10,71–73(2013). 

</body>
</html>

