<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html lang="en">
<head><title></title></head>
<body>

<h2>Quality control / Read quality with FastQC</h2>

<h3>Description</h3>

<p>Given a FASTQ file, this tool produces several quality control plots for the reads.</p>

<h3>Parameters</h3>

<ul>
<li>File type (bam, fastq) [fastq]</li>
<li>Create all plots (yes, no) [no]</li>
</ul>

<h3>Details</h3>

<p>FastQC supports all quality encoding variants of FASTQ files and BAM files. GZip compressed files are automatically detected and unzipped. By default plots are generated for per base quality and sequence content. User can also request to get plots for GC and N content, sequence length distribution, 
duplication levels, overrepresented sequences and Kmer content. 
For a detailed description of the plots, please see the <a href="http://www.bioinformatics.bbsrc.ac.uk/projects/fastqc/Help/3%20Analysis%20Modules/">FastQC manual</a>.</p>

<h3>Output</h3>

<p>Image files (.png) for the requested plots. The plotted information is also given in text format (fastqc_data.txt). </p>

<h3>Reference</h3>

This tool is based on the <a href="http://www.bioinformatics.bbsrc.ac.uk/projects/fastqc/">FastQC package</a> by Simon Andrews et al.

</body>

