<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html lang="en">
<head><title></title></head>
<body>

<h2>Nucleic acid distance algorithm</h2>

<h3>Description</h3>

<p>This tool is based on EMBOSS command <b>fdnadist</b> that is a wrapper command for 
command <b>dnadist</b> of the <a href="http://evolution.genetics.washington.edu/phylip.html">PHYLIP</a> package.
</p><p>
<b>Nucleic acid distance algorithm</b> can be used to compute four different distances between species from nucleic acid 
sequences.  The distances are the Jukes-Cantor formula, 
one based on Kimura's 2- parameter method, the F84 model used in DNAML, and the LogDet distance. 
The distances can also be corrected for gamma-distributed and gamma-plus-invariant-sites-distributed 
rates of change in different sites. Rates of evolution can vary among sites in a prespecified way, 
and also according to a Hidden Markov model. The program can also make a table of percentage similarity 
among sequences. </p>

<p>The resulting distance matrix can be used as an input for tool: <b>Phylogenies from distance matrix 
by N-J or UPGMA method</b>.</p>

<h3>Parameters</h3>
<li>Choose the method to use (fdnadist parameter: -method)</li> 
<li>Gamma distribution (fdnadist parameter: -gammatype)</li> 
<li>Number of substitution rate categories (fdnadist parameter: -ncategories)</li>
<li>Category rates (fdnadist parameter: -rate)</li>
<li>Coefficient of variation of substitution rate among sites (fdnadist parameter: -gammacoefficient)</li> 
<li>Fraction of invariant sites (fdnadist parameter: -invarcoefficient )</li> 
<li>Transition/transversion ratio (fdnadist parameter: -ttratio )</li> 
<li>Use empirical base frequencies from seqeunce input (fdnadist parameter: -freqsfrom )</li>
<li>Base frequencies for A C G T/U (use blanks to separate\) (fdnadist parameter: -basefreq )</li>
<li>Output as a lower triangular distance matrix (fdnadist parameter: -lower ) </li>
<li>Output as a human-readable distance matrix  (fdnadist parameter: -humanreadable ) </li>
<li>Print data at start of run (fdnadist parameter: -printdata )</li> 
<li>Collect a log file</li> 

</p>
<p>
For more details, please check the <a href="https://extras.csc.fi/emboss/doc/embassy/phylip/fdnadist.html">manual page of fdnadist command</a>.
</p>


<h3>Input and output</h3>

<p>
Input is a nucleotide multiple sequence alignment.
</p>
<p>
Output is a text file containing the distance matrix.
</p>


</body>

