<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html lang="en">
<head><title></title></head>
<body>

<h2PDNA parsimony algorithm</h2>

<h3>Description</h3>

<p>This tool is based on EMBOSS command <b>fdnapars</b> that is a wrapper command for 
command <b>dnapars</b> of the <a href="http://evolution.genetics.washington.edu/phylip.html">PHYLIP</a> package.
</p>
<b>DNA parsimony algorithm</b> estimates phylogenies by the parsimony method using nucleic acid sequences. 
It allows use the full IUB ambiguity codes, and estimates ancestral nucleotide states. 
Gaps are treated as a fifth nucleotide state. It can also be used for transversion parsimony. 
This method can cope with multifurcations, reconstruct ancestral states, use 0/1 character weights, 
and infer branch lengths.
</p>
<p>
<h3>Parameters</h3>

<p>
<li>Number of trees to save (fdnapars parameter -maxtrees )</li>
<li>More thorough search (fdnapars parameter -thorough )</li>
<li>Rearrange on just one best tree (fdnapars parameter -rearrange )</li>
<li>Use transversion parsimony (fdnapars parameter -transversion )</li>
<li>Number of times to randomise (fdnapars parameter -njumble )</li>
<li>Random number seed between 1 and 32767 (must be odd) (fdnapars parameter -seed )</li>
<li>Species number to use as outgroup (fdnapars parameter -outgrno )</li>
<li>Use threshold parsimony (fdnapars parameter -dothreshold )</li>
<li>Threshold value (fdnapars parameter -threshold )</li>
<li>Write out trees to tree file (fdnapars parameter -trout )</li>
<li>Print data at start of run (fdnapars parameter -printdata )</li>
<li>Print out steps in each site (fdnapars parameter -stepbox )</li>
<li>Print sequences at all nodes of tree (fdnapars parameter -ancseq )</li>
<li>Print out tree (fdnapars parameter -treeprint )</li>
<li>Use dot differencing to display results (fdnapars parameter -dotdiff )</li>
<li>Collect a log file</li>
</p>

<p>
For more details, please check the <a href="https://extras.csc.fi/emboss/doc/embassy/phylip/fdnapars.html">manual page of fdnapars command</a>.
</p>


<h3>Input and output</h3>

<p>
Input is a dna multiple sequence alignment.
</p>
<p>
Output consists of two files containing. One containing a human readable output report (dnapars.out.txt )
and a treefile (dnapars.tree.txt) that can be used as an input for tree drawing tools (Plot an unrooted tree, Plot cladogram).
</p>


</body>

