<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html lang="en">
<head><title></title></head>
<body>

<h2>Estimate nucleotide phylogeny by maximum likelihood</h2>

<h3>Description</h3>

<p><b>Estimate nucleotide phylogeny by maximum likelihood</b> tool is based on EMBOSS command <b>fdnaml</b> that is a wrapper command for 
command <b>dnaml</b> of the <a href="http://evolution.genetics.washington.edu/phylip.html">PHYLIP</a> package.
</p>
This tool estimates phylogenies from nucleotide sequences by maximum likelihood. 
The model employed allows for unequal expected frequencies of the four nucleotides, 
for unequal rates of transitions and transversions, and for different (prespecified) rates 
of change in different categories of sites, and also use of a Hidden Markov model of rates, 
with the program inferring which sites have which rates. This also allows gamma-distribution 
and gamma-plus-invariant sites distributions of rates across sites.  
</p>
<p>
<h3>Parameters</h3>
<p>
<li>Number of substitution rate categories (fdnaml parameter -Í„ncategories: )</li>
<li>Rate for each category (fdnaml parameter -rate: )</li>
<li>Use branch lengths from user trees (fdnaml parameter -lengths: )</li>
<li>Transition/transversion ratio (fdnaml parameter -ttratio: )</li>
<li>Use empirical base frequencies from seqeunce input (fdnaml parameter -freqsfrom: )</li>
<li>Base frequencies for A C G T/U (use blanks to separate\) (fdnaml parameter -basefreq: )</li>
<li>Rate variation among sites (fdnaml parameter -gammatype: )</li>
<li>Coefficient of variation of substitution rate among sites (fdnaml parameter -gammacoefficient: )</li>
<li>Number of categories (1-9\) (fdnaml parameter -ngammacat: )</li>
<li>Coefficient of variation of substitution rate among sites (fdnaml parameter -invarcoefficient: )</li>
<li>Number of categories (1-9\) including one for invariant sites (fdnaml parameter -ninvarcat: )</li>
<li>Fraction of invariant sites (fdnaml parameter -invarfrac: )</li>
<li>Number of HMM rate categories (fdnaml parameter -nhmmcategories: )</li>
<li>HMM category rates (fdnaml parameter -hmmrates: )</li>
<li>Probability for each HMM category (fdnaml parameter -hmmprobabilities: )</li>
<li>Rates at adjacent sites correlated (fdnaml parameter -adjsite: )</li>
<li>Mean block length of sites having the same rate (fdnaml parameter -lambda: )</li>
<li>Number of times to randomise (fdnaml parameter -njumble: )</li>
<li>Random number seed between 1 and 32767 (must be odd\) (fdnaml parameter -seed: )</li>
<li>Global rearrangements (fdnaml parameter -global: )</li>
<li>Species number to use as outgroup (fdnaml parameter -outgrno: )</li>
<li>Speedier but rougher analysis (fdnaml parameter -rough: )</li>
<li>Write out trees to tree file (fdnaml parameter -trout: )</li>
<li>Print data at start of run (fdnaml parameter -printdata: )</li>
<li>Print out tree (fdnaml parameter -treeprint: )</li>
<li>Reconstruct hypothetical sequence (fdnaml parameter -hypstate: )</li>
<li>Collect a log file</li>
</p>

<p>
For more details, please check the <a href="https://extras.csc.fi/emboss/doc/embassy/phylip/fdnaml.html">manual page of fdnaml command</a>.
</p>


<h3>Input and output</h3>

<p>
Input is a dna multiple sequence alignment.
</p>
<p>
Output consists of two files containing. One containing a human readable output report (dnapars.out.txt )
and a treefile (dnapars.tree.txt) that can be used as an input for tree drawing tools (Plot an unrooted tree, Plot cladogram).
</p>


</body>

