<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html lang="en">
<head><title></title></head>
<body>

<h2>Protein parsimony algorithm</h2>

<h3>Description</h3>

<p>This tool is based on EMBOSS command <b>fprotpars</b> that is a wrapper command for 
command <b>protpars</b> of the <a href="http://evolution.genetics.washington.edu/phylip.html">PHYLIP</a> package.
</p>
<b>Protein parsimony algorithm</b> estimates phylogenies from protein sequences using the parsimony method, 
in a variant which counts only those nucleotide changes that change the amino acid, on the assumption that 
silent changes are more easily accomplished. 
</p>
<p>
<h3>Parameters</h3>

<li>Number of times to randomise (protpars parameter -njumble )</li>
<li>Random number seed between 1 and 32767 (must be odd) (protpars parameter -seed )</li>
<li>Species number to use as outgroup (protpars parameter -outgrno )</li>
<li>Use threshold parsimony (protpars parameter -dothreshold )</li>
<li>Threshold value (protpars parameter -threshold )</li>
<li>Use which genetic code (protpars parameter -whichcode )</li>
<li>Write out trees to tree file (protpars parameter -trout )</li>
<li>Print data at start of run (protpars parameter -printdata )</li>
<li>Print out tree (protpars parameter -treeprint )</li>
<li>Print steps at each site (protpars parameter -stepbox )</li>
<li>Print sequences at all nodes of tree (protpars parameter -ancseq )</li>
<li>Use dot differencing to display results (protpars parameter -dotdiff )</li>
<li>Collect a log file</li>


</p>
<p>

For more details, please check the <a href="https://extras.csc.fi/emboss/doc/embassy/phylip/fprotpars.html">manual page of fprotpars command</a>.
</p>


<h3>Input and output</h3>

<p>
Input is a protein multiple sequence alignment.
</p>
<p>
Output consists of two files containing. One containing a human readable output report (protpars.out.txt )
and a treefile (protpars.tree.txt) that can be used as an input for tree drawing tools (Plot an unrooted tree, Plot cladogram).
</p>


</body>

