<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html lang="en">
<head><title></title></head>
<body>

<h2>Protein distance algorithm</h2>

<h3>Description</h3>

<p>This tool is based on EMBOSS command <b>fprotdist</b> that is a wrapper command for 
command <b>protdist</b> of the <a href="http://evolution.genetics.washington.edu/phylip.html">PHYLIP</a> package.
</p>
<b>Protein distance algorithm</b> computes a distance measure for aligned protein sequences, using maximum likelihood estimates based on the 
Dayhoff PAM matrix, the JTT matrix model, the PBM model, Kimura's 1983 approximation to these, or a 
model based on the genetic code plus a constraint on changing to a different category of amino acid. 
The distances can also be corrected for gamma-distributed and gamma-plus-invariant-sites-distributed 
rates of change in different sites. Rates of evolution can vary among sites in a prespecified way, 
and also according to a Hidden Markov model. The program can also make a table of percentage similarity 
among sequences. The distance matrix can be used as an input for tool: <b>Phylogenies from distance matrix 
by N-J or UPGMA method</b>.
</p>
<p>
<h3>Parameters</h3>
<li>Number of substitution rate categories (fprotdist parameter: -ncategories)
<li>Rate for each category (fprotdist parameter: -rate)</li>
<li>Choose the method to use (fprotdist parameter: -method)</li> 
<li>Rate variation among sites (fprotdist parameter: -gammatype)</li> 
<li>Coefficient of variation of substitution rate among sites (fprotdist parameter: -gammacoefficient)</li> 
<li>Coefficient of variation of substitution rate among sites (fprotdist parameter: -invarcoefficient )</li> 
<li>Choose the category to use (fprotdist parameter: -aacateg )</li> 
<li>Which genetic code (fprotdist parameter: -whichcode )</li> 
<li>Prob change category (1.0=easy) (fprotdist parameter: -ease )</li> 
<li> Transition/transversion ratio (fprotdist parameter: -ttratio )</li> 
<li>Base frequencies for A C G T/U (use blanks to separate) (fprotdist parameter: -basefreq )</li> 
<li>Print data at start of run (fprotdist parameter: -printdata )</li> 
<li>Collect a log file</li> 

</p>
<p>

For more details, please check the <a href="https://extras.csc.fi/emboss/doc/embassy/phylip/fprotdist.html">manual page of fprotdist command</a>.
</p>


<h3>Input and output</h3>

<p>
Input is a protein multiple sequence alignment.
</p>
<p>
Output is a text file containing the distance matrix.
</p>


</body>

