<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html lang="en">
<head><title></title></head>
<body>

<h2>Protein pattern search</h2>

<h3>Description</h3>

<p>
 Search for patterns in protein sequences.
</p>


<p>This tool runs EMBOSS command <b>fuzzpro</b> that searches for a specified 
PROSITE-style pattern in protein  sequences. Such patterns are specifications of a (typically short)
   length of sequence to be found. They can specify a search for an exact
   sequence or they can allow various ambiguities, matches to variable
   lengths of sequence and repeated subsections of the sequence. One or
   more nucleotide sequences are read from file. </p>
 <p>
   The standard IUPAC one-letter codes for the amino acids are used in the search patterns.
   The symbol 'x' is used for a position where any nucleotide is accepted.
   In standard PROSITE pattern notation square parentheses '[ ]', and curly 
   brackets '{ }' are used to define ambiguities, but at the moment these characters 
   can't be used with this tool in Chipster.
    
<p>
The input protein sequences can be in any of the <a href="http://emboss.sourceforge.net/docs/themes/SequenceFormats.html">EMBOSS compatible sequence formats</a>.
</p>
<p>
  For more details, please check the manual page of <a href="http://emboss.sourceforge.net/apps/release/6.5/emboss/apps/fuzzpro.html">fuzzpro</a> command.
</p>


</p>


<h3>Output</h3>

<p>
Output is a table or text formatted report file.
</p>


</body>

