<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html lang="en">
<head><title></title></head>
<body>

<h2>Getorf</h2>

<h3>Description</h3>

<p>
Find and extract open reading frames (ORFs)
 
</p>


<p>This tool uses EMBOSS command <b>getorf</b> to find and output the sequences 
of open reading frames  (ORFs) in one or more nucleotide sequences. An ORF may be defined as a
  region of a specified minimum size between two STOP codons, or between
   a START and a STOP codon. The ORFs can be output as the nucleotide
   sequence or as the protein translation. Optionally, the program will
   output the region around the START codon, the first STOP codon, or the
   final STOP codon of an ORF. The START and STOP codons are defined in a
   Genetic Code table; a suitable table can be selected for the organism
   you are investigating. The output is a sequence file containing
   predicted open reading frames longer than the minimum size, which
   defaults to 30 bases (i.e. 10 amino acids).



</p>
<p>
The input sequences can be in any of the <a href="http://emboss.sourceforge.net/docs/themes/SequenceFormats.html">EMBOSS compatible sequence formats</a>.
</p>
<p>
For more details, please check the <a href="http://emboss.sourceforge.net/apps/release/6.5/emboss/apps/getorf.html">manual page of getorf command</a>.
</p>


<h3>Output</h3>

<p>
Output is a fasta formatted sequence file.
</p>


</body>

