<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html lang="en">
<head><title>Chipster manual</title></head>
<body>

<h2>Chipster manual</h2>

<p>Chipster is a versatile data analysis platform with interactive visualizations and workflows. It offers a comprehensive collection of analysis tools for next generation sequencing (NGS),
microarray and proteomics data. The NGS functionality covers analysis from quality control and alignment to downstream applications such as pathway analysis 
and motif detection (please see the <a href="http://chipster.csc.fi/features.shtml">analysis tool summary</a> for details), 
and more analysis tools are added all the time. The built-in Chipster genome browser allows users visualize reads and results in their genomic context.
The microarray functionality covers expression, aCGH, miRNA, methylation and SNP arrays, and allows users to integrate expression data with miRNA and aCGH data. 
 
</p>
<b>General</b>
<ul>
<li> Getting started with Chipster: <a href="basic-functionality.html">5 minute tour</a>, <a href="tutorials.html">tutorials</a> </li>
<li> <a href="requirements.html">Software requirements</a> </li>
<li> <a href="sessions.html">Analysis sessions</a>: How to continue your analysis later and share it </li>
<li> <a href="visualisation-phenodata.html">Phenodata editor</a>: Describing your experimental setup </li>
<li> <a href="workflows.html">Workflows</a>: How to automate and share your analysis pipelines </li>
<li> <a href="developer.html">For developers</a> </li>
<li> <a href="copyright.html">Copyright</a> </li>
</ul>
<b>Functionality for NGS data</b>
<ul>
<li> <a href="genomeBrowser.html">Genome browser</a>: Viewing NGS reads and results in genomic context </li>
<li> <a href="tools-ngs.html">Analysis tools for NGS data</a> </li>
<ul>
<li> <a href="tools-ngs.html#filtering">Filtering</a> </li>
<li> <a href="tools-ngs.html#quality">Quality control</a> </li>
<li> <a href="tools-ngs.html#utilities">Utilities</a> </li>
<li> <a href="tools-ngs.html#regions">Matching genomic regions</a> </li>
<li> <a href="tools-ngs.html#alignment">Alignment</a> </li>
<li> <a href="tools-ngs.html#variants">Variants</a> </li>
<li> <a href="tools-ngs.html#RNA">RNA-seq</a> </li>
<li> <a href="tools-ngs.html#miRNA">miRNA-seq</a> </li>
<li> <a href="tools-ngs.html#ChIP">ChIP-seq and FAIRE-seq</a> </li>
<li> <a href="tools-ngs.html#CNA">CNA-seq</a> </li>
<li> <a href="tools-ngs.html#methyl">Methyl-seq</a> </li>
<li> <a href="tools-ngs.html#meta">Metagenomics</a> </li>
</ul>
</ul>
<b>Functionality for microarray data</b>
<ul>
<li> <a href="supported-chips.html">Supported microarray types</a> </li>
<li> <a href="import.html">Data import</a> </li>
<li> <a href="tools.html">Analysis tools for microarray data</a> </li>
<ul>
<li> <a href="tools.html#normalization">Normalization</a> </li>
<li> <a href="tools.html#quality">Quality control</a> </li>
<li> <a href="tools.html#preprocessing">Preprocessing</a> </li>
<li> <a href="tools.html#statistics">Statistics</a> </li>
<li> <a href="tools.html#clustering">Clustering</a> </li>
<li> <a href="tools.html#annotation">Annotation</a> </li>
<li> <a href="tools.html#pathways">Pathways</a> </li>
<li> <a href="tools.html#promoter">Promoter analysis</a> </li>
<li> <a href="tools.html#visualization">Visualization</a> </li>
<li> <a href="tools.html#utilities">Utilities</a> </li>
<li> <a href="tools.html#acgh">aCGH</a> </li>
</ul>
</ul>

<!--<li>Supporting material</li>
<ul>
<li> <a href="http://www.csc.fi/csc/julkaisut/oppaat/arraybook_overview">CSC's DNA microarray data analysis book</a></li>
<li> <a href="http://koti.mbnet.fi/tuimala/oppaat/r2.pdf">CSC's Bioconductor book by Jarno Tuimala</a></li>
</ul>-->


</ul>


</body>

</html>
