<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html lang="en">
<head><title></title></head>
<body>

<h2>Alingment statistics</h2>

<h3>Description</h3>

<p>Display basic information about a multiple sequence alignment.</p>

<p> 
   This tool runs EMBOSS command <b>infoalign</b> that lists basic information
    about sequences in a multiple sequence alignment. This includes the sequences'
   name, two measures of length, counts of gaps, and numbers of identical,
   similar and different residues or bases in this sequence when compared
   to a reference sequence, together with a simple statistic of the %
   change between the reference sequence and this sequence. <p>

<p>
   The reference sequence is the one against which all the other sequences
   are compared using a specified substitution matrix. It is either the
   calculated consensus sequence of the alignment (the default) or it can
   be one of the set of aligned sequences, specified by either the ordinal
   number of that sequence in the input file, or by name. There are
   various options to control how the consensus is calculated.
</p>


<p>
The input sequences can be in any of the <a href="http://emboss.sourceforge.net/docs/themes/AlignmentFormats.html">EMBOSS compatible sequence alignment formats</a>.
Note that this tool reads just one input file, that should contain all the sequences to be analyzed.
</p>
<p>
For more details, please check the manual pages of <a href="http://emboss.sourceforge.net/apps/release/6.5/emboss/apps/infoalign.html">infoalign</a> command.
</p>


<h3>Output</h3>

<p>
Output is a fasta formatted sequence file.
</p>


</body>

