<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html lang="en">
<head><title></title></head>
<body>

<h2>Find peaks using MACS2</h2>

<h3>Description</h3>



This tool identifies statistically significantly enriched genomic regions in ChIP- and DNase-seq data using the MACS2 algorithm (Model-Based Analysis of ChIP-Seq).  
<br>
<br>
<h3>Parameters</h3>

<ul>
<li>File format (ELAND, BAM, BED) [BAM]</li>
<li>Mappable genome size (hg18, hg19, mm9, mm10, rn5, user-specified) [hg19]</li>
<li>User-specified mappable genome size []</li>
<li>q-value cutoff (0..1) [0.01]</li>
<li>Read length (1..200) [0]</li>
<li>Keep duplicates (auto, all, 1) [auto]</li>
<li>Build peak model (yes, no) [yes]</li>
<li>Bandwidth (1..1000) [300]</li>
<li>Upper M-fold cutoff (1..100) [30]</li>
<li>Lower M-fold cutoff (1..100) [10]</li>
<li>Extension size (1..1000) [200]</li>
<li>Call broad peaks (yes, no) [no]</li>
</ul>

<h3>Details</h3>

<p>
MACS2 performs several steps as described below, ranging from duplicate filtering and peak model building to the actual peak detection and multiple testing correction. 
It has also an option to link nearby peaks together in order to call broad peaks.
</p>
<p>
If the read length parameter is set to zero, MACS2 detects read length automatically. MACS2 then proceeds to filter out duplicate reads. 
By default it calculates the maximum number of duplicate reads in a single position warranted by the sequencing depth, and removes redundant reads in excess of this number. 
Alternatively, you can select to keep only one read, or all duplicates. 
</p>
<p>
MACS2 models the distance between the paired forward and reverse strand peaks from the data. It slides a window across the genome to find enriched regions, which have M-fold more reads than background. 
The size of the window is twice the bandwidth parameter. The expected background is the number of reads times their length divided by the mappable genome size. 
Note that the mappable genome size is always less than the real genome size because of repetitive sequence.
The regions' fold enrichment must be higher than 10 and less than 30, but you can change these values if not enough regions are found. A smaller value for the lower cutoff provides more regions for model building, 
but it can also include spurious data into the model and thereby adversely affect the peak finding results. 
MACS2 uses 1000 enriched regions to model the distance d between the forward and reverse strand peaks.  
</p>
<p>
In the actual peak detection phase, MACS2 extends the reads in the 3' direction to the fragment length obtained from modeling. If the model building failed or if it was switched off, the reads are extended to the value of the extension size parameter. 
If a control sample is available, MACS2 scales the samples linearly to the same read number. It then selects candidate peaks by scanning the genome again, now using a window size which is twice the fragment length. 
MACS2 calculates a p-value for each peak using a dynamic Poisson distribution to capture local biases in read background levels. If a control sample is available, it is used to calculate the local background.
Finally, q-values are calculated using the Benjamini-Hochberg correction.
</p>

<h3>Output</h3>

<p>The analysis results consist of the following files:</p>

<ul>
<li>macs2-peaks.tsv: List of peaks and their length, summit location and height, as well as their fold change, p- and q-value. Note that you can visualize this file in the Chipster genome browser.</li>
<li>macs2-narrowpeak.bed: List of peak locations, q-values, fold change, p-value, q-value (again) and summit position relative to peak start, in narrow peak format (BED6+4).</li>
<li>macs2-summits.bed: List of peak summits and q-values in BED format.</li>
<li>macs2-log.txt: A log file listing the output from the various steps, which can be useful for diagnostic purposes and to get to know the details of the process.</li>
<li>macs2-model.pdf: If the peak model building is successful, a plot of the model is generated. The shape of the modeled peaks allows you to assess the quality of the model. </li>
</ul>
<h3>References</h3>

<p>Please cite the following article and the <a href="https://github.com/taoliu/MACS/">MACS2 website</a>.</p>

<p> Zhang Y, Liu T, Meyer CA, Eeckhoute J, Johnson DS, Bernstein BE, Nusbaum C,
Myers RM, Brown M, Li W, Liu XS. (2008) <a href="http://www.ncbi.nlm.nih.gov/pubmed/18798982">Model-based Analysis of ChIP-Seq (MACS), Genome Biology, 2008;9(9):R137</a>.
</p> 

</body>

