<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html lang="en">
<head><title></title></head>
<body>

<h2>Multiple sequence alignmnet with MAFF</h2>

<h3>Description</h3>

<p>
This tool calculates multiple sequence alignment using MAFFT multiple sequence alignment tool.  
</p>
MAFFT offers various multiple alignment strategies: the progressive ClusltaW like method, iterative refinement method with the WSP, and iterative refinment method using both the WSP and consistency scores. 
In general, there is a tradeoff between speed and accuracy. Faster methods are less accurate. 
<p>
In Chipster, MAFFT default selects the alignment method to be used based on the size of the sequence set to be aligned.
Optionally, user can define a specific alignment strategy. Following strategies are available </p>
<p><ul>
<li><b>FFT-NS-1</b> (Very fast. Recommended for more than 2000 sequences. progressive method). The procedure is: (1) make a rough distance matrix by counting the number of shared 6-tuples (see below) between every sequence pair, (2) build a guide tree and (3) align the sequences according to the branching order.  </li>
<li><b>FFT-NS-2</b> (Fast. Progressive method. The distance matrix used in FFT-NS-1 is very approximate and unreliable. In FFT-NS-2, (4) the guide tree is re-computed from the FFT-NS-1 alignment, and (5) the second progressive alignment is carried out. </li>
<li><b>FFT-NS-i</b> (Iterative refinement method.The accuracy of progressive alignment can be improved by the iterative refinement method (Berger and Munson 1991, Gotoh 1993). A simplified version of PRRN is implemented as the FFT-NS-i option of MAFFT. In FFT-NS-i, an initial alignment by FFT-NS-2 is subjected to an iterative refinement process.  
</li></ul></p>
<p>
In order to obtain more accurate alignments in difficult cases, 
three more complex alignment methods have been added toMAFFT. These methids use an 
 objective function combining the WSP score (Gotoh) explained above and 
 the COFFEE-like score (Notredame et al.), which evaluates the consistency 
 between a multiple alignment and pairwise alignments (Katoh et al. 2005).
</p><p>
For pairwise alignment, three different types of algorithms are implemented, 
global alignment (Needleman-Wunsch), local alignment (Smith-Waterman) with 
affine gap costs (Gotoh) and local alignment with generalized affine gap costs 
(Altschul). The differences in the accuracy values among these methods are 
small. However, each  of them has different characteristics, according to the
 algorithm in the pairwise alignment stage and thus the optimal method depends on 
 the alignment case. The tree methods are: 
</p>
<p><ul>
<li><b>E-INS-i</b> Slow. recommended for less than 200 sequences with multiple conserved domains and long gaps.</li> 
<li><b>L-INS-i</b> Slow. Recommended for less than 200 sequences with one conserved domain and long gaps.</li>
<li><b>G-INS-i</b> Slow. Recommended for less than 200 sequences with global homology</li>
</ul>

<h3>Details</h3>

<p>
More information can be found from the <a href="http://mafft.cbrc.jp/alignment/software/algorithms/algorithms.html">MAFFT home page</a>.
The input sequences must be in fasta format.
</p>

<p> 


</p>


<h3>Output</h3>

<p>
Output is a sequence file.
</p>


</body>

