<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html lang="en">
<head><title></title></head>
<body>

<h2>Local pairwise sequence alignment</h2>

<h3>Description</h3>

<p>
Waterman-Eggert local alignment of two sequences 
</p>

<p>
This tool runs EMBOSS command <b>matcher</b> that identifies local similarities in two input sequences using a
   rigorous algorithm based on Bill Pearson's lalign application, version
   2.0u4 (Feb. 1996). The substitution matrix, gap insertion and extension
   penalty are specified. The specified number of top-scoring pair-wise
   local sequence alignments are written to file.
</p>
<p>
The input sequences can be in any of the <a href="http://emboss.sourceforge.net/docs/themes/SequenceFormats.html">EMBOSS compatible sequence formats</a>.
</p>
<p>
For more details, please check the <a href="http://emboss.sourceforge.net/apps/release/6.5/emboss/apps/matcher.html">manual page of getorf command</a>.
</p>

<H3>Parameters</H3>

<li><b>How many alignments to show</b>This sets the number of alternative matches output. By default only the highest scoring alignment is shown. A value of 2 gives you other reasonable alignments. In some cases, for example multidomain proteins of cDNA and genomic DNA comparisons, there may be other interesting and significant alignments.</li>
<li><b>Gap penalty</b> Gap penalty</li>
<li><b>Gap length penalty</b> Gap length penalty</li>
<li><b>Scoring matrix</b> Choose the scoring matrix file used when comparing sequences. By default BLOSUM62 (for proteins\) or the file EDNAFULL (for nucleic sequences\) are used)</li>
<li><b>Output type for outfile</b> Choose format for output file</li>




<h3>Output</h3>

<p>
Output is a fasta formatted sequence file.
</p>


</body>

