<html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=us-ascii">
</head>
<body>
<h2>MEDIPS - methylation analysis</h2>

<h3>Description</h3>
<p>This tool performs methylation analysis of sequencing data for either a single condition, or comparing two conditions, e.g. treatment and control. The comparative analysis is performed by providing a second BAM file for the control in addition to the treatment BAM file. Additionally, a BAM file can also be specified for input sequencing data. Currently, the tool works only with human data.</p>

<h3>Parameters</h3>
<ul>
<li>Chromosome names in my BAM file look like (1, chr1) [1]</li>
<li>Genome (human hg19, human GRCh38, mouse mm10, rat rn5) [human GRCh38]</li>
<li>Use only promoters (yes, no) [no] </li>
<li>Fragment length (400, 800, 1600, 2400) [800] </li>
<li>Resolution for coverage calculation (25, 50, 100, 200) [50] </li>
<li>Smoothing extension (200, 400, 800, 1200) [400] </li>
<li>Promoter length, upstream from the TSS (1000, 2000, 5000) [1000] </li>
<li>Promoter length, downstream from the TSS (250, 500, 750, 1000) [500] </li>
<li>Save a BED file (yes, no) [yes] </li>
</ul>

<h3>Details</h3>
<p>Based on the input data, a genome-wide coverage is calculated. The resolution for the coverage calculation can be set, and it is by default 50 bps. Smoothing of the resolution is recommended, and can be set separately from the coverage resolution. The default length for smoothing is 400 bp. Normalization is controlled using the setting for fragment length. Users can choose to calculate the enrichment of methylated CpG islands for promoter regions only, and select the length of the promoter region used. The default is 1000 bps upstream and 500 bps downstream from the transcription start site.</p>

<p>During the analysis, the whole genome is split into bins of coverage resolution (by default 50 bps). The short read coverage is calculated using this resolution. Since the short reads are not the whole length sonicated DNA fragments, smoothing of the results is usually performed by extending the reads. During the extension, each read is extended to both telomeric and centromeric directions by the amount set in the smoothing extension parameter. For normalization the local CpG density needs to be estimated. The setting for fragment length specifies the area around each genomic bin that is used for calculating the local CpG density. The fragment length should be set close to the estimated size of the sonicated DNA fragments generated after the amplification.</p>

<h3>Output</h3>
<p>Methylation results are given in the file methylation.tsv. Users can also choose to have the relative methylation scores in saved as BED files, which can be visualized in the Chipster genome browser. When performing a comparative analysis of treatment vs. control, also p-values and fold changes are also returned.</p>

Several quality control plots are generated: 
<ul>
<li>The saturation analysis plot is used for deciding whether the amount of data is enough to cover the genome and generate reliable and reproducible methylation profiles.</li>
<li>The coverage analysis plot is used for getting an idea of number of CpGs covered by the short reads, and the depth of the coverage.</li>
<li>The calibration curve plot enables examination of the dependency of local CpG densities and signals derived from the data.</li>
</ul>

<h3>References</h3>
<p>For more information about the functionality, see the manual on Bioconductor site at
<a href="redir.aspx?C=f7adeb2573634fb8abd6c42c012c17b5&amp;URL=http%3a%2f%2fwww.bioconductor.org%2fpackages%2f3.0%2fbioc%2fhtml%2fMEDIPS.html" target="_blank">
http://www.bioconductor.org/packages/3.0/bioc/html/MEDIPS.html</a>.</p>
<p>If you publish results acquired using this tool, please cite also Chavez et al., Computational analysis of genome-wide DNA methylation during the differentiation of human embryonic stem cells along the endodermal lineage, Genome Res. 2010 Oct;20(10):1441-50.</p>
</body>
</html>
