<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html lang="en">
<head><title></title></head>
<body>

<h2>Metagenomics / Trim and filter reads with Mothur</h2>

<h3>Description</h3>

<p>
Trim and filter reads and remove duplicate reads.

</p>

<h3>Parameters</h3>
<ul>
<li>flip (yes, no) [no]</li>
<li>qaverage (0-40) []</li>
<li>qwindowaverage (Integer) []</li>
<li>qwindowsize (Integer) []</li>
<li>qstepsize (Integer) []</li>
<li>maxambig (0-10) []</li>
<li>maxhomop (0-50) []</li>
<li>minlength (0-1000) []</li>
<li>maxlength (0-1000) []</li>
<li>pdiffs (0-10) []</li>
<li>bdiffs (0-10) []</li>
</ul>

<h3>Details</h3>
<p>
The reads are checked against user-defined quality requirements. The output sequences will be trimmed to remove the 
user-provided primers, barcodes, and sequences that drop below a quality threshold. The reads are also checked for redundancy. If
a quality file is provided, a trimmed, non-redundant quality file is produce to match the trimmed, non-redundant sequence file.
<p>

<p>
This tool is based on the Mothur package.
</p>


</p>

<h3>Output</h3>

<p>The analysis output consists of the following:</p>
<ul>
<li>reads.groups.txt: A two column list with the first column indicating the sequence names in the input file and the second column the group that it came from.</li>
<li>reads.trim.names.txt: A list of the selected sequences and the redundant sequences corresponding to each.</li>
<li>reads.trim.unique.fasta: A FASTA-formatted sequence file trimmed, non-redundant reads.</li>
<li>reads.trim.unique.qual: Quality values (if provided as input) for the trimmed, non-redundant reads</li>
<li>summary.trim.unique.tsv: Summary statistics for the sequences</li>
</ul>
</p>


<h3>References</h3>
<p>Schloss PD, Westcott SL, Ryabin T, Hall JR, Hartmann M, Hollister EB, Lesniewski RA, Oakley BB, Parks DH, Robinson CJ, Sahl JW, Stres B, Thallinger GG, Van Horn DJ, Weber CF.
<a href="http://www.ncbi.nlm.nih.gov/pmc/articles/PMC2786419/">Introducing mothur: Open-source, platform-independent, community-supported software for describing and comparing microbial communities. Appl Environ Microbiol, 2009. 75(23):7537-41.</a> 
</p>

</body>

