<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html lang="en">
<head><title></title></head>
<body>

<h2>Metagenomics / Classify sequences to taxonomic units with Mothur</h2>

<h3>Description</h3>

<p>
Given a fasta file of aligned 16S rRNA sequences and a groups file generated by Mothur, this tool assigns sequences to taxonomic units (~ species). 
In addition to classification, it generates a count table and a phenodata file, which you can use to assign samples to different experimental groups 
for further statistical analysis.
</p>

<h3>Parameters</h3>

Cutting level for taxonomic names in count table (0-9) [0] 

<h3>Details</h3>
<p>
Assigns aligned 16S rRNA sequences to taxonomic units using the Wang method and the Silva reference set and its taxonomy file. The bootstrap confidence score is calculated with 1000 iterations.
</p>
<p>
This tool is based on the <a href="http://www.mothur.org/wiki/Classify.seqs">Classify.seqs</a> command of the Mothur package.
</p>

<h3>Output</h3>

<p>The analysis output consists of the following:</p>
<ul>
<li>reads-taxonomy-assignment.txt: Taxonomy assignment for each sequence </li> 
<li>classification-summary.txt: Taxonomic outline indicating the number of sequences that were found at each level </li> 
<li>counttable.tsv: Count table where rows are samples and columns are taxonomic groups </li> 
<li>phenodata.tsv: Phenodata file allowing the user to assign samples to experimental groups </li> 
</ul>
</p>


<h3>References</h3>
<p>Schloss PD, Westcott SL, Ryabin T, Hall JR, Hartmann M, Hollister EB, Lesniewski RA, Oakley BB, Parks DH, Robinson CJ, Sahl JW, Stres B, Thallinger GG, Van Horn DJ, Weber CF.
<a href="http://www.ncbi.nlm.nih.gov/pmc/articles/PMC2786419/">Introducing mothur: Open-source, platform-independent, community-supported software for describing and comparing microbial communities. Appl Environ Microbiol, 2009. 75(23):7537-41.</a> 
</p>

</body>

