<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html lang="en">
<head><title></title></head>
<body>

<h2>Translated nucleotide BLAST against users own protein sequence set</h2>

<h3>Description</h3>

<p>This tool runs a nucleotide BLAST search using the given protein sequence set as the target database (blastx). 
The query sequence file can contain up to 100 000 sequences. The size of the target database is not limited.    
</p>

<h3>Parameters</h3>


<li><b>Expectation threshold for saving hits</b> E-value specifies the statistical significance threshold for reporting 
matches against database sequences. The default value 10 means that 10 such matches are expected to be found merely 
by chance. Lower thresholds are more stringent, leading to fewer chance matches being reported.</li>

<li><b>Genetic code</b>. Genetic code to use to translate the query into protein sequence. In addition to the standard genetic code 
you choose one of the mitochondrial or other non-standard genetic codes to be used</li>

<li><b>Word size</b> The length of the seed that initiates an alignment. BLAST works by finding word-matches 
between the query and database sequences. One may think of this process as finding hot-spots that BLAST can 
then use to initiate extensions that might eventually lead to full-blown alignments. For BLASTP searches 
non-exact word matches are taken into account based upon the similarity between words. 
The amount of similarity can be varied so one normally uses just the word-sizes 2 and 3 for these searches.</li>

<li><b>Maximun number of hits to collect per sequence</b> This parameter limits the number of hit sequences 
reported for one query sequence. By default up to 100 hits are reported, but if you wish to collect all the hits, 
and not just the best ones you should in many cases increase this value significantly.</li> 

<li><b>Output format type</b> The BLAST results can be presented in many different formats. The classical BLAST report is not
optimal for big data query sets or in the cases where the results will be analyzed with other tools. In addition to the text based 
BLAST reports, the results can be presented as table or XML file. You can also produce a fasta formatted sequence file containing 
the matching hit sequence regions or a list of hit sequence names.

<li><b>Filter low complexity regions</b> Use SEG program for filtering low complexity regions in the query sequence.</li> 

<li><b>Entrez query to limit search</b> You can use Entrez query syntax to search a subset of the selected BLAST database. 
This can be helpful to limit searches to molecule types, sequence lengths or to exclude organisms.</li>

<li><b>Location on the query sequence</b> Location of the search region in the query sequence, for example: 23-66.</li>

<li><b>Matrix"</b>b> Weight matrix assigns a score for aligning pairs of residues, and determines overall alignment score. 
Experimentation has shown that the BLOSUM62 matrix is among the best for detecting most weak protein similarities. 
For particularly long and weak alignments, the BLOSUM45 matrix may prove superior. 
For proteins, shorter than 85 residues, the BLOSUM80 matrix may provide better hits"</li>

<li><b>Gap opening penalty</b> Cost to open a gap. Integer value from 6 to 25. The default value of this parameter 
depends on the selected scoring matrix. Note that if you assign this value, you must define also the gap extension penalty</li>

<li><b>Gap extension penalty</b> Gap extension penalty.  Integer value from 1 to 3.The default value of 
this parameter depends on the selected scoring matrix. Note that if you assign this value, 
you must define also the gap opening penalty </li>

<li><b>Output a log file</b> Collect a log file for the BLAST run.</li>


<h3>Details</h3>
This tool requires two protein sequence sets as input data. One dataset, the <i>database sequences</i>, is idexed and used as the target database while 
the other is used as a query sequence set




</body>

