<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html lang="en">
<head><title></title></head>
<body>

<h2>Utilities / Define NGS experiment</h2>

<h3>Description</h3>

<p>
Merges a set of count files into a single table and makes a phenodata file for it.
</p>

<h3>Parameters</h3>

<ul>
<li>Type of experiment (ChIP-seq, RNA-seq, miRNA-seq) [RNA-seq]</li>
<li>Does your data contain genomic coordinates (yes, no) [yes]</li>
<li>Count column</li>
<li>Impute missing data (0-1000000) [0]</li>
</ul>

<h3>Details</h3>

<p>
This tool takes a set of count files for individual samples and merges them into one count table. 
You can create the sample count files from BAM files using one of the following tools:
</p>
<ul>
<li>RNA-seq / Count aligned reads per genes with HTSeq</li>
<li>RNA-seq / Count aligned reads per exons for DEXSeq</li>
<li>miRNA-seq / Convert genomic BAM file to count table</li>
<li>miRNA-seq / Convert miRBased BAM file to count table</li>
</ul>
<p>
In addition to merging count files, this tool also creates a phenodata file which allows you to assign samples to different experimental groups for statistical testing etc.
The library size column of the phenodata file gives you an option to enter the initial library size of the samples. If you give the library sizes, 
the differential expression analysis tools edgeR and DESeq tools will use them for normalization, rather than using column totals.
</p>
<p>
If your count files contain columns for genomic location information (chr, start, end), you can add this information to the count table.
</p>
<p>
Since zero counts can cause problems for downstream applications which transform data or calculate ratios, the tool also offers the possibility to impute a fixed value to replace the zeros. 
 
</p>

<h3>Output</h3>

<p>Count table and a corresponding phenodata table.</p>

</body>

