<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html lang="en">
<head><title></title></head>
<body>

<h2>ChIP-seq and FAIRE-seq/  Find motifs with GADEM and match to JASPAR</h2>

<h3>Description</h3>

Given a set of genomic regions, this tool discovers sequence motifs and matches them against the JASPAR database of known transcription factor binding sites. 
For each motif, 10 best matches to JASPAR are reported.

<br>
<br>
<h3>Parameters</h3>

<ul>
<li>P-value cutoff (0...1) [0.0002]</li>
<li>E-value cutoff (0...1) [0.01]</li>
<li>Genome ([BSgenome.Hsapiens.UCSC.hg17, BSgenome.Hsapiens.UCSC.hg18, BSgenome.Hsapiens.UCSC.hg19, BSgenome.Mmusculus.UCSC.mm8, BSgenome.Mmusculus.UCSC.mm9, BSgenome.Rnorvegicus.UCSC.rn4]) [BSgenome.Hsapiens.UCSC.hg18]</li>
</ul>

<br>
<h3>Details</h3>

<p>For a thorough description of the technical details please consult the original publications cited in the References section below. Briefly, the analysis proceeds through the following steps:

<ul>
<li>Given a set of genomic regions, the tool first performs a de novo motif discovery in an unseeded fashion. 
The starting set of position weighted matrices (PWM) is obtained through a combination of space dyads and expectation maximization procedures. 
Further optimization is achieved using genetic algorithms techniques.</li>
<li>The discovered motifs are matched against the PWM of known transcription factors in the JASPAR database.</li>
<li>The top ten best matching transcription factors are reported for each motif, logo plots created, and E-values scoring the match strength are calculated. </li>
</ul>
<br>

<h3>Output</h3>

<p>The analysis output consists of the following:</p>
<ul>
<li>motif-analysis-summary.tsv:	This file summarizes the analysis with information on </li>
<dl>
    <dd>- number of consensus motifs</dd>
    <dd>- number of matches per motif</dd>
    <dd>- matches repartition</dd>
    <dd>- the UPAC nucleotide sequences for the motifs </dd>
    <dd>- number of occurrences of consensus sequence for each motif </dd>
    <dd>- the position weighted matrices for each motif </dd>
</dl>
<br>
<li>logo-plot-(...).pdf: For each consensus motif, the ten best matching transcription factors from the JASPAR database are represented with a logo plot.</li>
</ul>

<br>
<h3>References</h3>

<p>This tool uses the following Bioconductor packages:</p>
<ul>
<li>MotIV<br></li>
<li>rGADEM<br></li>
<li>BSgenome annotation packages</li>
</ul>


<br>For more details refer to these publications:</br>

<br>S. Mahony, P.V. Benos "STAMP: a web tool for exploring DNA-binding motif similarities." Nucl
Acids Res, (2007) 35:W253-258</br>

<br>S Mahony, PE Auron, PV Benos, "DNA familial binding profiles made easy: comparison of
various motif alignment and clustering strategies", PLoS Computational Biology (2007) 3(3):e61</br>

<br>L. Leiping. GADEM: A Genetic Algorithm Guided Formation of
Spaced Dyads Coupled with an EM Algorithm for Motif Discovery.
J Comput Biology, (2009) Feb;16(2):317-29.</br>
</body>

