<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html lang="en">
<head><title></title></head>
<body>

<h2>RNA-seq /  Differential expression analysis using edgeR</h2>

<h3>Description</h3>

Differential expression analysis using the exact statistical methods of the edgeR Bioconductor package ("classic edgeR"). 
Please note that this tool is suitable only for two group comparisons. 
For multifactor experiments you can use the tool "Differential expression using edgeR for multivariate experiments", 
which uses generalized linear models -based statistical methods ("glm edgeR").


<br>
<br>
<h3>Parameters</h3>

<ul>
<li> Column describing groups [group]</li>
<li> Apply normalization (yes, no) [yes]</li>
<li> Dispersion method (common, tagwise) [tagwise]</li>
<li> Dispersion estimate (0-1) [0.1] </li>
<li> Multiple testing correction (none, Bonferroni, Holm, Hochberg, BH, BY) [BH]</li>
<li> P-value cutoff (0-1) [0.05]</li>
<li> Plot width (200-3200 [600]</li>
<li> Plot height (200-3200) [600]</li>
</ul>

<h3><br>Details</h3>

<br>This tool takes as input a table of raw counts from the different samples. The count file has to be associated with a phenodata file describing the experimental groups. 
These files are best created by the tool "Utilities / Define NGS experiment", which combines count files for different samples to one table, and creates a phenodata file for it.
<br><br>
Normalization factors are calculated using the library size given by the user in the phenodata.tsv or by summing the counts for each sample.
TMM method is then used to calculate the normalization factors in order to reduce RNA compositon bias 
(which can arise for example when only a small number of genes
are very highly expressed in one experiment condition but not in the other).
<br><br>
Dispersion is estimated using the quantile-adjusted conditional maximum likelyhood method (qCML). It can estimate a common dispersion for all the genomic features,
or a separate (tagwise) dispersion for each individual feature using an empirical Bayes strategy. 
<br><br>
It is highly recommended to always have at least two biological replicates for each experiment condition.
If this is not possible, one can run the analysis by manually setting
the dispersion factor through the 'Dispersion estimate' parameter.
By default the dispersion estimate is set to 0.1, which is somewhere in-between what is usually observed for technical replicates (0.01) and human data (0.4). 
It is recommended to experiment with different values for this parameter.
<br><br>
Once negative binomial models are fitted and dispersion estimates are obtained, 
edgeR proceeds with testing for differential expression using the exact test, which is based on the qCML methods.
<br><br>

<h3>Output</h3>

<p>The analysis output consists of the following files:</p>


<ul>
<li>de-list-edger.tsv:	Table containing the results of the statistical testing, including fold change estimates and p-values.<br></li>
<li>de-list-edger.bed:  The BED version of the results table contains genomic coordinates and log fold change values.<br></li>
<li>ma-plot-raw-edger.pdf: A scatter plot of the raw count value averages between experiment conditions.<br></li>
<li>ma-plot-normalized-edger.pdf: A scatter plot of the normalized count value averages between experiment conditions. <br></li>
<li>ma-plot-significant-edger.pdf: A scatter plot where the significantly differentially expressed features are highlighted. <br></li>
<li>mds-plot-edger.pdf: A plot showing the results of multidimensional scaling of the data to visualize sample similarities. <br></li>
<li>edger-log.txt: Log file. <br></li>
<li>p-value-plot-edger.pdf: Plot of the raw and adjusted p-value distributions of the statistical test. <br></li>
<br></ul>
<h3><br>References</h3>

<p>This tool uses the edgeR package for statistical analysis. Please read the following article for more detailed information:</p>

<p> MD Robinson, DJ McCarthy, and GK Smyth. <a href="http://www.ncbi.nlm.nih.gov/pubmed/19910308">edgeR: a bioconductor package for differential expression analysis of digital gene expression data. Bioinformatics, 26 (1):139-40, Jan 2010.</a></p> 


<p> </a>.</p> 

</body>
