<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html lang="en">
<head><title></title></head>
<body>

<h2>Normalisation / Illumina with normexp background correction </h2>

<h3>Description</h3>

<p>Normalize Illumina arrays to remove systematic bias using the summary probe profile file and summary control probe profile file generated by GenomeStudio.</p>

<h3>Parameters</h3>

<ul>
<li>Normalization method (none, scale, quantile) [quantile]</li>
<li>Chip type (empty, ...) [empty]</li>
<li>Identifier type (ProbeID) [ProbeID]</li>
<li>Produce flags (yes, no) [no]</li>
<li>Include original annotations (yes, no) [no]</li>
</ul>

<h3>Details</h3>

<p>This tool normalizes Illumina summary probe and summary control probe profile files. 
Import the files DIRECTLY, not via the Import tool. Note that the decimal separator in the files needs to be a dot, not a comma.
This tool performs the normexp background correction using the limma Bioconductor package.</p>

<p>Within array normalization method options include scaling the chips to the same
median, and quantile normalizing the chips to make the expression value follow the same distribution on
all chips. Please note that during normalization the data is also log2-transformed.</p>

<p>If the flags are produced, the Illumina detection values are added to the output table.</p>

<p>You can choose to include the original probe annotations to the results. Please keep in mind that these can be very outdated.</p>

<p>It is obligatory to enter the chiptype. Otherwise annotation-based analyses do not work.</p>

<p>Identifier type specifies how the data was generated in GenomeStudio/BeadStudio. There are typically several probes per each 
gene. TargetIDs summarize all these probes as an expression estimate for one gene. ProbeID does not make
this summarization, and several probes per gene remain in the dataset. TargetIDs are alphanumeric codes, 
such as 0610005A07RIK or GI_10047089-S, whereas ProbeIDs are numerical codes, such as 5570647.</p>

<h3>Output</h3>

<p>A tab-delimited text file containing gene names, expression estimates and optionally call values ("flags") and original annotations. 
This file is suitable for all further analyses.</p>

<h3>References</h3>


<p>This tool uses the Bioconductor package limma.</p>

<p>For normalization, please cite:</p>

<p>Smyth, G. K., and Speed, T. P. (2003). Normalization of cDNA microarray data.
Methods 31, 265-273.</p>

<p>For background correction steps, please cite:</p>

<p>Ritchie, M. E., Silver, J., Oshlack, A., Silver, J., Holmes, M., Diyagama, D.,
Holloway, A., and Smyth, G. K. (2007). A comparison of background correction
methods for two-colour microarrays. Bioinformatics.</p>


</body>

