<!-- saved from url=(0022)http://internet.e-mail -->
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html lang="en">
<head><title></title></head>
<body>


<h2>Normalisation / Illumina - methylumi pipeline</h2>

<h3>Description</h3>

<p>Normalization for Illumina Infinium human methylation 27k arrays using the <a href="http://www.bioconductor.org/packages/2.7/bioc/vignettes/lumi/inst/doc/methylationAnalysis.pdf">lumi methodology</a>.
Please note that this tool requires your raw data to be in one FinalReport file. The filename must end with ".txt", and you should NOT use the Import tool 
when bringing the data into Chipster (in the Import files -window change the action to "Import directly").</p>

<h3>Parameters</h3>

<ul>
<li>Color balance adjustment (none, quantile, ssn) [quantile]</li>
<li>Background correction (none, bgAdjust2C, forcePositive) [none]</li>
<li>Normalization (none, quantile, ssn) [quantile]</li>
<li>Chiptype (human) [human]</li>
<li>Image width (200-3200) [600]</li>
<li>Image height (200-3200) [600]</li>
</ul>

<h3>Details</h3>

<p>This tool adjusts color balance and background, and normalizes Illumina methylation data provided in GenomeStudio FinalReport format. The
color balance can be adjusted using smooth quantile normalization or simple scaling normalization (ssn).
The background adjustment methods bgAdjust2C and forcePositive are based on the assumption that most of the CpG sites are 
unmethylated. Normalization is performed at the probe level (normalizing the intensities
of methylated and unmethylated probes instead of the summarized methylation levels) using either the quantile or ssn method.
</p>

<h3>Output</h3>

<p>Four tab-delimited text files (normalized.tsv, unmethylated.tsv, methylated.tsv, phenodata.tsv) and a quality control plot (QC-plot.png).
The file normalized.tsv contains normalized methylated / unmethylated ratios on log2-scale, and it is suitable for all further analyses. 
Unmethylated.tsv and methylated.tsv contain data for unmethylated and methylated probes. The QC plot shows distibutions of methylated and 
unmethylated probes where unmethylated data is shown in green and methylated data is shown in red.</p>

<h3>References</h3>

<p>This tool uses Bioconductor package lumi. Please cite the articles:</p>

<p>Du P, Kibbe WA, Lin SM (2008) <a href="http://bioinformatics.oxfordjournals.org/cgi/content/abstract/24/13/1547">lumi: a pipeline for processing Illumina microarray, Bioinformatics, 24, 1547-1548</a>.</p> 
<p>Du P, Zhang X, Huang CC, Jafari N, Kibbe WA, Lin SM (2010) <a href="http://www.biomedcentral.com/content/pdf/1471-2105-11-587.pdf">Comparison of Beta-value and M-value methods for quantifying
methylation levels by microarray analysis, BMC bioinformatics, 11, 587</a>.</p> 

</body>

