<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html lang="en">
<head><title></title></head>
<body>

<h2>Normalisation / Illumina SNP arrays</h2>

<h3>Description</h3>

<p>Preprocesses Illumina SNP array genotype call-files and/or calls allels and their corresponding confidence values from Illumina GenCall software output.</p>

<h3>Parameters</h3>

<ul>
<li>Allel calling method</li>
<li>cdfName</li>
<li>Signal-to-noise ratio</li>

</ul>

<h3>Details</h3>

<p>Reformats Illumina GenCall output to Chipster format or computes the genotype values using the CRLMM method as implemented in the package
crlmm. The cdfName parameter defines the CDF name to be used in the CRLMM analysis (human1mduov3b, human1mv1c, human370quadv3c, 
human370v1c, human550v3b, human610quadv1b, human650v3a, human660quadv1a, humanomni1quadv1b, humanomniexpress12v1b). By default, CDF names 
are extracted from the chips themselves for CRLMM analysis. In the case, your chip is not among the listed ones, contact support.
</p>

<p>
In the CRLMM implementation, bad-quality chips are filtered out from the data based on their signal-to-noise ratio. Lowering the cutoff can 
decrease the overall quality of SNP calls but will increase the amount of chips passed to further analysis.
</p>

<h3>Output</h3>

<p>A tab-delimited text file containing gene names and expression estimates. This file is
suitable for all further analyses.</p>

<h3>References</h3>

<p> This tool uses Bioconductor package oligo. You can cite the package as:</p>

<p>Carvalho B, Bengtsson H, Speed TP, Irizarry RA. Exploration, normalization, and genotype calls of high density
oligonucleotide snp array data. Biostatistics, Dec 2006.</p>

<p>Carvalho BS, Louis TA, Irizarry RA. Quantifying uncertainty in genotype calls. Bioinformatics. 2010, 26:242-9.</p>

<p>Ritchie ME, Carvalho BS, Hetrick KN, Tavaré S, Irizarry RA. R/Bioconductor software for Illumina's Infinium 
whole-genome genotyping BeadChips. Bioinformatics. 2009, 25:2621-3.</p>

</body>


