<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html lang="en">
<head><title></title></head>
<body>

<h2>Normalisation / Affymetrix</h2>

<h3>Description</h3>

<p>Preprocesses Affymetrix CEL-files to give expression estimates and call values for genes.</p>

<h3>Parameters</h3>

<ul>
<li>Normalisation method (MAS5, PLIER, RMA, GCRMA, Li-Wong) [RMA]</li>
<li>Stabilize variance (yes, no) [no]</li>
<li>Custom chiptype (Use original Affymetrix annotations, hgu133a, hgu133av2...) [Use original Affymetrix annotations]</li>
</ul>

<h3>Details</h3>

<p>There are five different normalization methods available. Regardless of method, the normalized data will always be expressed in base 2 log-transformed scale.</p>

<p>MAS5 is the method originally implemented by Affymetrix in their GCOS software.
Plier is a more recent method developed by Affymetrix that results in more precise estimates of lower intensity probes than MAS5.
Variance stabilizing normalization can be combined with these methods to further improve the precision of low intensity expression estimates.</p>

<p>Methods RMA, GCRMA and Li-Wong give gene expression estimates that have some nice statistical properties, making the data more closely resemble a normal distribution and improving on the homogeneity of variances.
Compared to MAS5 they have been shown to result in greater statistical power in downstream analyses.
Note that the parameter setting for variance stabilization does not have an effect here, since it's an inherent part of these methods.</p>

<p>Call values (present, absent, marginal) are always calculated using the MAS5 algorithm regardless of the selected 
preprocessing algorithm.</p>

<p>It is a known problem with Affymetrix expression arrays that a sizeable part of the probes have been mis-annotated 
to wrong genes. The probes have been re-annotated according to the current knowledge to 'correct' genes in alternative
CDF environments. If the user wants to preprocess the Affymetrix chips using these reannotations, then one of the custom
chiptypes needs to be selected. The reannotations are from <a href="http://brainarray.mbni.med.umich.edu/Brainarray/Database/CustomCDF/genomic_curated_CDF.asp">University of Michigan</a>.
If you choose to use the original Affymetrix annotations, the tool will use the information in the .CEL-files to determine the chiptype. 
NOTE: If your files are from Gene or Exon array, please use the corresponding normalization tools:
<a href=" http://chipster.csc.fi/manual/norm-affy-exon.html">Normalisation / Affymetrix exon arrays</a> 
<a href=" http://chipster.csc.fi/manual/norm-affy-gene.html">Normalisation / Affymetrix gene arrays</a> 
 </p>
 
<h3>Output</h3>

<p>A tab-delimited text file containing gene names, expression estimates and call values ("flags"). This file is
suitable for all further analyses.</p>

<h3>References</h3>

<p>Affymetrix. Affymetrix Microarray Suite User Guide. Affymetrix, Santa Clara, CA,
version 5 edition, 2001.</p>
<p>Rafael A. Irizarry, Bridget Hobbs, Francois Collin, Yasmin D. Beazer-Barclay, Kristen J.
Antonellis, Uwe Scherf, and Terence P. Speed. Exploration, normalization, and sum-
maries of high density oligonucleotide array probe level data. Biostatistics, 2003. Apr;4(2):249-64.</p>
<p>C. Li and W.H. Wong. Model-based analysis of oligonucleotide arrays: Expression index
computation and outlier detection. Proceedings of the National Academy of Science
U S A, 98:31{36, 2001.</p>
</body>

