<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html lang="en">
<head><title></title></head>
<body>

<h2>Normalisation / Affymetrix SNP arrays</h2>

<h3>Description</h3>

<p>Preprocesses Affymetrix SNP array CEL-files to call allels and their corresponding confidence values.</p>

<h3>Parameters</h3>

<ul>
<li>CRLMM-version</li>
<li>cdfName</li>
<li>Signal-to-noise ratio</li>

</ul>

<h3>Details</h3>

<p>Computes the genotype values using the SNPRMA and CRLMM methods. At present, two implementations exists for the CRLMM function. An old
version is available in package oligo and a newer, improved version in the package crlmm. If the new version of CRLMM has been selected, the 
parameter cdfName defines the CDF name to be used in the analysis ('GenomeWideSnp5' or 'GenomeWideSnp6'). By default, this information is 
extracted from the chip.
</p>

<p>
If the new version of CRLMM function has been selected, bad-quality chips are filtered out from the data based on their signal-to-noise ratio. 
Lowering the cutoff can decrease the overall quality of SNP calls but will increase the amount of chips passed to further analysis.
</p>

<h3>Output</h3>

<p>A tab-delimited text file containing gene names and expression estimates. This file is
suitable for all further analyses.</p>

<h3>References</h3>

<p> This tool uses Bioconductor package oligo. You can cite the package as:</p>

<p>Carvalho BS, Bengtsson H, Speed TP, Irizarry RA. Exploration, normalization, and genotype calls of high density
oligonucleotide snp array data. Biostatistics, Dec 2006.</p>

<p>Carvalho BS, Louis TA, Irizarry RA. Quantifying uncertainty in genotype calls. Bioinformatics. 2010, 26:242-9.</p>

</body>


