<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html lang="en">
<head><title></title></head>
<body>

<h2>Filter reads for several criteria</h2>

<h3>Description</h3>

<p>This tool filters reads for several criteria such as length, low complexity, Ns and duplicates. The input data can be in FASTQ or FASTA format. 
Note that if you paired-end reads and you need to keep them in the matching order, you should give both files to this tool at the same time. The tool will automatically 
make sure that only filtered pairs are given in the output.
</p>

<h3>Details</h3>
<p>
Only those filtering methods are used for which you assign a value. 
You can use several filtering methods at the same time. For example filtering with parameters 
<b>Maximum length: 100</b> and <b>Maximum count of Ns: 1</b>
would produce a reads set where all the reads are shorter than 101 bases 
and they contain less than 2 Ns. For detailed description of the different filtering conditions, please check the
manual  of the  <a href="http://prinseq.sourceforge.net/">PRINSEQ package</a>.
From the table below, you can check what PRINSEQ command line options the parameter definitions 
correspond to:
<p>
<table>
<tr><td><b>Parameter name</b></td><td><b>Command line</b></td><td><b>Description</b> </td></tr>
<tr><td>Max length</td><td>-max_len</td><td> Select only reads that are shorter than the given value.</td></tr>
<tr><td>Min length</td><td>-min_len</td><td> Select only reads that are longer than the given value.</td></tr>
<tr><td>Max GC content</td><td>-max_gc</td><td> Select only reads that has GC content that is less than the given value.</td></tr>
<tr><td>Min GC content</td><td>-min_gc</td><td> Select only reads that has GC content that is more than the given value.</td></tr>
<tr><td>Min quality score </td><td>-min_qual_score </td><td>Filter reads with GC content below than the given value.</td></tr>
<tr><td>Max quality score</td><td>-max_qual_score</td><td> Filter reads with GC content above then the given value.</td></tr>
<tr><td>Min mean quality</td><td>-min_qual_mean</td><td> Filter reads with quality score mean below the given value.</td></tr>
<tr><td>Max mean quality </td><td>-max_qual_mean </td><td>Filter reads with quality score mean above the given value.</td></tr>
<tr><td>Max percentage of Ns </td><td>-ns_max_p</td><td> Filter reads for which the percentage of Ns is higher than the given value.</td></tr>
<tr><td>Max count of Ns</td><td>-ns_max_n </td><td>  Filter reads for which the count of Ns is higher than the given value.</td></tr>
<tr><td>Max number of reads</td><td>-seq_num </td><td> Only keep the given number number of reads that pass all other filters.</td></tr>
<tr><td>Type of duplicates to filter</td><td>-derep </td><td>  Type of duplicates to filter.</td></tr>
<tr><td>Number of allowed duplicates</td><td>-derep_min </td><td>This option specifies the number of allowed duplicates. For example, to remove reads that occur more than 2 times, you would specify value 3.</td></tr>
<tr><td>DUST filter threshold</td><td> -lc_method dust, -lc_threshold </td><td>Use DUST algorithm with the given maximum alloed low complexity score, between 0 and 100.</td></tr>
<tr><td>Entropy filter threshold </td><td>-lc_method entropy, -lc_threshold </td><td> Use Entropy algorithm with the given minimum allowedlow complexity score, between 0 and 100.</td></tr>
<tr><td>Base quality encoding</td><td>-phred64</td><td> Select "Sanger" for Illumina 1.8+, Sanger, Roche/454, Ion Torrent and PacBio data.</td></tr>
</table>

</p> 
<h3>Output</h3>

<p>
The reads that pass the filtering condition are saved to file called accepted.fastq or accepted.fasta.
You can also choose to store the reads that are filtered out to a separate file (rejected.fastq or rejected.fasta).
You can also print out a log file that contains information about the filtering task and
statistics about how many reads were accepted and rejected.
</p>


<h3>Reference</h3>

This tool is based on the <a href="http://prinseq.sourceforge.net/">PRINSEQ package</a>.

</body>

