<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html lang="en">
<head><title></title></head>
<body>

<h2>Utilities / Count alignments in BAM</h2>

<h3>Description</h3>

<p>
Given a BAM file, this tool counts how many alignments it contains, taking the mapping quality into account if needed.
</p>

<h3>Parameters</h3>

<ul>
<li>Minimum mapping quality required for counting (0-1000) [0]</li>
</ul>

<h3>Details</h3>

<p> 
By default all alignments are counted, but users can choose to count only alignments which have the specified
mapping quality or higher. Note that a read can have several alignments, so this tool doesn't give you the number of reads mapped.
</p>


<h3>Output</h3>

<p>
Output is a text file containing the number of alignments.
</p>


<h3>References</h3>

<p>This tool is based on the <a href = "http://samtools.sourceforge.net/">SAMtools package</a>. 
Please cite the article <a href = "http://www.ncbi.nlm.nih.gov/pubmed/19505943">The Sequence alignment/map (SAM) format and SAMtools</a> by Li H., Handsaker B., Wysoker A., Fennell T., Ruan J., Homer N., Marth G., Abecasis G., Durbin R. and 1000 Genome Project Data Processing Subgroup (2009) Bioinformatics, 25, 2078-9. [PMID: 19505943]. 
</p>

</body>

