<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html lang="en">
<head><title></title></head>
<body>

<h2>Variants / Call SNPs and short INDELs</h2>

<h3>Description</h3>

<p>Calls SNPs and short INDELs for one or more diploid individuals.</p>

<h3>Parameters</h3>

<ul>
<li>Reference sequence () [human]</li>
<li>Chromosome names in my BAM file look like (1, chr1) [chr1]</li>
<li>Call only SNPs, INDELs not considered (yes, no) [no]</li>
<li>Minimum read depth (Integer) [2]</li>
<li>Maximum read depth (Integer) [1000]</li>
<li>Minimum mapping quality for an alignment to be used (Integer) [0]</li>
<li>Minimum base quality for a base to be considered (Integer) [20]</li>
<li>Disable probabilistic realignment for computation of BAC (yes, no) [no]</li>
<li>Downgrading coefficient (Integer) [0]</li>
<li>Call variants only for a certain region (String) []</li>
<li>Output per sample read depth (yes, no) [no]</li>
<li>Output per sample number of high-quality non-reference bases (yes, no) [no]</li>
<li>Output all INFO fields (yes, no) [no]</li>
</ul>

<h3>Details</h3>

<p> 
You can provide your own reference sequence in FASTA format or choose one of the provided reference genomes. 
If you use the reference genomes provided by Chipster, please indicate the style of chromosome naming in your BAM file (e.g. chr1 or 1). 
</p>
<p>
Alignment records are grouped by sample identifiers in @RG header lines. If sample identifiers are absent, each input file is regarded as one 
sample.
</p>
<p> 
Maximum read depth should be adjusted to about twice the average read depth.
</p>
<p>
You can tailor the FORMAT filed in the output VCF by adding per sample number of high-quality bases (DP) and high-quality non-reference bases (DV).
Also the INFO fields can be modified: By default only AC, DP, DP4, INDEL and IDV fields are included in the output. If you select yes, all the samtools' default INFO fields are included. These are AC, AN, BQB, DP, DP4, HDB, ICB, IDV, IMF, INDEL, MQ, MQB, MQOF, MQSB, RPB, SGB and VDB.
</p>


<h3>Output</h3>

<p>
Output is a VCF file. You can view VCF files in Chipster as text (which includes the header) or spreadsheet (which can be shorted). 
The location column in the spreadsheet can be used in Chipster genome browser to navigate quickly through a list of variants.
</p>


<h3>References</h3>

<p>This tool is based on the <a href = "http://samtools.sourceforge.net/">SAMtools package</a>. 
Please cite the article <a href = "http://www.ncbi.nlm.nih.gov/pubmed/19505943">The Sequence alignment/map (SAM) format and SAMtools</a> by Li H., Handsaker B., Wysoker A., Fennell T., Ruan J., Homer N., Marth G., Abecasis G., Durbin R. and 1000 Genome Project Data Processing Subgroup (2009) Bioinformatics, 25, 2078-9. [PMID: 19505943]. 
</p>

</body>

