<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html lang="en">
<head><title></title></head>
<body>

<h2>Display a multiple sequence alignment as formatted text</h2>

<h3>Description</h3>

<p>Display a multiple sequence alignment in pretty format</p>
<p> 
   This tool runs EMBOSS command <b>showalign</b> that reads a set of aligned protein or a nucleic acid sequences,
   and writes them to file in formatted text file. Similarities and differences of each sequence to a
   reference sequence are highlighted for specified types of matches. The
   reference sequence can be the calculated consensus sequence (default)
   or one of the input set (specified by name or the ordinal number of
   that sequence in the file). The output sequences can be displayed in
   either the input order (the default), sorted in order of their
   similarity to the reference sequence, or sorted alphabetically by their
   names. There are many other options to control the content and format
   of the output.
  </p> 
   


<p>
The input sequences can be in any of the <a href="http://emboss.sourceforge.net/docs/themes/AlignmentFormats.html">EMBOSS compatible sequence alignment formats</a>.
Note that this tool reads just one input file, that should contain all the sequences to be analyzed.
</p>
<p>
For more details, please check the manual pages of <a href="http://emboss.sourceforge.net/apps/release/6.5/emboss/apps/showalign.html">showalign</a> command.
</p>


<h3>Output</h3>

<p>
Output is a fasta formatted sequence file.
</p>


</body>

