<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html lang="en">
<head><title></title></head>
<body>

<h2>Skip redundant sequences</h2>

<h3>Description</h3>

<p>Remove redundant sequences from an input set</p>
<p> 
   This tool runs EMBOSS command <b>skipredundant</b> that removes
   redundancy from an input file of sequences, either at a single
   threshold of sequence similiarty (e.g. 40%) or within a threshold range
   of sequence similiarty (e.g. 40% - 70%). Redundancy is calculated for each 
   pair of sequences in turn and thus this tool is not suitable for large data sets.
   </p>
   <p>
    Redundancy in a collection of sequences occurs when
   one or more similar sequences are present. The inclusion of very
   similar sequences in certain analyses will introduce undesirable bias.
   For example, a family may possess 100 sequences in the sequence
   database, but 90 of these might be essentially the same sequence, e.g.
   very close relatives or mutations of a single sequence. Although 100
   sequences are known, the family only contains 11 sequences that are
   essentially unique. For many applications it is desirable or even
   essential to remove redundant sequences from a set in order to produce
   a smaller set that is representative of the whole.
  </p> 
   


<p>
The input sequences can be in any of the <a href="http://emboss.sourceforge.net/docs/themes/AlignmentFormats.html">EMBOSS compatible sequence alignment formats</a>.
Note that this tool reads just one input file, that should contain all the sequences to be analyzed.
</p>
<p>
For more details, please check the manual pages of <a href="http://emboss.sourceforge.net/apps/release/6.5/emboss/apps/skipredundant.html">skipredundant</a> command.
</p>


<h3>Output</h3>

<p>
Output is a fasta formatted sequence file.
</p>


</body>

