<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html lang="en">
<head><title></title></head>
<body>

<h2>Statistics / Two group tests</h2>

<h3>Description</h3>

Test every gene's expression for a difference between two groups.

<h3>Parameters</h3>

<ul>
<li>Test type (empiricalBayes, t-test, F-test, Mann-Whitney, LPE) [t-test] </li>
<li>Method of multiple testing correction (none, Bonferroni, Holm, Hochberg, BH, BY) [BH]</li>
<li>P-value cut-off (0...1) [0.05]</li>
</ul>

<h3>Details</h3>

<p>Runs the specified test, and returns the genes that have a corrected p-value smaller than the specified 
p-value cut-off. If you want to get a p-value for every single gene in your data set, set the p-value cut-off to 1. This
should return a new gene list of equal length to the original one.</p>

<p>Test empiricalBayes is esssentially a t-test where estimates are shrank using Bayesian methods. 
This results into better estimates of individual gene's statistical significance. In addition, the empiricalBayes
is much faster than the usual t-test. LPE test is especially well suited for small sample sizes. All other test, except
F-test, compare differences in two groups mean expression. F-test compares the differences in variance between the two 
groups.</p>

<p>If interested in performing a paired t, Mann-Whitney or empiricalBayes test, the phenodata column describing the sample pairs
must be defined with the pairing parameter. In the phenodata matrix, a unique identifier (like 1,2,3 or pair1,pair2,pair3, or patient1,
patient2, patient3, etc) must be given for each pair of samples. Please note that if a sample does not have a pair, it is not used 
in assessing the significance.
</p>

<p>Multiple testing correction options are Bonferroni, Holm, and Hochberg for family-wise error rate (FWER) and
Benjamini-Hochberg and Benjamini-Yakutieri for false discovery rate (FDR). Of these Bonferroni is the most
conservative, returning the smallest number of significant results, and FDR-based adjustments are less
conservative, and return more significant results.</p>

<h3>Output</h3>

<p>A text file containing the gene expression values and the p-value for the test. </p>


<h3>References</h3>

<p>This tool uses Bioconductor packages limma and LPE. Please cite the following article, if you used empirical Bayes
from limma package:</p>

<p>Smyth, G. K. (2004). Linear models and empirical Bayes methods for assessing
differential expression in microarray experiments. Statistical Applications in Ge-
netics and Molecular Biology, Vol. 3, No. 1, Article 3.</p>

<p>Cite the following articles, if you used LPE-test from LPE package:</p>

<p>Jain et. al. Local-pooled-error test for identifying differentially expressed genes with
a small number of replicated microarrays, Bioinformatics, 2003, Vol 19, No. 15, pp:
1945-1951.</p>

<p>Jain et. al. Rank-invariant resampling based estimation of false discovery rate for analysis
of small sample microarray data, BMC Bioinformatics, 2005, Vol 6, 187.</p>


</body>

