<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html lang="en">
<head><title></title></head>
<body>

<h2>Statistics / Time series</h2>

<h3>Description</h3>

Finds periodically expressed genes from the selected gene list.

<h3>Parameters</h3>

<ul>
<li>Analysis type (periodicity, ica, maSigPro) [periodicity]</li>
<li>Time column (Phenodata column describing the time to test)</li>
<li>Replicate column for maSigPro (Phenodata column indicating the replicate groups)</li>
<li>First experimental column for maSigPro (For maSigPro, the first column giving the assignment of arrays to experimental groups)</li>
<li>Last experimental column for maSigPro (For maSigPro, the first column giving the assignment of arrays to experimental groups)</li>
<li>P-value cut-off (0...1) [0.05]</li>
<li>P-value adjustment method (none, Bonferroni, Holm, Hochberg, BH, BY) [BH]</li>
<li>Standard deviation for ica (0...10) [2]</li>
<li>k for maSigPro [9]</li>
<li>degree for maSigPro [2]</li>
<li>rsq for maSigPro [0.7]</li>
<li>Image width (200...3200) [600]</li>
<li>Image height (200...3200) [600]</li>
</ul>

<h3>Details</h3>

<p>Periodically expressed genes in a single time course microarray experiment can be found using either the Fisher's G test or
independent component analysis (ICA). If your data includes multiple time series for different experimental conditions (e.g. one for 
control samples and another one for treated ones) use maSigPro. For the Fisher's test and maSigPro, genes that have the
corrected p-value less than the specified threshold are returned. For ICA, genes that are within the specified standard deviation 
from the mean, are returned.</p>

<p>In order to use maSigPro, you need to have a phenodata-matrix containing columns for the time variable, replicate variable and one for each
experimental group. The time column should describe the time-point of the chip. Replicates column is an index column that indicates the replicated 
arrays: all arrays belonging to the same experimental condition must be given the same number. In addition to these two, the phenodata should include 
binary columns that give the assignment of arrays to experimental groups. There are as many binary columns as experimental groups and arrays take
the value 1 or 0 whether they belong or not to that experimental group. The experiment-group columns should be side by side in the phenodata and you
should mark the first (left-most) and the last (right-most) such column by using variables First/Last experimental column for maSigPro. For more help,
please see the manual of maSigPro.
</p>

<img src="masigpro.png">

<h3>Output</h3>

<p>A list of genes expression and p-values (periodicity analysis) or a list of genes
with expression values and their cluster membership (ICA). For ICA, a plot of genes
in the clusters is also generated.</p>

<h3>References</h3>

<p>Please cite the following article, if you used maSigPro package:</p>

<p>Conesa,A. (2006). maSigPro: a method to identify significantly differential expression profiles in time-course microarray experiments. Bioinformatics, 
Vol 22, No 9, 1096-1102.</p>


</body>

