<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html lang="en">
<head><title>Chipster manual</title></head>
<body>

<h2>Analysis tools for next generation sequencing data</h2>

<p>Chipster's NGS analysis tools are grouped in the categories listed below. Each analysis tool has its own detailed manual page, which consists of four sections: 
General description, parameters, details, output and reference.
 The parameter section contains the list of parameters, their possible values (1-100) and their default values [10]. The manual for Chipster's built-in genome browser can be found <a href="genomeBrowser.html">here</a>.</p>

<ol>
<li> <a href="#filtering">Filtering</a> </li>
<li> <a href="#quality">Quality control</a> </li>
<li> <a href="#utilities">Utilities</a> </li>
<li> <a href="#regions">Matching genomic regions</a> </li>
<li> <a href="#alignment">Alignment</a> </li>
<li> <a href="#RNA">RNA-seq</a> </li>
<li> <a href="#miRNA">miRNA-seq</a> </li>
<li> <a href="#ChIP">ChIP-seq</a> </li>
<li> <a href="#CNA">CNA-seq</a> </li>
<li> <a href="#methyl">Methyl-seq</a> </li>
</ol>
<hr>
<ul>
<b><a name="filtering">Filtering</a></b>
<li> <a href="fastx-quality-filter.html">Filter out low quality reads</a> </li>
<li> <a href="fastx-clipper.html">Filter reads for adapters, length and Ns</a></li>
<li> <a href="ngs-filter-annotations.html">Filter by column term </a> </li>
<li> <a href="ngs-filter-results-column.html">Filter by column value </a> </li>
<br>
<br>
<b><a name="quality">Quality control</a></b>
<li> <a href="fastqc.html">Read quality with FastQC</a></li>
<li> <a href="fastx-statistics.html">Quality statistics with FASTX</a></li>
<br>
<br>
<b><a name="utilities">Utilities</a></b>
<li> Define NGS experiment </li>
<li> <a href="fastx-trimmer.html">Trim reads</a> </li>
<li> <a href="fastx-renamer.html">Rename reads</a> </li>
<li> <a href="fastx-collapser.html">Remove duplicate reads from FASTQ or FASTA</a> </li>
<li> <a href="samtools-rmdup.html">Remove duplicate reads from BAM </a> </li>
<li> <a href="samtools-index.html">Index BAM </a> </li>
<li> <a href="samtools-sort-index-BAM.html">Sort and index BAM</a> </li>
<li> <a href="samtools-convert-SAMtoBAM-sort-index.html">Convert SAM to BAM, sort and index</a> </li>
<li> <a href="bedtools-bamtobed.html">Convert BAM to BED </a> </li>
<li> <a href="fastx-fastq-to-fasta.html">Convert FASTQ to FASTA</a> </li>
<li> <a href="samtools-merge.html">Merge BAM </a> </li>
<li> <a href="samtools-idxstats.html">Statistics for BAM </a> </li>
<li> <a href="samtools-subset.html">Make a subset of BAM  </a> </li>
<li> <a href="samtools-count.html">Count alignments in BAM  </a> </li>
<li> <a href="bedtools-sort.html">Sort BED </a> </li>
<li> <a href="bedtools-fastafrombed.html">FASTA from BED </a> </li>
<li> <a href="bedtools-slopbed.html">Slop BED </a> </li>
<li> Extract features from BED </li>

<br>
<br>
<b><a name="regions">Matching genomic regions</a></b>
<li> <a href="FindOverlappingTool.html">Find overlapping regions </a> </li>
<li> <a href="RemoveOverlappingTool.html">Remove overlapping regions </a> </li>
<li> <a href="CombineRegionsTool.html">Combine region files </a> </li>
<li> <a href="bedtools-subtractbed.html">Subtract BED </a> </li>
<li> <a href="bedtools-intersectbed.html">Intersect BED </a> </li>
<li> <a href="bedtools-closestbed.html">Closest BED </a> </li>
<li> <a href="bedtools-windowbed.html">Window BED </a> </li>
<li> <a href="bedtools-mergebed.html">Merge BED </a> </li>
<li> <a href="bedtools-coveragebed.html">Coverage BED </a> </li>
<li> <a href="bedtools-genomecoveragebed.html">Genome coverage BED </a> </li>
<li> <a href="bedtools-overlap.html">Overlap BED </a> </li>
<li> <a href="bedtools-pairtobed.html">Overlap BEDPE with BED </a> </li>
<li> <a href="bedtools-pairtopair.html">Compare two BEDPE files </a> </li>
<br>
<br>
<b><a name="alignment">Alignment</a></b>
<li> <a href="bowtie.html">Bowtie for single-end reads </a> </li>
<li> <a href="bowtie-paired-end.html">Bowtie for paired-end reads</a> </li>
<li> <a href="bowtie-with-index-building.html">Bowtie for single-end reads and own genome </a> </li>
<li> <a href="bowtie-paired-end-with-index-building.html">Bowtie for paired-end reads and own genome </a> </li>
<li> <a href="bwa.html">BWA for single-end reads </a> </li>
<li> <a href="bwasw.html">BWA-SW for single-end reads </a> </li>
<li> <a href="bwa-paired-end.html">BWA for paired-end reads </a> </li>
<li> <a href="bwa-with-index-building.html">BWA for single-end reads and own genome</a> </li>
<li> <a href="bwasw-with-index-building.html">BWA-SW for single-end reads and own genome</a> </li>
<li> <a href="bwa-paired-end-with-index-building.html">BWA for paired-end reads and own genome </a> </li>
<li> <a href="tophat.html">TopHat for paired end reads </a> </li>
<li> <a href="tophat-single-end.html">TopHat for single end reads </a> </li>
<br>
<br>
<b><a name="RNA">RNA-seq</a></b>
<li> <a href="htseq-count.html">Map aligned reads to genes with HTSeq</a> </li>
<li> <a href="htseq-count-own-gtf.html">Map aligned reads to genes with HTSeq using own GTF </a> </li>
<!--  <li> Map aligned reads to genes with coverageBed using own BED </li>-->
<li> Differential expression analysis using edgeR</li>
<li> Differential expression analysis using Cufflinks</li>
<br>
<br>
<b><a name="miRNA">miRNA-seq</a></b>
<li> Convert genomic BAM to edgeR input format </li>
<li> Convert miRBased BAM to edgeR input format </li>
<li> Differential expression analysis using edgeR</li>
<li> Find miRNA targets </li>
<li> GO enrichment for miRNA targets</li>
<li> KEGG enrichment for miRNA targets</li>
<li> Correlate miRNA with target expression</li>
<li> Up-down analysis of miRNA targets</li>
<li> Extract miRNA target expression data</li>
<br>
<br>
<b><a name="ChIP">ChIP-seq</a></b>
<li> <a href="ngs-find-peaks-macs-one.html">Find peaks using MACS, treatment only </a> </li>
<li> <a href="ngs-find-peaks-macs-two.html">Find peaks using MACS, treatment vs. control </a> </li>
<li> <a href="ngs-find-motifs-jaspar.html">Find common motifs and match to JASPAR </a> </li>
<li> <a href="ngs-find-nearest-genes.html">Find the nearest genes for regions </a> </li>
<li> Find unique and annotated genes</li>
<li> <a href="pathways-ngs-hyperg-go.html">GO enrichment for list of genes </a> </li>
<br>
<br>
<b><a name="CNA">CNA-seq</a></b>
<li> Count reads in fixed bins </li>
<li> Plot copy number profile </li>
<li> Compare to reference sample</li>
<li> Define CNA-seq experiment </li>
<br>
<br>
<b><a name="methyl">Methyl-seq</a></b>
<li> <a href="convertBAMtoMEDIPS.html">Convert BAM file to MEDIPS input format </a> </li>
<li> <a href="combineMEDIPS.html">Combine MEDIPS input files </a> </li>
<li> <a href="MEDIPS_one-sample.html">MEDIPS -methylation analysis, treatment only </a> </li>
<li> <a href="MEDIPS_two-samples.html">MEDIPS -methylation analysis, treatment and control </a> </li>




      


<!--        <li> <a href=" "> </a> </li>        -->
</ul>

</body>

</html>