<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html lang="en">
<head><title>Chipster manual</title></head>
<body>

<h2>Analysis tools for next generation sequencing data</h2>

<p>Chipster's NGS analysis tools are grouped in the categories listed below. Chipster also contains a <a href="genomeBrowser.html">built-in genome browser</a>,
allowing you to view reads and results in their genomic context.</p>

<ol>
<li> <a href="#quality">Quality control</a> </li>
<li> <a href="#filtering">Filtering</a> </li>
<li> <a href="#utilities">Utilities</a> </li>
<li> <a href="#regions">Matching genomic regions</a> </li>
<li> <a href="#alignment">Alignment</a> </li>
<li> <a href="#variants">Variants</a> </li>
<li> <a href="#RNA">RNA-seq</a> </li>
<li> <a href="#miRNA">miRNA-seq</a> </li>
<li> <a href="#ChIP">ChIP-seq and FAIRE-seq</a> </li>
<li> <a href="#CNA">CNA-seq</a> </li>
<li> <a href="#methyl">Methyl-seq</a> </li>
</ol>
<hr>
<ul>
<b><a name="quality">Quality control</a></b>
<li> <a href="fastqc.html">Read quality with FastQC</a></li>
<li> <a href="fastx-statistics.html">Quality statistics with FASTX</a></li>
<li> <a href="prinseq-statistics.html">Read quality statistics with PRINSEQ </a> </li>
<br>
<br>
<b><a name="filtering">Filtering</a></b>
<li> <a href="fastx-quality-filter.html">Filter out low quality reads (FASTX)</a> </li>
<li> <a href="prinseq-length-filter.html">Filter reads for length (PRINSEQ)</a> </li>
<li> <a href="prinseq-N-filter.html">Filter reads for Ns (PRINSEQ)</a> </li>
<li> <a href="prinseq-complexity-filter.html">Filter reads for low complexity (PRINSEQ)</a> </li>
<li> <a href="prinseq-duplicate-filter.html">Filter reads for duplicates (PRINSEQ)</a> </li>
<li> <a href="prinseq-filter.html">Filter reads for several criteria (PRINSEQ)</a> </li>
<li> <a href="fastx-clipper.html">Filter reads for adapters, length and Ns (FASTX)</a></li>
<li> <a href="ngs-filter-annotations.html">Filter by column term </a> </li>
<li> <a href="ngs-filter-results-column.html">Filter by column value </a> </li>
<br>
<br>
<b><a name="utilities">Utilities</a></b>
<li> <a href=" ngs-create-experiment.html">Define NGS experiment </a> </li>
<li><a href="tagcleaner-predict.html">Predict primers/adaptors (TagCleaner)</a></li>
<li><a href="tagcleaner-statistics.html">Statistics for primers/adaptors (TagCleaner)</a></li>
<li><a href="tagcleaner-trim.html">Trim primers/adaptors (TagCleaner)</a></li>
<li> <a href="fastx-trimmer.html">Trim reads (FASTX)</a> </li>
<li> <a href="prinseq-quality-trimmer.html">Trim reads by quality (PRINSEQ) </a> </li>
<li> <a href="prinseq-AT-trimmer.html">Trim reads for poly-A/T tails (PRINSEQ) </a> </li>
<li> <a href="prinseq-trimmer.html">Trim reads for several criteria (PRINSEQ) </a> </li>
<li> <a href="fastx-renamer.html">Rename reads</a> </li>
<li> <a href="samtools-rmdup.html">Remove duplicate reads from BAM </a> </li>
<li> <a href="samtools-index.html">Index BAM </a> </li>
<li> <a href="samtools-sort-index-BAM.html">Sort and index BAM</a> </li>
<li> <a href="samtools-convert-SAMtoBAM-sort-index.html">Convert SAM to BAM, sort and index</a> </li>
<li> <a href="bedtools-bamtobed.html">Convert BAM to BED </a> </li>
<li> <a href="fastx-fastq-to-fasta.html">Convert FASTQ to FASTA</a> </li>
<li> <a href="samtools-merge.html">Merge BAM </a> </li>
<li> <a href="samtools-idxstats.html">Statistics for BAM </a> </li>
<li> <a href="samtools-subset.html">Make a subset of BAM  </a> </li>
<li> <a href="samtools-count.html">Count alignments in BAM  </a> </li>
<li> <a href="samtools-consensus.html">Create consensus sequence from BAM </a> </li>
<li> <a href="annotate-variants.html">Annotate variants </a> </li>
<li> <a href="bedtools-sort.html">Sort BED </a> </li>
<li><a href="sort-gtf.html">Sort GTF</a></li>
<li><a href="sort-vcf.html">Sort VCF</a></li>
<li> <a href="bedtools-fastafrombed.html">FASTA from BED </a> </li>
<li> <a href="bedtools-slopbed.html">Slop BED </a> </li>
<li> Extract features from BED </li>
<br>
<br>
<b><a name="regions">Matching genomic regions</a></b>
<li> <a href="FindOverlappingTool.html">Find overlapping regions </a> </li>
<li> <a href="RemoveOverlappingTool.html">Remove overlapping regions </a> </li>
<li> <a href="CombineRegionsTool.html">Combine region files </a> </li>
<li> <a href="bedtools-subtractbed.html">Subtract BED </a> </li>
<li> <a href="bedtools-intersectbed.html">Intersect BED </a> </li>
<li> <a href="bedtools-closestbed.html">Closest BED </a> </li>
<li> <a href="bedtools-windowbed.html">Window BED </a> </li>
<li> <a href="bedtools-mergebed.html">Merge BED </a> </li>
<li> <a href="bedtools-coveragebed.html">Coverage BED </a> </li>
<li> <a href="bedtools-genomecoveragebed.html">Genome coverage BED </a> </li>
<li> <a href="bedtools-overlap.html">Overlap BED </a> </li>
<li> <a href="bedtools-pairtobed.html">Overlap BEDPE with BED </a> </li>
<li> <a href="bedtools-pairtopair.html">Compare two BEDPE files </a> </li>
<br>
<br>
<b><a name="alignment">Alignment</a></b>
<li> <a href="bowtie.html">Bowtie for single-end reads </a> </li>
<li> <a href="bowtie-paired-end.html">Bowtie for paired-end reads</a> </li>
<li> <a href="bowtie-with-index-building.html">Bowtie for single-end reads and own genome </a> </li>
<li> <a href="bowtie-paired-end-with-index-building.html">Bowtie for paired-end reads and own genome </a> </li>
<li><a href="bowtie2.html">Bowtie2 for single end reads</a>.</li>
<li><a href="bowtie2-paired-end.html">Bowtie2 for paired end reads</a>.</li>
<li><a href="bowtie2-with-index-building.html">Bowtie2 for single end reads and own genome</a>.</li>
<li><a href="bowtie2-paired-end-with-index-building.html">Bowtie2 for paired end reads and own genome</a>.</li>
<li> <a href="bwa.html">BWA for single-end reads </a> </li>
<li> <a href="bwasw.html">BWA-SW for single-end reads </a> </li>
<li> <a href="bwa-paired-end.html">BWA for paired-end reads </a> </li>
<li> <a href="bwa-with-index-building.html">BWA for single-end reads and own genome</a> </li>
<li> <a href="bwasw-with-index-building.html">BWA-SW for single-end reads and own genome</a> </li>
<li> <a href="bwa-paired-end-with-index-building.html">BWA for paired-end reads and own genome </a> </li>
<li><a href="tophat2.html">TopHat2 for paired end reads</a> </li>
<li><a href="tophat2-single-end.html">TopHat2 for single end reads</a></li>
<li> <a href="tophat.html">TopHat for paired end reads </a> </li>
<li> <a href="tophat-single-end.html">TopHat for single end reads </a> </li>
<br>
<br>
<b><a name="variants">Variants</a></b>
<li><a href="samtools-snp-indel-single.html"> Call SNPs and short INDELs</a></li>
<li><a href="vcftools.html">Filter and analyze variants</a></li>
<li><a href="annotate-variant.html">Annotate variants</a></li>
<br>
<br>
<b><a name="RNA">RNA-seq</a></b>
<li> <a href="htseq-count.html">Map aligned reads to genes with HTSeq</a> </li>
<li> <a href="htseq-count-own-gtf.html">Map aligned reads to genes with HTSeq using own GTF </a> </li>
<!--  <li> Map aligned reads to genes with coverageBed using own BED </li>-->
<li><a href="dea-cufflinks.html">Differential expression analysis using Cufflinks</a></li>
<li> <a href="ngs-dea-edger-RNA.html"> Differential expression analysis using edgeR</a></li>
<li> <a href="edger-multivariate.html"> Differential expression analysis using edgeR for multivariate experiments</a></li>
<li><a href="dea-deseq.html">Differential expression analysis using DESeq</a></li>
<li><a href="dexseq-exoncount.html">RNA-seq / Map aligned reads to exons for DEXSeq</a></li>
<li><a href="DEXSeq.html">RNA-seq / Differential exon expression using DEXSeq</a></li>
<br>
<br>
<b><a name="miRNA">miRNA-seq</a></b>
<li><a href="convert-bam-to-edger.html">Convert genomic BAM to edgeR input format</a></li>
<li><a href="convert-miRBase-bam-to-edger.html">Convert miRBased BAM to edgeR input format</a></li>
<li><a href="ngs-dea-edger.html">Differential expression analysis using edgeR</a></li>
<li><a href="dea-deseq.html">Differential expression analysis using DESeq</a></li>
<li><a href="ngs-annotate-miRNA-targets.html">Find miRNA targets</a></li>
<li><a href="ngs-pathways-mirna-hyperg-go.html">GO enrichment for miRNA targets</a></li>
<li><a href="ngs-pathways-mirna-hyperg-kegg.html">KEGG enrichment for miRNA targets</a></li>
<li><a href="ngs-correlation-analysis-mirna.html">Correlate miRNA with target expression</a></li>
<li><a href="ngs-up-down-analysis-mirna.html">Up-down analysis of miRNA targets</a></li>
<li><a href="ngs-extract-target-gene-expression.html">Extract miRNA target expression data</a></li>
<br>
<br>
<b><a name="ChIP">ChIP-seq and FAIRE-seq</a></b>
<li> <a href="ngs-find-peaks-macs-one.html">Find peaks using MACS, treatment only </a> </li>
<li> <a href="ngs-find-peaks-macs-two.html">Find peaks using MACS, treatment vs. control </a> </li>
<li> <a href="ngs-find-motifs-jaspar.html">Find common motifs and match to JASPAR </a> </li>
<li> <a href="ngs-find-nearest-genes.html">Find the nearest genes for regions </a> </li>
<li> <a href="ngs-find-unique-genes.html">Find unique and annotated genes</a></li>
<li> <a href="pathways-ngs-hyperg-go.html">GO enrichment for list of genes </a> </li>
<li><a href="ngs-find-peaks-fseq.html">ChIP-seq and FAIRE-seq / Find broad peaks using F-seq</a></li>
<br>
<br>
<b><a name="CNA">CNA-seq</a></b>
<li><a href="cna-bin-hits.html">Count reads in fixed bins</a></li>
<li><a href="cna-define-experiment.html ">Define CNA-seq experiment </a> </li>
<li><a href="cna-correct.html">Correct for GC content</a></li>
<li><a href="acgh-normalize.html">Normalize copy number data</a></li>
<li><a href="cna-plot-profile.html">Plot copy number profiles</a></li>
<li><a href="acgh-segment.html">Segment copy number data</a></li>
<li><a href="acgh-call-aberrations.html">Call aberrations from segmented copy number data</a></li>
<li><a href="acgh-identify-regions.html">Identify common regions from called copy number data</a></li>
<li><a href="acgh-plot-frequencies.html">Plot copy number aberration frequencies</a></li>
<li><a href="acgh-cluster.html">Cluster called copy number data</a></li>
<li><a href="acgh-group-test.html">Group tests for called copy number data</a></li>
<li><a href="acgh-survival-test.html">Survival test for called copy number data</a></li>
<li><a href="acgh-plot-survival.html">Plot survival curves for called copy number data</a></li>
<li><a href="acgh-convert-from-probes-to-genes.html">Detect genes from called copy number data</a></li>
<li><a href="acgh-pathways-go.html">GO enrichment for called gene copy numbers</a></li>
<li><a href="acgh-match-probes.html">Match copy number and expression features</a></li>
<li><a href="acgh-plot-combined-profile.html">Plot profiles of matched copy number and expression</a></li>
<li><a href="acgh-expression-test.html">Test for copy number induced expression changes</a></li>
<li><a href="acgh-plot-combined-expression.html">Plot copy number induced gene expression</a></li>
<li><a href="acgh-add-cytobands.html">Add cytogenetic bands</a></li>
<li><a href="acgh-count-overlapping-cnvs.html">Count overlapping CNVs</a></li>
<br>
<br>
<b><a name="methyl">Methyl-seq</a></b>
<li> <a href="convertBAMtoMEDIPS.html">Convert BAM file to MEDIPS input format </a> </li>
<li> <a href="combineMEDIPS.html">Combine MEDIPS input files </a> </li>
<li> <a href="MEDIPS_one_sample.html">MEDIPS -methylation analysis, treatment only </a> </li>
<li> <a href="MEDIPS_two_samples.html">MEDIPS -methylation analysis, treatment and control </a> </li>


      


<!--        <li> <a href=" "> </a> </li>        -->
</ul>

</body>

</html>