<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html lang="en">
<head><title>Chipster manual</title></head>
<body>

<h2>Analysis tools for microarray data</h2>

<p>Every tool is described in detail on its own help page. 
The help page is divided into four parts: General description, Parameters, Details on how the tool works, and Output 
where the results are described. The Parameters part contains the list of parameters, their possible values (1-100)
and their default values [10].</p>

<ol>
<li> <a href="#normalization">Normalization</a> </li>
<li> <a href="#quality">Quality control</a> </li>
<li> <a href="#preprocessing">Preprocessing</a> </li>
<li> <a href="#statistics">Statistics</a> </li>
<li> <a href="#clustering">Clustering</a> </li>
<li> <a href="#annotation">Annotation</a> </li>
<li> <a href="#pathways">Pathways</a> </li>
<li> <a href="#promoter">Promoter analysis</a> </li>
<li> <a href="#acgh">Copy number aberrations</a> </li>
<li> <a href="#visualization">Visualization</a> </li>
<li> <a href="#utilities">Utilities</a> </li>
</ol>
<ul>
<b><a name="normalization">Normalization</a></b>
<li> <a href="norm-affy.html">Affymetrix </a> </li>
<li> <a href="norm-affy-exon.html">Affymetrix exon arrays</a> </li>
<li> <a href="norm-affy-snp.html">Affymetrix SNP arrays</a> </li>
<li> <a href="norm-affy-gene.html">Affymetrix gene arrays</a> </li>
<li> <a href="norm-agilent.html">Agilent 2-color</a> </li>
<li> <a href="norm-agilent-one-color.html">Agilent 1-color </a> </li>
<li> <a href="norm-agilent-mirna.html">Agilent miRNA </a> </li>
<li> <a href="norm-cdna.html">cDNA</a> </li>
<li> <a href="norm-illumina.html">Illumina</a> </li>
<li> <a href="norm-illumina-lumi.html">Illumina - lumi pipeline</a> </li>
<li> <a href="norm-illumina-methylumi.html">Illumina - methylumi pipeline</a> </li>
<li> <a href="norm-lme.html">Random effects </a> </li>
<li> <a href="norm-combat.html">ComBat - batch normalisation</a></li>
<li> <a href="norm-specific-samples.html">Specific samples </a> </li>
<li> <a href="norm-specific-genes.html">Specific genes </a> </li>
<li> <a href="norm-chip-average.html">Chip average </a> </li>
<li> <a href="norm-gene-average.html">Gene average </a> </li>
<li> <a href="norm-prenormalized.html">Process prenormalized</a> </li>
<li> <a href="norm-process-prenormalized-affy.html">Process prenormalized affy</a> </li>
<br>
<br>
<b><a name="quality">Quality control</a></b>
<li> <a href="qc-affy.html">Affymetrix </a> </li>
<li> <a href="qc-affy-rle-nuse.html">Affymetrix RLE/NUSE</a> </li>
<li> <a href="qc-agilent.html">Agilent </a> </li>
<li> <a href="qc-agilent-one-color.html">Agilent 1-color </a> </li>
<li> <a href="qc-cdna.html">cDNA </a> </li>
<li> <a href="qc-illumina.html">Illumina </a> </li>
<br>
<br>
<b><a name="preprocessing">Preprocessing</a></b>
<li> <a href="filter-cv.html">Filter by CV</a> </li>
<li> <a href="filter-expression.html">Filter by expression</a> </li>
<li> <a href="filter-flags.html">Filter by flags</a> </li>
<li> <a href="filter-sd.html">Filter by standard deviation</a> </li>
<li> <a href="filter-iqr.html">Filter by interquartile range</a> </li>
<li> <a href="filter-by-column-term.html">Filter by column term</a> </li>
<li> <a href="filter-by-column-value.html">Filter by column value</a> </li>
<li> <a href="impute.html">Impute missing values</a> </li>
<li> <a href="na-omitted.html">Remove missing values </a> </li>
<br>
<br>
<b><a name="statistics">Statistics</a></b>
<li> <a href="stat-one-group.html">One sample tests</a> </li>
<li> <a href="ordination-nmds.html">NMDS </a> </li>
<li> <a href="ordination-pca.html">PCA </a> </li>
<li> <a href="ordination-ca.html">DCA </a> </li>
<li> <a href="stat-several-groups.html">Several groups tests</a> </li>
<li> <a href="stat-singleslide.html">Single-slide methods</a> </li>
<li> <a href="stat-timeseries.html">Time series</a> </li>
<li> <a href="stat-two-groups.html">Two groups tests</a> </li>
<li> <a href="stat-ROTS.html">ROTS</a> </li>
<li> <a href="stat-sam.html">SAM </a> </li>
<li> <a href="up-down-analysis-mirna.html">UP-DOWN analysis of miRNA targets </a> </li>
<li> <a href="correlation-analysis-mirna.html">Correlate miRNA with target expression </a> </li>
<li> <a href="stat-correlate-with-phenodata.html">Correlate with phenodata</a> </li>
<li> <a href="estimate-samplesize.html">Sample size estimation</a> </li>
<li> <a href="sample-size-with-bh.html">Sample size calculations with an adapted BH method</a></li>
<li> <a href="stat-linear-modelling.html">Linear modelling</a> </li>
<li> <a href="stat-linear-modelling-designmat.html">Linear modelling using user-defined design matrix</a></li>
<li> <a href="stat-chisq-snp.html">Association analysis </a></li>
<li> <a href="stat-adjust-pvalues.html">Adjust p-values </a></li>
<li> <a href="calculate-descriptive-statistics.html">Calculate descriptive statistics </a> </li>
<br>
<br>
<b><a name="clustering">Clustering</a></b>
<li> <a href="cluster-hierarchical.html">Hierarchical</a> </li>
<li> <a href="cluster-kmeans.html">K-means</a> </li>
<li> <a href="cluster-kmeans-testK.html">K-means - estimate K</a> </li>
<li> <a href="classification-knn.html">KNN classification</a> </li>
<li> <a href="cluster-som.html">SOM</a> </li>
<li> <a href="cluster-qt.html">QT</a> </li>
<li> <a href="classification.html">Classification</a> </li>
<br>
<br>
<b><a name="annotation">Annotation</a></b>
<li> <a href="annotate-genelist2html.html">Gene list</a> </li>
<li> <a href="annotate-add-to-data.html">Add annotations to data</a> </li>
<li> <a href="annotate-miRNA.html">Annotate miRNA</a> </li>
<li> <a href="annotate-miRNA-targets.html">Find miRNA targets</a> </li>
<br>
<br>
<b><a name="pathways">Pathways</a></b>
<li> <a href="stat-hyperG-GO.html">Hypergeometric test for GO</a> </li>
<li> <a href="stat-hyperG-KEGG-PFAM.html">Hypergeometric test for KEGG / PFAM</a> </li>
<li> <a href="pathways-hypergeometric-cpdb.html">Hypergeometric test for ConsensusPathDB</a> </li>
<li> <a href="stat-hyperG-cytoband.html">Hypergeometric test for cytobands</a> </li>
<li> <a href="pathways-reactome.html">Associations to Reactome pathways</a> </li>
<li> <a href="pathways-intact.html">Protein interactions from IntAct</a> </li>
<li> <a href="stat-hyperG-safe.html">SAFE</a> </li>
<li> <a href="stat-geneset.html">Gene set test</a> </li>
<li> <a href="pathways-hyperg-mirna-kegg.html">KEGG enrichment for miRNA targets</a> </li>
<li> <a href="pathways-mirna-hyperg-go.html">GO enrichment for miRNA targets</a> </li>
<br>
<br>
<b><a name="promoter">Promoter analysis</a></b>
<li> <a href="promoter-tfbs.html">Weeder</a> </li>
<li> <a href="promoter-tfbs-cosmo.html">Cosmo</a> </li>
<li> <a href="promoter-cbust.html">ClusterBuster</a> </li>
<li> <a href="promoter-retrprom.html">Retrieve promoters</a> </li>
<br>
<br>
<b><a name="acgh">Copy number aberrations</a></b>
<li> <a href="acgh-import-from-cangem.html">Import from CanGEM</a> </li>
<li><a href="acgh-normalize.html">Normalize copy number data</a></li>
<li> <a href="acgh-plot-profile.html">Plot copy number profiles</a> </li>
<li> <a href="acgh-smooth.html">Smooth waves from normalized aCGH data</a> </li>
<li><a href="acgh-segment.html">Segment copy number data</a></li>
<li><a href="acgh-call-aberrations.html">Call aberrations from segmented copy number data</a></li>
<li><a href="acgh-identify-regions.html">Identify common regions from called copy number data</a></li>
<li><a href="acgh-plot-frequencies.html">Plot copy number aberration frequencies</a></li>
<li><a href="acgh-cluster.html">Cluster called copy number data</a></li>
<li><a href="acgh-group-test.html">Group tests for called copy number data</a></li>
<li><a href="acgh-survival-test.html">Survival test for called copy number data</a></li>
<li><a href="acgh-plot-survival.html">Plot survival curves for called copy number data</a></li>
<li><a href="acgh-convert-from-probes-to-genes.html">Detect genes from called copy number data</a></li>
<li><a href="acgh-pathways-go.html">GO enrichment for called gene copy numbers</a></li>
<li><a href="acgh-match-probes.html">Match copy number and expression features</a></li>
<li><a href="acgh-plot-combined-profile.html">Plot profiles of matched copy number and expression</a></li>
<li><a href="acgh-expression-test.html">Test for copy number induced gene expression</a></li>
<li><a href="acgh-plot-combined-expression.html">Plot copy number induced gene expression</a></li>
<li> <a href="acgh-fetch-probe-positions-from-cangem.html">Fetch probe positions from CanGEM</a> </li>
<li><a href="acgh-fetch-probe-positions-from-geo.html">Fetch probe positions from GEO</a></li>
<li><a href="acgh-add-cytobands.html">Add cytogenetic bands</a></li>
<li><a href="acgh-count-overlapping-cnvs.html">Count overlapping CNVs</a></li>
<br>
<br>
<b><a name="visualization">Visualization</a></b>
<li> <a href="plot-chrom-pos.html">Chromosomal position </a> </li>
<li> <a href="plot-idiogram.html">Idiogram</a> </li>
<li> <a href="plot-volcano.html">Volcano plot</a> </li>
<li> <a href="plot-volcano-data-exists.html">Volcano plot from existing results</a> </li>
<li> <a href="plot-boxplot.html">Boxplot</a> </li>
<li> <a href="plot-corrgram.html">Correlogram</a> </li>
<li> <a href="plot-dendrogram.html">Dendrogram</a> </li>
<li> <a href="plot-heatmap.html">Heatmap</a> </li>
<li> <a href="plot-histogram.html">Histogram</a> </li>
<br>
<br>
<b><a name="Utilities">Utilities</a></b>
<li> <a href="merge-tables.html">Merge tables </a> </li>
<li> <a href="merge-datasets.html">Merge data sets</a></li>
<li> <a href="average-replicates.html">Average replicates </a> </li>
<li> <a href="calculate-fold-change.html">Calculate fold change </a> </li>
<li> <a href="extract-genes.html">Extract genes </a> </li>
<li> <a href="extract-target-gene-expression.html">Extract data for miRNA targets </a> </li>
<li> <a href="extract-genes-from-clustering.html">Extract genes from clustering </a> </li>
<li> <a href="extract-genes-from-go.html">Extract genes from GO term</a></li>
<li> <a href="extract-genes-from-stattest.html">Extract genes from a statistical test results </a> </li>
<li> <a href="extract-samples-from-dataset.html">Extract samples from dataset </a> </li>
<li> <a href="generate-phenodata.html">Generate phenodata </a> </li>
<li> <a href="search-correlation.html">Search by correlation</a> </li>
<li> <a href="search-queryword.html">Search by gene name </a> </li>
<li> <a href="filter-by-column.html">Filter by column </a> </li>
<li> <a href="delete-columns.html">Delete columns </a> </li>
<li> <a href="delete-and-subtract-columns.html">Delete and subtract columns</a></li>
<li> <a href="change-interpretation.html">Change interpretation </a> </li>
<li> <a href="sort-samples.html">Sort samples </a> </li>
<li> <a href="sort-genes.html">Sort genes </a> </li>
<li> <a href="sampling.html">Random Sampling </a> </li>
<li> <a href="combine-probes-to-genes.html">Combine probes to genes </a> </li>
<li> <a href="export-tab2mage.html">Export tab2mage format</a> </li>
<li> <a href="export-soft.html">Export GEO's SOFT format</a> </li>
<li> <a href="import-soft2.html">Import from GEO</a> </li>
<li> <a href="import-ArrayExpress.html">Import from ArrayExpress</a> </li>
<li> <a href="intersect-lists.html">Intersect lists</a> </li>

  
<!--        <li> <a href=" "> </a> </li>        -->
</ul>

</body>

</html>