<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html lang="en">
<head><title></title></head>
<body>

<h2>Alignment / TopHat2 TopHat2 for paired end reads and own genome</h2>

<h3>Description</h3>

<p>
Aligns paired end Illumina RNA-seq reads to a genome provided as a FASTA file in order to identify exon-exon splice junctions. 
</p>

<h3>Parameters</h3> 

<ul>
<li>When GTF file is used, ignore novel junctions (yes, no) [yes]</li>
<li>Base quality encoding used (phred+33, phred+64) [phred+33]</li>
<li>Expected inner distance between mate pairs (Integer) [200]</li>
<li>Standard deviation for the inner distances between mate pairs (Integer) [20]</li>
<li>How many hits is a read allowed to have (1-1000000) [20]</li>
<li>Number of mismatches allowed in final alignment [2]</li>
<li>Minimum anchor length (3-1000) [8]</li>
<li>Maximum number of mismatches allowed in the anchor (0, 1, 2) [0]</li>
<li>Minimum intron length (10-1000) [70]</li>
<li>Maximum intron length (1000-1000000) [500000]</li>
<li>Report only paired alignments (yes, no) [yes]</li>
<li>Library type (fr-unstranded, fr-firststrand, fr-secondstrand) [fr-unstranded]</li>
</ul>

<h3>Details</h3>

<p>TopHat2 maps Illumina RNA-Seq reads to a genome in order to identify exon-exon splice junctions. The alignment process consists of several steps. 
If annotation is available as a GTF file, TopHat will extract the transcript sequences and use Bowtie2 to align reads to this virtual transcriptome first. 
Only the reads that do not fully map to the transcriptome will then be mapped on the genome. 
The reads that still remain unmapped are split into shorter segments, which are then aligned to the genome. Segment mappings are used to find potential splice sites. 
Sequences flanking a splice site are concatenated, and unmapped segments are mapped to them. Segment alignments are then stitched together to form whole read alignments. 
</p>
<p>
The "anchor length" means that TopHat2 will report junctions spanned by reads with at least this many bases on each side of the junction. 
Note that individual spliced alignments may span a junction with fewer than this many bases on one side. However, every junction involved 
in spliced alignments is supported by at least one read with this many bases on each side. By default no mismatches are allowed in the anchor, but you can change this. 
</p>
<p>
TopHat2 will ignore donor-acceptor pairs which are closer than the minimum intron length or further than the maximum intron length apart.  
</p>
<p>
For paired-end reads, TopHat2 processes the two reads separately through the same
mapping stages described above. In the final stage, the independently aligned reads are
analyzed together to produce paired alignments, taking into consideration additional
factors including fragment length and orientation.
The expected mean inner distance between mate pairs means the fragment length minus the reads. For example, if your fragment size is 300 bp and read length is 50 bp, the inner distance is 200.
</p>

<p>After running TopHat2, Chipster indexes the BAM file using the <a href="http://samtools.sourceforge.net/">SAMtools</a> package. This way the results are ready to be visualized in the genome browser.</p>


<h3>Output</h3>
<p>
This tool returns the following files:
<ul>
<li>tophat.bam: BAM file containingthe alignments.</li>
<li>tophat.bam.bai: Index file for the BAM file.</li>
<li>tophat-summary.txt: Summary of the alignment results.</li>
<li>junctions.bed: Each junction consists of two connected BED blocks, where each block is as long as the maximal overhang of any read spanning the junction. 
The score is the number of alignments spanning the junction.</li>
</ul>
In cases where Tophat2 for some reason fails, file tophat2.log is returned instead.
</p> 

<h3>Reference</h3>

<p>This tool is based on the <a href="http://tophat.cbcb.umd.edu/">TopHat</a> package. Please cite the following article:</p>
Kim D, Petrtea G, Trapnell C, et al. <a href="http://genomebiology.com/2013/14/4/R36/abstract">TopHat2: accurate alignments of transcriptomes in the presence of insertions, deletions and gene fusions.</a>  Genome Biology (2013) 14: R36. 
</p>
 


</body>