<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html lang="en">
<head><title></title></head>
<body>

<h2>Preprocessing / Trim reads with Trimmomatic</h2>

<h3>Description</h3>

<p>
This tool performs a variety of trimming tasks for Illumina paired end and single end data.
</p>

<h3>Parameters</h3>

<ul>
<li>Adapter set (none, TruSeq2-SE.fa, TruSeq3-SE.fa, TruSeq2-PE.fa, TruSeq3-PE.fa, TruSeq3-PE-2.fa, NexteraPE-PE.fa) [none]</li>
<li>Adapter clipping parameters (String) []</li>
<li>Quality scale used in the fastq file (phred33, phred64) [phred33]
<li>Minumum quality to keep a leading base (Integer) []</li>
<li>Minumum quality to keep a trailing base (Integer) []</li>
<li>Number of bases to keep from the start (Integer) []</li>
<li>Number of bases to remove from the start (Integer) []</li>
<li>Sliding window trimming parameters (String) []</li>
<li>Adaptive quality trimming parameters (String) []</li>
<li>Minimum average quality of reads to keep (Integer) []</li>
<li>Minimum length of reads to keep (Integer) []</li>
<li>Write a log file (yes, no) [yes]</li>
</ul>

<h3>Details</h3>

<p>
Only the trimming steps with user specified parameters are performed. The steps with empty parameters are skipped. The steps are performed in following order: 
<ol>
<li>Adapter clipping (ILLUMINACLIP)</li>
<li>Trim leading bases by quality (LEADING)</li>
<li>Trim trailing bases by quality (TRAILING)</li>
<li>Number of bases to keep from the start (CROP)</li>
<li>Number of bases to remove from the start (HEADCROP)</li>
<li>Sliding window trimming (SLIDINGWINDOW)</li>
<li>Adaptive quality trimming (MAXINFO)</li>
<li>Minimum average quality of reads to keep (AVGQUAL)</li>
<li>Minimum length of reads to keep (MINLEN)</li>
</ol>
</p>
<p>
Adapter clipping is best performed first, as other clipping functions may remove parts of adapter sequence and thus make adapters more 
difficult to find. Minimum length filtering should be done last, or additional clipping may produce reads shorter than the minimum 
length specified. Other steps are less sensitive to the order they are performed, but if you wish to run the steps in a specific order 
that differs from the default order, you have to run each step separately.
</p>
<p>
For details on each trimming step and their parameters see the <a href = "http://www.usadellab.org/cms/uploads/supplementary/Trimmomatic/TrimmomaticManual_V0.32.pdf">Trimmomatic manual</A>.
</p>
<p>
The tool is based on the Trimmomatic package.
</p>


<h3>Output</h3>
<p>
The trimmed reads are in gzipped fastq format. Output files depend on whether the trimmimng was done on single end or paired end reads.
</p>
<p>
For single end reads:
<ul>
<li>trimmed.fq.gz:	Trimmed reads</li>
</ul>
For paired end reads:
<ul>
<li>trimmed_reads1_paired.fq.gz:	Trimmed reads from the first input file with a mate pair in the second input file.</li>
<li>trimmed_reads1_unpaired.fq.gz:	Trimmed reads from the first input file without a mate pair in the second input file.</li>
<li>trimmed_reads2_paired.fq.gz:	Trimmed reads from the second input file with a mate pair in the first input file.</li>
<li>trimmed_reads2_unpaired.fq.gz:	Trimmed reads from the second input file without a mate pair in the first input</li>
</ul>
Optionally:
<ul>
<li>trimlog.txt:	Trimming log file</li>
</ul>

</p>


<h3>References</h3>

<p>This tool uses the the Trimmomatic package. Please cite the article:</p>
<p>Bolger, A. M., Lohse, M., & Usadel, B. (2014). Trimmomatic: A flexible trimmer for Illumina Sequence Data. Bioinformatics, btu170.</p>

<p>Please see the <a href = "http://www.usadellab.org/cms/?page=trimmomatic">Trimmomatic homepage</A> for more details.</p>
</body>

