<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html lang="en">
<head><title></title></head>
<body>

<h2>Utilities / Retrieve unique alignments from BAM</h2>

<h3>Description</h3>

<p>
Given a BAM file, this tool retrieves unique alignments.
</p>

<h3>Parameters</h3>

Does the alignment file contain paired-end data (yes, no) [no]

<h3>Details</h3>

<p>
This tool filters out unaligned reads and alignments which have a mapping quality less than 4. For paired end data it keeps only reads which are mapped in a proper pair.
</p>


<h3>Output</h3>

<p>
Output is a BAM file where each read has the tag NH:i:1 (this tag is used by tools which count RNA-seq reads per genes/exons).
</p>


<h3>References</h3>

<p>This tool is based on the <a href = "http://samtools.sourceforge.net/">SAMtools package</a>. 
Please cite the article <a href = "http://www.ncbi.nlm.nih.gov/pubmed/19505943">The Sequence alignment/map (SAM) format and SAMtools</a> by Li H., Handsaker B., Wysoker A., Fennell T., Ruan J., Homer N., Marth G., Abecasis G., Durbin R. and 1000 Genome Project Data Processing Subgroup (2009) Bioinformatics, 25, 2078-9. [PMID: 19505943]. 
</p>

</body>

