<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html lang="en">
<head><title></title></head>
<body>

<h2>Statistics / Up-down analysis of miRNA targets</h2>

<h3>Description</h3>

<p>Given miRNA expression data and gene expression data from the same two-group comparison experiment,
 this tool identifies the target genes which behave oppositely to the expression of the corresponding miRNA.</p>

<h3>Parameters</h3>

<ul>
<li>average.method (mean, median) [median] </li>
<li>group.column.mirna [group]</li>
<li>group.column.gene [group]</li>
<li>id.type (probe_id, entrez_id) [probe_id]</li>

</ul>

<h3>Details</h3>

<p>This tool integrates miRNA and gene expression data from two-group comparison experiments, and identifies which of the target genes behave oppositely to any given miRNA.
By default, the tool assumes that the "group" column of the phenodata table is used to define which samples belong to a certain experiment groups. NOTE: you need to select both the file with the miRNA expression and the file with the matching gene expression.<br>

<br>The analysis consists of two parts:<br>

<br>Firstly, predicted target genes for the list of miRNAs are fetched from the TargetScan and PicTar databases. An intersection
is made to include only those genes that are predicted in both databases. Then, another intersection is made with the list of genes
in the matching gene expression data set.<br>

<br>Next, the average expression (either mean or median) is calculated for the two experiment groups for each data set and the fold difference calculated.
For each up-regulated miRNA the corresponding target genes that are down-regulated are identified and a result table listing the miRNAs, target genes and the fold
difference for each is reported. Down-regulated miRNAs and their up-regulated gene targets are reported in a separate result file.
NOTE; since there is no statistical analysis performed and hence no confidence measures are reported, it is advisable that the input data has been judiciously filtered
and/or subjected to statistical analysis for differential expression prior to invoking this tool.<br>

 
</p>

<h3>Output</h3>

<p>The output consists of two text files, one listing the up-regulated miRNAs and their down-regulated gene targets, and one listing the down-regulated miRNAs
and their up-regulated gene targets.</p>

</p>

<h3>References</h3>
<p>This tool leverages some of the functionality from the RmiR Bioconductor package. More info on the databases of predicted miRNA targets can be found at:<br/>
<br/>TargetScan<br/>
http://www.targetscan.org/
<br/><br> PicTar<br/>http://pictar.mdc-berlin.de/</p>

</body>

