<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html lang="en">
<head><title></title></head>
<body>

<h2>Filter variants</h2>

<h3>Description</h3>

<p>Filters variants in VCF files.</p>

<h3>Parameters</h3>

<ul>
<li>Keep only indels (yes, no) [no]</li>
<li>Remove indels (yes, no) [no]</li>
<li>Minimum number of alleles (Integer) [0]</li>
<li>Maximum number of alleles (Integer) [0]</li>
<li>Minimum quality (Decimal) [0]</li>
<li>Filter by flags (yes, no) [no]</li>
<li>List removed files (yes, no) [no]</li>
</ul>

<h3>Details</h3>

<p> 
Filters VCF files by various criteria.
</p>


<h3>Output</h3>

<p>
Output is a VCF file. You can view VCF files in Chipster as text (which includes the header) or spreadsheet (which can be sorted). 
The location column in the spreadsheet can be used in Chipster genome browser to navigate quickly through a list of variants.
</p>


<h3>References</h3>

<p>This tool is based on the <a href = "http://vcftools.sourceforge.net/">VCFtools package</a>. 
Please cite the article <a href = "http://bioinformatics.oxfordjournals.org/content/27/15/2156">The Variant Call Format and VCFtools</a> by Danecek P., Auton A., Abecasis G, Albers C. A., Banks E., DePristo M. A., Handsaker R., Lunter G., Gabor Marth G., Sherry S. T., McVean G., Durbin R. and 1000 Genomes Project Analysis Group (2011) Bioinformatics, 27, 2156-8. [PMID: 21653522] 

</p>

</body>

