<!DOCTYPE html>
<html>

  <head>
    <meta charset='utf-8' />
    <meta http-equiv="X-UA-Compatible" content="chrome=1" />
    <meta name="description" content="Chipster : Open source platform for data analysis" />

    <link rel="stylesheet" type="text/css" media="screen" href="stylesheets/stylesheet.css">

    <title>Chipster</title>
  </head>

  <body>

    <!-- HEADER -->
    <div id="header_wrap" class="outer">
        <header class="inner">
          <a id="forkme_banner" href="https://github.com/chipster/chipster">View on GitHub</a>

          <h1 id="project_title">Chipster</h1>
          <h2 id="project_tagline">Open source platform for data analysis</h2>

	<!--
            <section id="downloads">
              <a class="zip_download_link" href="https://github.com/chipster/chipster/zipball/master">Download this project as a .zip file</a>
              <a class="tar_download_link" href="https://github.com/chipster/chipster/tarball/master">Download this project as a tar.gz file</a>
            </section>
	-->
        </header>
    </div>

    <!-- MAIN CONTENT -->
    <div id="main_content_wrap" class="outer">
      <section id="main_content" class="inner">


<h2><a name="release-notes" class="anchor"></a>Release notes</h2>

<h3 id="2.12.1">Version 2.12.1 <em>(21/May/2014)</em></h3>

<p>See <a href="http://chipster.csc.fi/manual/newVersion.html">http://chipster.csc.fi/manual/newVersion.html</a>.</p></p>  


<h3 id="2.12.0">Version 2.12.0 <em>(13/May/2014)</em></h3>

<p>See <a href="http://chipster.csc.fi/manual/newVersion.html">http://chipster.csc.fi/manual/newVersion.html</a>.</p></p>  


<h3 id="2.11.0">Version 2.11.0 <em>(7/Apr/2014)</em></h3>

<p>See <a href="http://chipster.csc.fi/manual/newVersion.html">http://chipster.csc.fi/manual/newVersion.html</a>.</p></p>  


<h3 id="2.10.0">Version 2.10.0 <em>(31/Jan/2014)</em></h3>

<p>Maintenance release. See <a href="http://chipster.csc.fi/manual/newVersion.html">http://chipster.csc.fi/manual/newVersion.html</a>.</p></p>  


<h3 id="2.9.0">Version 2.9.0 <em>(4/Dec/2013)</em></h3>

<p>See <a href="http://chipster.csc.fi/manual/newVersion.html">http://chipster.csc.fi/manual/newVersion.html</a>.</p></p>  


<h3 id="2.8.1">Version 2.8.1 <em>(16/Sep/2013)</em></h3>

<p></p>  

<h4>Improvements</h4>
<ul>
<li>Tophat2 updated to tophat-2.0.9</li>
</ul>

<h4>Bug fixes</h4>
<ul>
<li>miRNA names forced to lower case in microarray tool "Go enrichment for miRNA targets"</li>
</ul>


<h3 id="2.8.0">Version 2.8.0 <em>(30/Aug/2013)</em></h3>

<p>See <a href="http://chipster.csc.fi/manual/newVersion.html">http://chipster.csc.fi/manual/newVersion.html</a>.</p></p>


<h3 id="2.7.0">Version 2.7.0 <em>(28/Jun/2013)</em></h3>

<p>See <a href="http://chipster.csc.fi/manual/newVersion.html">http://chipster.csc.fi/manual/newVersion.html</a>.</p></p>  


<h3 id="2.6.1">Version 2.6.1 <em>(20/Jun/2013)</em></h3>

<h4>Improvements</h4>
<ul>
<li>Cufflinks update to 2.1.1</li>
</ul>

<h4>Bug fixes</h4>
<ul>
<li>Various fixes and updates to genomes, bwa, bowtie and bowtie2 indexes</li>
<li>HTSeq link fix</li>
<li>Fix Normalisation / Affymetrix gene arrays 
</ul>



<h3 id="2.6.0">Version 2.6.0 <em>(30/May/2013)</em></h3>

<p>See <a href="http://chipster.csc.fi/manual/newVersion.html">http://chipster.csc.fi/manual/newVersion.html</a>.</p></p>  


<h3 id="2.5.1">Version 2.5.1 <em>(22/Apr/2013)</em></h3>

<p></p>  

<h4>New features</h4>
<ul>
<li>Copy number data visualisation in Genome browser</li>
</ul>

<h4>Bug fixes</h4>
<ul>
<li>Add missing miRBase19 bowtie and bowtie2 indexes</li>
</ul>



<h3 id="2.5.0">Version 2.5.0 <em>(11/Apr/2013)</em></h3>

<p>See <a href="http://chipster.csc.fi/manual/newVersion.html">http://chipster.csc.fi/manual/newVersion.html</a>.</p></p>  



<h3 id="2.4.0">Version 2.4.0 <em>(8/Feb/2013)</em></h3>

<p></p>  

<h4>Improvements</h4>
<ul>
<li>Scrollable samples in the genome browser</li>
<li>Chicken, cow and human mitochondrial genome added to genome browser</li>
<li>Variants / Call SNPs and short INDELs: reference genomes for pig, chicken and cow have been added</li>

</ul>

<h4>New features</h4>
<ul>
<li>Normalization / Agilent miRNA too restored</li>
</ul>


<h3 id="2.3.2">Version 2.3.2 <em>(24/Jan/2013)</em></h3>

<p></p>  

<h4>Improvements</h4>
<ul>
<li>Updated Agilent 1-color normalisation</li>
<li>Updated Two groups test</li>
<li>Arabidopsis lyrata and Sus scrofa for bwa and bowtie2</li>
<li>Vcftool, added minimum quality parameter</li>

</ul>

<h4>New features</h4>
<ul>
<li>Cufflinks2</li>
</ul>

<h4>Bug fixes</h4>
<ul>
<li>BWA unzipping</li>
</ul>



<h3 id="2.3.1">Version 2.3.1 <em>(22/Nov/2012)</em></h3>

<p></p>  

<h4>Improvements</h4>
<ul>
<li>Annotate variants tool</li>
<li>Manual updates</li>
</ul>

<h4>Bug fixes</h4>
<ul>
<li>BWA tool</li>
</ul>



<h3 id="2.3.0">Version 2.3.0 <em>(12/Nov/2012)</em></h3>

<p></p>  

<h4>Improvements</h4>
<ul>
<li>Yersinia similis to genome browser</li>
</ul>

<h4>New features</h4>
<ul>
<li>Bowtie2 for single end reads and own genome</li>
<li>Bowtie2 for paired end reads and own genome</li>
<li>Normalisation / Illumina - methylumi pipeline</li>
</ul>




<h3 id="2.2.4">Version 2.2.4 <em>(07/Nov/2012)</em></h3>

<p></p>  

<h4>Improvements</h4>
<ul>
<li>Sheep and mm10 genomes to bwa</li>
<li>Fruit fly and hgug4851a R annotation packages</li>
</ul>

<h4>New features</h4>
<ul>
<li>Bowtie2</li>
<li>One tool can now exist in several categories and modules</li>
</ul>

<h4>Bug fixes</h4>
<ul>
<li>Bwa</li>
<li>Bowtie paired end tools</li>
<li>FASTX clipper</li>
<li>Visualisation / Heatmap</li>
<li>Stat / Timeseries</li>
</ul>


<h3 id="2.2.3">Version 2.2.3 <em>(26/Oct/2012)</em></h3>

<p></p>  

<h4>Improvements</h4>
<ul>
<li>Use file extension as the label in the workflow view</li>
</ul>

<h4>Bug fixes</h4>
<ul>
<li>Remove obsolete genomes</li>
<li>Remove Filter by column tool</li>
</ul>



<h3 id="2.2.2">Version 2.2.2 <em>(22/Oct/2012)</em></h3>

<p>Sheep genome, minor improvements, bug fixes.</p>  

<h4>Improvements</h4>
<ul>
<li>View gtf files in the spreadsheet view</li>
<li>Minor tool improvements:
	<ul>
	<li>Import from GEO</li>
	<li>Bedtools</li>
	<li>Time series</li>
	<li>Heatmap</li>
	</ul>
</li>
</ul>

<h4>New features</h4>
<ul>
<li>Sheep genome to genome browser, Tophat, Bowtie</li>
</ul>

<h4>Bug fixes</h4>
<ul>
<li>Fix for failing to save over existing session in Windows</li>
</ul>



<h3 id="2.2.1">Version 2.2.1 <em>(11/Oct/2012)</em></h3>

<p>Minor technical updates.</p>  

<h4>Improvements</h4>
<ul>
<li>Removed timeouts for tasks.</li>
<li>Improved logging in filebroker.</li>
<li>Minor updates to TopHat tools.</li>

</ul>


<h3 id="2.2.0">Version 2.2.0 <em>(27/Sep/2012)</em></h3>

<p>Chipster 2.2.0 contains a new tool category Variants with tools for analyzing VCF files. Visualization of these files is supported already (see the 12.9.2012 release notes).
There are also new tools for RNA-seq analysis, and many tools have been updated to use Ensembl v68 annotations. 
The full list of NGS analysis tools is available <a href="tools-ngs.html">here</a>.</p>  


<h4>New features</h4>
<ul>
<li><i>RNA-seq / Map aligned reads to exons for DEXSeq</i>. This tool counts the reads that fall into each non-overlapping exonic part.</li>
<li><i>RNA-seq / Differential exon expression using DEXSeq</i>. This Bioconductor-based tool infers differential exon usage.</li>
<li><i>Variants / Call SNPs and short INDELs</i>. This SAMtools and bcftools -based tool combines the old variant calling tools, allowing you to call variants from one or multiple individuals.</li>
<li><i>Variants / Filter and analyze variants</i>. This VCFtools-based tool allows you to filter and analyze variants in a VCF file.</li>
<li><i>Alignment / TopHat2 for paired end reads</i>. This tools uses TopHat 2.0.4. and Bowtie 2. We will TopHat 1.3.0 is still available for a while, but we encourage you to use TopHat 2 for new analysis sessions. </li>
<li><i>Alignment / TopHat2 for single end reads</i>. As above.</li>
</ul>

<h4>Improvements</h4>
<ul>
<li><i>RNA-seq / Differential expression with Cufflinks</i>. Added mouse genome mm10, updated GTFs to Ensembl 68. Note that Cufflinks version is still 1.0.3, but this will be updated to 2.0.2 in October.</li>
<li><i>RNA-seq / Map aligned reads to genes using HTSeq</i>. Added mouse genome mm10, updated GTFs to Ensembl 68.</li>
<li><i>Alignment / TopHat for paired end reads and single end reads</i>. Updated GTFs to Ensembl 68.</li>
</ul>

<h3 id="2.1.0">Version 2.1.0 <em>(12/Sep/2012)</em></h3>

<h4>New features</h4>
<ul>
<li>Major improvements in the genome browser, including several new genomes, support for VCF files and links to Ensembl and UCSC genome browser</li>
<li>New tool: RNA-seq / Differential expression analysis using edgeR for multivariate experiments</li>
<li>New tool: Utilities / Annotate variants</li>
</ul>

<h4>Improvements</h4>
<ul>
<li>Genome browser: improved visualization of INDELs and coverage</li>
<li>Updated to use PRINSEQ version 0.19.3</li>
<li>Support for HiSeq FASTQ format</li>
</ul>

<h3 id="2.0.3">Version 2.0.3 <em>(23/Aug/2012)</em></h3>

<h4>New features</h4>
<ul>
<li>Support for mouse genome mm10</li>
</ul>

<h4>Improvements</h4>
<ul>
<li>Small improvements to NGS tools</li>
</ul>


<h3 id="2.0.2">Version 2.0.2 <em>(6/Jul/2012)</em></h3>

<h4>New features</h4>
<ul>
<li>Various NGS analysis tools for SNP calling, read quality statistics and read filtering (PRINSEQ)</li>
<li>Microarray tool Annotation / Add genomic location to data</li>
<li>Automatic update functionality for virtual machine based server installations</li>
</ul>

<h4>Improvements</h4>
<ul>
<li>Unzipping support to all aligners</li>
<li>Bowtie: Genomes for Dog (UCSC canFam2) and Gasterosteus aculeatus (BROADS1.67) have been added</li>
<li>TopHat: Parameter for the standard deviation of inner distance has been added</li>
</ul>



<h3 id="2.0.1">Version 2.0.1 <em>(17/Apr/2012)</em></h3>

<p><em>A maintenance release that was developed while virtual machine based distribution bundles were created for Chipster.</em></p>

<h4>New features</h4>
<ul>
<li>Support for optional input files ("INPUT OPTIONAL")</li>
<li>Compute service can access filebroker file repository locally (entry "local-filebroker-user-data-path" in config)</li>
</ul>

<h4>Improvements</h4>
<ul>
<li>Automatic configuration of the server installation ("configure.sh auto")</li>
<li>Other small tweaks to make virtual machine based installations more straightforward</li>
</ul>

<h4>Bug fixes</h4>
<ul>
<li>Fixed issue with compute service failing to clean old work files if work dir is symlinked</li>
</ul>


<h3 id="2.0.0">Version 2.0.0 <em>(24/Jan/2012)</em></h3>

<p><em>Chipster 2.0 means a shift in focus, as the software adds support for Next-Generation Sequencing (NGS) data analysis. It supports
ChIP-seq, miRNA-seq, RNA-seq and MeDIP-seq analysis. Besides that, an integrated genome viewer allows the 
user to efficiently browse large NGS datasets.</em></p>

<p><em>New major version also means that some of the underlying mechanisms were updated and hence this version is not fully backwards compatible
with 1.x versions. These changes include improved session system, server module system and improved tool description language (SADL).</em></p>

<h4>New features</h4>
<ul>
<li>Rich selection of NGS data analysis tools</li>
<ul>
<li>Support for ChIP-seq, miRNA-seq, RNA-seq and MeDIP-seq analysis</li>
<li>NGS quality control tools</li>
<li>FastX</li>
<li>FastQC</li>
<li>Bowtie</li>
<li>BWA</li>
<li>MACS</li>
<li>edgeR</li>
<li>HTSeq
<li>coverageBed</li>
<li>Cufflinks</li>
<li>MEDIPS</li>
<li>Various tools for matching data against relevant databases</li>
<li>Region manipulation with BEDTools and custom integrated tools</li>
<li>Various utitilies for data format conversions</li> 
</ul>
<li>Import tool for automatical indexing and sorting of BAM/SAM files</li> 
<li>Integrated genome browser</li>
<ul>
<li>Shows NGS reads in their genomic context using Ensembl annotations</li>
<li>Smooth realtime navigation from chromosome to nucleotide level</li>
<li>Supported track types: gene, transcript, cytoband, reads, SNP's, read coverage, strand specific coverage, quality coverage, density, regions (BED)</li>
<li>Support for SAM/BAM, TSV, BED and Wiggle files</li>
<li>Support for spliced reads</li>
<li>Search by gene name or location</li>
<li>Integrated into BED tables so that genomic locations can be opened with a single click</li> 
</ul>
<li>Improved session system</li>
<ul>
<li>New file extension: .zip</li>
<li>Sessions are automatically saved on the background</li>
<li>If computer crashes, session can be restored when Chipster is restarted and no work is lost</li>
</ul>  
<li>Improved integrated PDF viewer</li>
<li>Server module system</li>
<ul>
<li>Modules for different areas can be managed independently</li>
<li>All tool scripts can be inspected and updated easily</li>
</ul>
<li>New and improved SADL tool description language</li>
<li>Continuous testing support</li>
<ul>
<li>Runs Chipster sessions (.zip) directly, allowing easy creation and modification of test sessions</li>
<li>Results are matched against those in the test sessions</li>
<li>Allows continuous quality assurance for tool content and backend systems</li>
</ul>
</ul>

<h4>Improvements</h4>
<ul>
<li>Temporary copies of data are not made any more at the client side</li>
<li>When compute service and file broker have a shared disk area, data is not copied over but used directly</li>
</ul>  
</ul>


<h3 id="1.4.7">Version 1.4.7 <em>(23/May/2011)</em></h3>

<h4>New features</h4>
<ul>
<li>Tools and annotation packages based on R / Bioconductor functionality have been updated to R version 2.12.1</li>
<li>Mac OS X support for server side also</li>
<li>Integrated PDF viewer</li>
</ul>

<h4>Improvements</h4>
<ul>
<li>All tools that generate static graphical visualizations have been modified to yield output in PDF format</li>
<li>Now obsolete utility tools dropped to make tool selection more clear</li>
<li>New tool: Utilities / Extract data for miRNA targets</li>
<li>Improvements to analysis tools</li>
<li>Server components unsubscribe from temp topics as soon as possible</li>
</ul>

<h3 id="1.4.6">Version 1.4.6 <em>(7/Dec/2010)</em></h3>

<h4>New features</h4>
<ul>
<li>New analysis tool: Utilities / Extract genes from KEGG pathway</li>
</ul>

<h4>Improvements</h4>
<ul>
<li>Utilities / Sort samples can used also before statistical testing</li>
<li>In hierarchical clustering tools maximum number of genes/samples increased to 20000</li>
<li>ActiveMQ broker binds to all interfaces by default</li>
</ul>

<h4>Bug fixes</h4>
<ul>
<li>Message broker (ActiveMQ) instability solved</li>
</ul> 

<h3 id="1.4.5">Version 1.4.5 <em>(14/Oct/2010)</em></h3>

<h4>New features</h4>
<ul>
<li>New tool: aCGH / Smooth waves from normalized aCGH data</li>
<li>New tool: aCGH / Calculate aberration frequencies from called aCGH data</li>
<li>New tool: aCGH / Cluster called aCGH data</li>
<li>New tool: aCGH / Group tests for called aCGH data</li>
<li>New tool: aCGH / Convert called aCGH data from probes to genes</li>
<li>New tool: aCGH / GO enrichment for copy number aberrations</li>
<li>New tool: aCGH / Match copy number and expression probes</li>
<li>New tool: Utilities / Merge datasets</li>
<li>New tool: Utilities / Extract genes from GO term</li> 
</ul>

<h4>Improvements</h4>
<ul>
<li>Improved Pathways / Hypergeometric test for GO: Now includes parameters for p-value adjustment method, GO category type (BP, MF, CC), minimum size of category, and conditional testing. Results are given both as html and as a table which can be used for further filtering.</li> 
<li>Improved Pathways / GO enrichment for miRNA targets: Now includes parameters for p-value adjustment method, GO category type (BP, MF, CC), minimum size of category, and conditional testing. Results are given both as html and as a table which can be used for further filtering.</li> 
<li>Improved Annotations / Find miRNA targets: Modified the output to allow use of downstream tools, such as Venn Diagram visualization to compare output from different gene target databases and select consensus genes.</li> 
<li>Improved Preprocessing / Filter using a column value: Removed the restrictive range from the cutoff parameter.</li> 
<li>Improved Statistics / Calculate descriptive statistics: Modified the behavior for chips so that it now calculates the statistics for all column.</li> 
<li>Improved Statistics / Gene set test: Modified to exclude single gene gene sets and added group labels in plots and added more columns with information to results table.</li> 
<li>Improved Utilities / Combine probes to genes: Modified to work on input that contains results from the Annotations / Add annotations to data tool.</li> 
<li>Improved Utilities / Average replicate chips: Modified so that it generates a new phenodata file, enabling downstream analysis of the averaged data.</li> 
<li>Improved Utilities / Extract samples from dataset: Modified to include annotations and to exclude samples with no class assigned.</li> 
</ul>

<h3 id="1.4.4">Version 1.4.4 <em>(20/Apr/2010)</em></h3>

<h4>Bug fixes</h4>
<ul>
<li>Fixed the 3D scatter plot bug</li>
<li>Fixes to Average replicate chips, Affy normalisation and Linear modelling tools</li>
</ul>

<h3 id="1.4.3">Version 1.4.3 <em>(13/Apr/2010)</em></h3>

<h4>New features</h4>
<ul>
<li>Annotation / Find miRNA targets</li>
<li>Tools for for aCGH data analysis</li>
<li>Tools for aCGH and gene expression integration</li>
<li>New filtering tools</li>
<li>Provide support for Affy arabidopsis annotations</li>
<li>miRNA and gene expression up/down analysis</li>
<li>miRNA and gene expression correlation</li>
</ul>

<h4>Improvements</h4>
<ul>
<li>Volcano plot y-axis logarithm changed from ln to log10</li>
</ul>

<h4>Bug fixes</h4>
<ul>
<li>Fixed: ArrayIndexOutOfBoundsException when loading session</li>
<li>Installation tool fixes</li>
</ul>


<h3 id="1.4.2">Version 1.4.2 <em>(11/Feb/2010)</em></h3>

<h4>New features</h4>
<ul>
<li>Ready-made analysis workflows for miRNA and proteomics data</li>
<li> New tool: Statistics / Correlate miRNA with target expression</li>
</ul>

<h4>Improvements</h4>
<ul>
<li>Clustering / KNN classification: Improved validation of classifiers with a test set</li>
<li>Hierarchical clustering: Color scheme has been changed to blue-red in order to cater for color blind users</li>
<li>Possibility to add gene / protein annotations to visualizations</li>
</ul>
 

<h3 id="1.4.1">Version 1.4.1 <em>(22/Jan/2010)</em></h3>

<h4>New features</h4>
<ul>
<li>Pathways / GO enrichment for miRNA targets</li>
<li>Pathways / KEGG enrichment for miRNA targets</li>
<li>Utilities / Import from ArrayExpress</li>
<li>Quality control / Affymetrix exon arrays - using RLE and NUSE</li> 
<li>Support for UniProt identifiers added to pathway tools</li>
<li>3D scatter plot for PCA: Samples can be colored based on phenodata columns which contain text</li>
<li>Migration tool (migrate.sh)</li>
</ul>

<h4>Bug fixes</h4>
<ul>
<li>Fixed: Wrong path for environment.xml in setup.sh</li>
<li>Analysis script bug fixes</li>
<li>Visualisation bug fixes</li>
</ul>
        
    

<h3 id="1.4.0">Version 1.4.0 <em>(10/Dec/2009)</em></h3>

<p><em>The major improvement in version 1.4.0 is update to R 2.9, which also has allowed to support
some new applications areas. Server side installation has been improved and server environment made
more robust. Some client quirks are also ironed out.</em></p>

<h4>New features</h4>
<ul>
<li>Complete R 2.9.x support</li>
<li>Updated annotation packages: Bioconductor 2.4, BrainArray v12 entrezg, X:map exon cdfs v1.1</li>
<li>Complete Affymetrix Gene ST support</li>
<li>Agilent miRNA normalization</li>
<li>Chipster environment build tool (setup.sh)</li>
<li>Configuration option to disable proxy bypass</li>
<li>Configurable tool root directory in compute service</li>
</ul>
    
<h4>Improvements</h4>
<ul>
<li>Uniprot ID support to Web services tools for pathway analysis</li>
<li>Improved Web services tools for pathway analysis (more informative output, also in table format)</li>
<li>Caching of result files for better performance</li>
<li>No warning about unsaved session shown when quick link panel is showing</li>
<li>Gene count limit added to hierarchical clustering to prevent oversized analysis jobs</li>
<li>Ordinary PCA 3d scatter is disabled when by sample PCA is done</li>
<li>Guard against extra spaces in copy/paste parameter panel</li>
<li>Fixed type font to Show source screen</li>
</ul>

<h4>Bug fixes</h4>
<ul>
<li>Fixed: Import tool &quot;change decimal separator&quot; does not work</li>
<li>Fixed: In some rare cases session saving fails and results into corrupted session file</li>
<li>Fixed: Workflows occasionally fail with no tool specific error, when using data from loaded session</li>
<li>Fixed: Excessive file transmissions can hang the message broker</li>
</ul>


<h3 id="1.3.0">Version 1.3.0 <em>(16/Jul/2009)</em></h3>

<p><em>Version 1.3.0 contains improvements mostly for the server side. Client side improvements include 
new analysis tools for Gene ST arrays, improvements to several visualisation, quick menu for loading data
and better blocking (what happens when client is busy working on something). There are also
new Web services integrated as analysis tools.</em></p>

<p><em>On server side, both the directory layout and configuration is completely redesigned. It is more consistent,
simpler and easier to work with. Both security and reliability of the server improvement have been improved.</em></p>

<h4>New features</h4>

<ul>
<li>Affymetrix Gene ST normalisation</li>
<li>Integrated Web Services as analysis tools: IntAct, Reactome, CPDB</li>
<li>Select all key shortcut for the workflow view</li>
<li>Row selection support for Expression Profile</li>
<li>HC selection system integrated to rest of the client</li>
<li>More information shown on progress bar when only one job is run</li>
<li>Nice and safe GUI blocking: everything is greyed out</li>
<li>File transfer progress shown in Task list</li>
<li>Quick links for dataset view</li>
<li>Automated tool for workflow unit testing</li>
<li>Internally opening sessions and workflows over URL</li>
<li>Repository of ready-made workflows bundled with client</li>
<li>Cluster colors to PCA visualisation</li>
<li>Zoom for HC</li>
<li>Completely refactored directory layout for client and server components</li>
<li>Lightweight manager application</li>
<li>Automatic upgrade tool for server installations</li>
<li>Better support for BeanShell jobs</li>
</ul>

   
<h4>Improvements</h4>

<ul>
<li>Progress bar shows file transfer progress</li>
<li>Blocking UI operations shown in application title</li>
<li>Bundled users file</li>
<li>Windows wrapper support (experimental)</li>
<li>Removed unused colors from HC visualisation property dialog</li>
<li>Less strict selection for 3D scatterplot colors</li>
<li>Changed 3D scatterplot color scale</li>
<li>3D PCA color selection allows floats</li>
<li>Volcano plot: guard against rounding to 0</li>
<li>Refactored Viski visualisation library</li>
<li>Configuration supports empty lists</li>
<li>Implemented new and improved configuration system: a lot smaller configuration files, supports default overriding, robust checking of values</li>
<li>Guard against unauthorised file brokering</li>
<li>Nicer filebroker index page</li>
<li>Client working directories follow common conventions on Mac</li>
<li>Proper gene name showing to HC visualisation</li>
<li>Max number of jobs is configured by config tool</li>
</ul>

<h4>Bug fixes</h2>

<ul>
<li>When datasets are deleted disk space is made free immediately</li>
<li>Obsolete operations are handled when opening session</li>
<li>Fixed: Session saving can fail after deleting dataset</li>
<li>Fixed: ID issue that sometimes made complex session saving fail</li>
<li>Fixed: Deleting dataset between derived data and phenodata can make session loading fail</li>
<li>Using sort and keyboard does not confuse phenodata editor</li>
<li>Phenodata columns can be removed before one is added</li>
<li>Expression profile selection bug fixed</li>
<li>Fixed: scatterplot zooms if selection rectangle is done partly outside plot</li>
<li>Fixed: ExpressionProfiles can be selected for dataset with zero rows</li>
<li>Fixed: Expression profiles visualisation can cause too many files open</li>
<li>Fixed:Phenodata can't be edited without double clicking on cell</li>
<li>Fixed client threading so that JGraph does not occassionally throw exceptions on the background</li>
<li>Better wording in PCA visualisation</li>
<li>ActiveMQ broker might on OutOfMemoryException with certain type of load, fixed by reconfiguring threading</li>
<li>Failover works reliably: broker threading issue caused it to fail sometimes</li>
<li>Services are shut down more gracefully (shutdown hook that closes broker connection)</li>
</ul>
  
   


<h3 id="1.2.4">Version 1.2.4 <em>(27/Mar/2009)</em></h3>

<h4>New features</h4>
<ul>
<li>Automatic proxy bypass for local networks with non-compliant HTTP proxy servers</li>
<li>New analysis tools for miRNA annotation with target predictions, classification and ordination, gene name handling, and more</li>
<li>Support for Illumina Human HT-12 chips</li>
</ul>

<h4>Bug fixes</h4>
<ul>
<li>Launching Chipster from a desktop icon (instead via browser) does not prevent updating</li>
<li>Venn diagram's create dataset feature works also for plain gene lists</li>
</ul>

<h3 id="1.2.3">Version 1.2.3 <em>(15/Jan/2009)</em></h3>

<h4>Improvements</h4>
<ul>
<li>Possibility to add new scripts with custom scripts functionality</li>
</ul>

<h3 id="1.2.2">Version 1.2.2 <em>(08/Jan/2009)</em></h3>

<h4>New features</h4>
<ul>
<li>New scripts: Statistics/ Adjust P-values, Normalization/ Normalize to specific samples, Normalization/ Illumina SNP arrays, Statistics/ Association analysis, Utilities/ Combine probes to genes</li>
</ul>

<h4>Improvements</h4>
<ul>
<li>HC visualisation usability improvements</li>
<li>Promoter analysis: if there are several probe sets for the same transcript (RefSeq id), the promoter sequence is used only once</li>
<li>Improvements to visualisations: Venn diagram, 3D scatter plot, Volcano plot, Phenodata editor</li>
<li>Increased fileserver cache lifetime</li>
</ul>

<h4>Bug fixes</h4>
<ul>
<li>Made client to use release version of ActiveMQ library instead of snapshot</li>
<li>Handling for renamed/missing parameters when loading session</li>
<li>Filter by CV range bug</li>
</ul>


<h3 id="1.2.1">Version 1.2.1 <em>(11/Nov/2008)</em></h3>


<h4>New features</h4>
<ul>
<li>Interactive Volcano plot</li>
</ul>

<h4>Improvements</h4>
<ul>
<li>Annotation support for new chip types: Illumina Human V3, Illumina Mouse V2, Agilent Zebrafish V2, Agilent Rice</li>
<li>Improvements to analysis tools: Statistics/ Linear modelling, Statistics/ NMDS, Statistics/ One sample tests, Utilities/ Calculate descriptive statistics, Preprocessing/ Filter by expression, Normalization/ Agilent 1-color</li>
<li>Made merge tables not require expression values</li>
<li>Made "identifier" to be supported as alternative row name column</li>
<li>Improved compute service stability</li>
</ul>

<h4>Bug fixes</h4>
<ul>
<li>Parsing specially formatted analysis scripts hanged</li>
<li>Agilent normalisation produced columns without .tsv suffix</li>
<li>Jetty could fail with dispatch error</li>
</ul>

<h3 id="1.2.0">Version 1.2.0 <em>(13/Oct/2008)</em></h3>

<p><em>Version 1.2.0 introduces new advanced analysis tools (Illumina lumi pipeline, SAFE pathways, ROTS statistical tool) and 
major improvements to user interface. User interface has been thoroughly analysed and many improvements made to make user
experience smoother. Also new key features have been added: old limited workspace system has been replaced with session
system, own sample names can be used and gene names are fetched in the beginning. Visualisation system has been 
reworked to support simultanous visualisations and to synchronise selections across visualisation. Datasets can be created
from selections made with visualisations.</em></p>

<p><em>Session system allows user to save the ongoing progress (all data) to separate files that are compressed
for efficient storage. Sessions also include now full metadata, so they can be used together with workflow and analysis history 
tools. For improved remote usage sessions also record remoting information, meaning that data is not transferred again even when 
client is restarted or session moved to another client.</em></p>

<h4>New features</h4>
<ul>
<li>New session system (replaces workspaces): sessions saved to compressed files</li>
<li>Support for loading old format workspaces</li>
<li>New analysis tools: Illumina lumi pipeline, SAFE pathways, ROTS statistical tool</li>
<li>Small analysis utility tools (delete columns etc.)</li>
<li>Gene names fetched already in the normalisation</li>
<li>Possibility to change shown sample names from &quot;description&quot; column in phenodata</li>
<li>Possibility to open visualisation to new windows for comparisons etc.</li>
<li>Universal selection (same selection shown in all visualisation etc.)</li>
<li>Possibility to filter datasets by selection</li>
<li>Easier phenodata linking</li>
<li>Bigger font option (View->Text size) for demo and teaching purposes</li>
<li>Check for too large visualisations (long running)</li>
<li>Demo mode for 3D scatterplot</li>
<li>New installation and configuration tool (server side feature)</li>
<li>Server environment configuration utility tool (server side feature)</li>
<li>Automatic expiration to local accounts (server side feature)</li>
</ul>

<h4>Improvements</h4>
<ul>
<li>Colored expression profile visualisation (GeneSpring style)</li>
<li>Real sample names stored to phenodata</li>
<li>Workflow running and saving from any location in directory structure</li>
<li>Remote information stored to sessions so data is not transferred again even when client is restarted</li>
<li>Analysis history with source code available in saved sessions also</li>
<li>Session save/load dialogs</li>
<li>Changed annotation button to use R/Bioconductor annotations</li>
<li>3 digit internal chip names</li>
<li>Better use of user interface (Java Swing) conventions in dataset folder view</li>
<li>Visualisation of empty dataset does not cause a confusing error any more</li>
<li>Improved &quot;Complete the rest&quot; in import tool</li>
<li>Workflow view usability improvements: automatic zooming, better placing of items</li>
<li>More space given for the workflow view</li>
<li>Minor Task manager improvements</li>
<li>Faster scrolling to dataset folder view</li>
<li>Better dataset grouping</li>
<li>Easier to use link menus</li>
<li>Reversed heatmap colors</li>
<li>Implemented .tre extension for hierarchical clustering to clarify visualisation options</li>
<li>Increased job limit to 10</li>
</ul>

<h4>Bug fixes</h4>
<ul>
<li>Illumina chip names did not come from column names</li>
<li>File grouping broken after deleting datasets</li>
<li>Parameters in notes panel do not update under some specific conditions</li>
<li>COSMO result file is interpreted also as a HC result</li>
<li>Workflow view jumps unnecessarily in some situations</li>
<li>Misleading warning about binary content with Genepix files</li>
<li>And many other small bugs fixed</li>
</ul>

<h3>Version 1.1.3 <em>(01/Sep/2008)</em></h3>


<h4>Bug fixes</h4>
<ul>
<li>Generated phenodata is not bound properly to actual dataset</li>
<li>Wizard groups are confused if there are more then 9</li>
<li>Error in standalone client (non Web Start) classpath</li>
<li>HC colors were reversed compared to common conventions</li>
</ul>

<h3>Version 1.1.2 <em>(14/Aug/2008)</em></h3>

<h4>Bug fixes</h4>
<ul>
<li>Phenodata help button sometimes invisible</li>
</ul>
    
<h3>Version 1.1.1 <em>(13/Jun/2008)</em></h3>

<h4>New features</h4>
<ul>
<li>Config option to disable HTTP chunked mode</li>
</ul>

<h4>Bug fixes</h4>
<ul>
<li>GeneCruiser annotation result problems</li>
</ul>


<h3>Version 1.1.0 <em>(02/Jun/2008)</em></h3>

<h4>New features</h4>
<ul>
<li>Branched workflows</li>
<li>Sequence analysis</li>
<li>Better parser for HC trees</li>
<li>Parameters visible also when data is not selected</li>
<li>Guard against loading ZIPs or other large binary files with import tool</li>
<li>Autoexpiration to local accounts</li>
</ul>

<h4>Improvements</h4>
<ul>
<li>Update to R 2.6</li>
<li>Workflow API</li>
<li>Confusing logarithms in scatter plots</li>
<li>Client startup does not fail with faulty descriptions</li>
<li>Fixed size font to TXT visualisation</li>
<li>Small import tool improvements</li>
<li>Configuration version check</li>
</ul>

<h4>Bug fixes</h4>
<ul>
<li>Fixed problem with import tool & many Illumina files</li>
<li>Details field in error dialog made non-editable</li>
<li>Directory browsing disabled on file broker</li>
<li>TaskCount NullPointer at startup</li>
<li>Known filetypes were not imported directly</li>
</ul>


<h3>Version 1.0.4 <em>(16/Apr/2008)</em></h3>

<h4>Improvements</h4>
<ul>
<li>Help mapping fixes</li>
<li>Removed autolayout from workflow view</li>
<li>Friendlier zip warning dialog</li>
</ul>

<h4>Bug fixes</h4>
<ul>
<li>Some files cannot be opened after loading workspace</li>
<li>Bug in history save button</li>
<li>Bug in workspace saving</li>
</ul>


<h3>Version 1.0.3 <em>(11/Apr/2008)</em></h3>

<h4>Improvements</h4>
<ul>
<li>Always create custom-scripts dir</li>
<li>Failed login reported</li>
</ul>

<h4>Bug fixes</h4>
<ul>
<li>File choosers are slow with WinXP&Zips</li>
<li>GeneCruiser AMIGO links broken</li>
<li>Workspace snapshot saving fails on a fresh environment</li>
<li>Glitches in help page mapping</li>
</ul>


<h3>Version 1.0.2 <em>(26/Mar/2008)</em></h3>

<h4>New features</h4>
<ul>
<li>Nagios check for multiple analysers</li>
</ul>

<h4>Improvements</h4>
<ul>
<li>Updated unreliable ActiveMQ 5.0</li>
<li>New databean does not cause deselection (and visualisation disappearing)</li>
</ul>


<h3>Version 1.0.1 <em>(20/Mar/2008)</em></h3>

<h4>Improvements</h4>
<ul>
<li>Refactored visualisation threading</li>
<li>Replaced buggy JRE LinkedBlockingQueue</li>
<li>Improved wording in client menus and dialogs</li>
</ul>

<h3>Version 1.0.0 <em>(14/Mar/2008)</em></h3>

<h4>Improvements</h4>
<ul>
<li>Read permission check to broker</li>
<li>Better data type checking for workflows</li>
<li>Updated help mappings</li>
</ul>

<h4>Bug fixes</h4>
<ul>
<li>Visualisation threading improved</li>
<li>Export dataset or folder - multiple files with same name handled correctly</li>
<li>"Stream closed" exceptions</li>
<li>Visualisation parameter side panel behaves strangely</li>
<li>Saving workspace sometimes fails</li>
<li>Hierarchical clustering: index error fixed</li>
<li>Deleting non-empty folders fails</li>
<li>Saved workflows not visible</li>
<li>NullPointerException after editing phenodata</li>
<li>Double columns in phenodata editor</li>
<li>Opening visualisation acts randomly</li>
<li>Running workflow fails</li>
<li>HC visualisation instability issues</li>
<li>Null More helps</li>
</ul>


<h3>Version 0.3.1 <em>(28/Feb/2008)</em></h3>

<h4>New features</h4>
<ul>
<li>New analysis tools: ordination-nmds, filter-cv, qc-agilent-one-color, qc-agilent</li>
</ul>

<h4>Bug fixes</h4>
<ul>
<li>Dataset delete fails</li>
<li>Analysis script bug fixes</li>
</ul>


<h3>Version 0.3.0 <em>(22/Feb/2008)</em></h3>

<h4>New features</h4>
<ul>
<li>Task manager UI component</li>
<li>Handling for binary .CEL-files</li>
<li>Separate authenticator</li>
<li>Restrictions on file size and job count</li>
<li>Analyser distribution</li>
<li>Warn if the user is about the close Chipster</li>
</ul>

<h4>Improvements</h4>
<ul>
<li>Production level messaging settings</li>
<li>Properly packaged fileserver</li>
<li>Replaced JavaHelp with HTML based help</li>
<li>Better selection management</li>
<li>Configured broker to authorise messages to authenticated topic</li>
<li>Tree panel look and feel improved</li>
<li>Better error dialog</li>
<li>About in Help improved</li>
</ul>

<h4>Bug fixes</h4>
<ul>
<li>Jetty server dies in some cases</li>
<li>Script annotate-genelist2 does not work</li>
<li>Fix graph view</li>
<li>Fix workflow view layout</li>
<li>Wizard visualisation fails</li>
<li>Snapshot saving can fail and is ambiguous</li>
<li>Harmless NPE in phenodata editor</li>
<li>Datasets cannot be deleted from tree or workflow views</li>
<li>Weeder's result seqs.html fails to open with double click</li>
<li>Import fails in some cases</li>
</ul>

<h3>Version 0.2.2 <em>(10/Dec/2007)</em></h3>

<h4>Improvements</h4>
<ul>
<li>Converted wizard to plain Java</li>
</ul>

<h4>Bug fixes</h4>
<ul>
<li>File choosers are slow with WinXP&Zips</li>
<li>Generating new phenodata fails at client</li>
<li>Memory leak in visualisation</li>
<li>Fixes to scripts</li>
</ul>


<h3>Version 0.2.1 <em>(29/Nov/2007)</em></h3>

<h4>Bug fixes</h4>
<ul>
<li>User notes with newlines break workspaces</li>
<li>Metacolumn selector does not work with stat-linear-model</li>
</ul>


<h3>Version 0.2.0 <em>(28/Nov/2007)</em></h3>
 
<h4>New features</h4>
<ul>
<li>Persistent data sessions</li>
<li>Parameter type for input dataset selection</li>
<li>Parameter type for selecting phenodata field</li>
<li>Histogram: ability to choose for what chip the data is shown</li>
<li>Line and multiline graphs</li>
<li>Box plot (with R)</li>
<li>Gene selection to 2D scatterplot</li>
<li>Production level authentication system (JAAS, Unix auth.)</li>
<li>User editable workflows</li>
<li>Analysis script override</li>
</ul>

<h4>Improvements</h4>
<ul>
<li>JobExecutor refactoring</li>
<li>Hierarchical clustering should show original chip names</li>
<li>orig_name column not always generated to phenodata</li>
<li>Histogram and scatterplots for non-Affy data</li>
<li>Disabled linking from Phenodata</li>
<li>Bigger window for viewing the R error message</li>
<li>Possibility to use import tool for Illumina files (csv)</li>
<li>Gene name tooltip to expression profile</li>
<li>Cluster number made visible to cluster profiles visualisation</li>
<li>Illumina normalisation refactoring</li>
<li>3d-scatterplot tooltips</li>
<li>Expression profile look</li>
</ul>

<h4>Bug fixes</h4>
<ul>
<li>HC samplewise clustering looks bad</li>
<li>GeneCruiser annotation window</li>
<li>Fix: user notes appear in the history only after another dataset has been selected</li>
<li>Child-relationships are broken in databeans</li>
<li>Normalisation for multichip Illumina is not available</li>
<li>Histogram bug</li>
<li>Affy wizard fails</li>
<li>Visualisation fails for merged data</li>
<li>Error on saving workflow: chipster-script could not be created</li>
<li>Metacolname parameter defaults don't work</li>
<li>HC did not work fully</li>
<li>Utilities / Average replicate chips: edit parameters does not work</li>
</ul>


<h3>Version 0.1.3 <em>(31/Aug/2007)</em></h3>

<h4>New features</h4>
<ul>
<li>Support for large data</li>
</ul>

<h4>Bug fixes</h4>
<ul>
<li>Analyser hangs with multimegabyte data</li>
</ul>


<h3>Previous versions</h3>

<p>There are no release notes for internal development versions 0.1.1 <em>(08/May/2006)</em>, 0.1.0 <em>(10/Apr/2006)</em>,
 0.0.2 <em>(03/Feb/2005)</em> and 0.0.1 <em>(20/Jan/2005)</em> (there is no 
version 0.1.2 due to change in version numbering scheme).<p>



    <!-- FOOTER  -->
    <div id="footer_wrap" class="outer">
      <footer class="inner">
	<!--
        <p class="copyright">Chipster maintained by <a href="https://github.com/chipster">chipster</a></p>
        <p>Published with <a href="http://pages.github.com">GitHub Pages</a></p>
	-->
      </footer>
    </div>

    

  </body>
</html>
